---
title: "The Natural History of Clonal Haematopoiesis"
subtitle: "Simulations and modelling"
author: 
    - Jos√© Guilherme de Almeida 
    - Moritz Gerstung
output: 
  html_document:
    theme: lumen
    highlight: tango
    code_folding: hide
    df_print: paged
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 4
    number_sections: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,fig.align = "center")
```

```{r import_functions, include=FALSE}
source("Scripts/vaf_dynamics_functions.R")
set.seed(42)
source("Scripts/prepare_data.R")
dir.create("figures/simulations/",showWarnings = F)
```

# Validating inference through simulations 

Here we present the modelling work that validates this approach. We simulate several Fischer-Wright dynamics with passenger and driver mutations across a grid of different driver mutation rates ($0.1*10^{-9}$ to $40.0*10^{-9}$) and fitness values ($0.001$ to $0.040$) per generation for a fixed population of 200000 HSC. We assume that HSCs divide 13 times *per* year, yielding for 100 years, 1300 generations. An example of some of these runs is presented below. We also assume that 10 driver and 200 passenger sites exist for each of these mutation rates and fitness values.

```{r get_all_runs}
all_files <- list.files(path = "/hps/nobackup/research/gerstung/josegcpa/projects/05VAF_DYNAMICS/hsc_output",
                        full.names = T) %>%
  lapply(function(x) list.files(x,pattern = 'csv',full.names = T)) %>%
  unlist
pop_size <- 2e5
```

We wish to determine if a model such as the one posited by us and in a similar scenario can accurately infer the fitness effects. As such, we need to simulate the same data collection circumstances, namely:

  * Sample one individual between 3 and 5 times with the earliest timepoint being sampled from a Gamma distribution fitted to the first timepoints from the collected data and the following times separated by approximately 3 years (40 generations)
  * Coverage is also simulated according to what was observed in our data, simmulating this using a Gamma distribution fitted to the coverage for all timepoints
  * Sampling at each timepoint for each individual is done by sampling from a beta binomial distribution with the sampled coverage as the number of trials, the probabilities as the allele frequencies of the driver mutations, and a fixed technical dispersion (with $\alpha = \frac{p\beta}{1-p}$ and $\beta=dispersion$). We only account for drivers because we are working with targetted sequencing, which focuses on mutations known to be associated with CH

## Inspecting and fitting distributions for age and coverage

```{r preprocessing_data}
plot(main="Coverage density",density(full_data$TOTALcount))
full_data %>%
  select(SardID,Age) %>%
  distinct %>% 
  group_by(SardID) %>%
  filter(Age == min(Age)) %>%
  ungroup %>%
  select(Age) %>%
  unlist %>% 
  density %>%
  plot(main = "Age at first timepoint density")
```

```{r fitting_distributions}
min_age_data <- full_data %>%
  select(SardID,Age) %>%
  distinct %>% 
  group_by(SardID) %>%
  filter(Age == min(Age)) %>%
  ungroup %>%
  select(Age) %>%
  unlist

coverage_distr <- fitdistrplus::fitdist(full_data$TOTALcount,
                                        method = "mme",
                                        #probs = c(0.1,0.9),
                                        distr = "gamma")
min_age_distr <- fitdistrplus::fitdist(min_age_data,
                                       method = "mge",
                                       distr = "gamma")

plot(min_age_distr)
plot(coverage_distr)

sample_age <- function(n) {
  s <- rgamma(n,min_age_distr$estimate[1],min_age_distr$estimate[2])
  s <- ifelse(s < min(min_age_data),min(min_age_data),s)
  s <- ifelse(s > max(min_age_data),max(min_age_data),s)
  return(s)
}

sample_coverage <- function(n) {
  s <- rgamma(n,coverage_distr$estimate[1],coverage_distr$estimate[2])
  s <- ifelse(s < min(full_data$TOTALcount),
              min(full_data$TOTALcount),s)
  s <- ifelse(s > max(full_data$TOTALcount),
              max(full_data$TOTALcount),s)
  return(round(s))
}
```

While these fits are not perfect they are good enough to help us simulate the conditions under the experimental data was obtained.

## Simulating the sampling process

```{r load_simulation_data}
load("models/simulation_model.RData")
```

```{r visualise_dynamics,fig.width=20,fig.height=5}
mutation_trajectories_plot <- Data %>%
  ggplot(aes(x = Gen,y = VAF,group = paste(Individual,Mutation),colour = mutation_rate/1e-9)) +
  geom_line(alpha = 0.3) +
  theme_gerstung() + 
  theme(legend.position = "bottom") + 
  scale_color_material(name = "Mutation rate (x10^-9)",
                       palette = "light-blue") + 
  facet_wrap(~fitness,nrow = 3,scales = "free") + 
  ggtitle("Samples from our simulated cohort for each fitness value") + 
  scale_y_continuous(limits = c(0,0.5)) + 
  scale_x_continuous(c(min(Data$Gen),max(Data$Gen)))
mutation_trajectories_plot
```

We now need to set a model similar to the one described below, which is similar to the one used on the Sardinian cohort. The main difference here is that the model here presented does not have three distinct levels of a genetic effect (gene, domain and site) and has a single effect whose prior is the addition of the three previous priors ($N(0,0.1) + N(0,0.1) + N(0,0.1) = N(0,0.14)$). We also choose to not have mutations with $VAF > 0.45$ for all timepoints because this is an unrealistic scenario which we do not observe in our data.

## Assessing the quality of the fit

### MCMC convergence diagnostic

```{r diagnostics,fig.height=7,fig.width=10}
mcmc_trace(draws,regex_pars = "genetic")

rg_stat <- coda::gelman.diag(draws)
ess <- coda::effectiveSize(draws)
plot(rg_stat$psrf[,1],main = "Distribution of Rubin-Gelman Statistic",ylab = "Rubin Gelman Statistic",xlab = "Variable index")
plot(ess,main = "Effective sample sizes for all variables",ylab = "Effective sample size",xlab = "Variable index",ylim = c(0,max(ess)),axes=F)
axis(side = 2, at=seq(0,max(ess),by = 500),las=2)
axis(side = 1)
```

We can see clearly from analysing the draws for the genetic coefficients that the MCMC has converged. Additionally, the above plots shows Rubin-Gelman statistic values are very close to 1 for all variables and effective sample sizes are at least 1000 These 3 separate analysis combined together indicate that convergence has been achieved and that our estimates are acceptable.

### Goodness of fit (residual effect)

To assess the goodness of fit we assume that, for each trajectory, the probability density of our true counts under the inferred model - $p$ - should follow a uniform distribution if the fit is perfect. As such, the normal normal quantiles of $p$ should follow a normal distribution and the sum of their squares should follow a Chi-squared distribution. This allows us to have a "goodness of fit" for each trajectory varying between 0 (not good) and 1 (perfect), which can be stated as $s=CDF_{\chi^2}(\sum_{i=1}^{n}Q_{N(0,1)}(CDF_{BBinom(coverage,\alpha_{inferred},\beta_{inferred})}(counts))^2)$ (here $p=CDF_{BBinom(coverage,\alpha_{inferred},\beta_{inferred})}(counts)$). Using this, we now define a residual effect (RE) as $RE = 1 - s$ and 0.7 as a threshold for declaring a trajectory as good ($RE\lt0.7$) or bad ($RE\geq0.7$).

```{r calculate_gof,fig.height=3,fig.width=3}
all_predicted_variables <- variable_summaries(do.call(rbind,draws)) %>% 
  as_tibble() %>%
  mutate(variable_no = as.numeric(str_match(variable,'[0-9]+(?=\\])'))) %>% 
  mutate(true_fitness = unique(simulated_samples$fitness)[variable_no]) %>% 
  mutate(true_fitness_per_year = log(exp(true_fitness) ^ 13)) %>%
  spread(key = labels,value = values)
genetic_effect_inferred <- all_predicted_variables %>%
  subset(grepl('genetic',variable)) %>%
  arrange(variable_no)
clone_effect_inferred <- all_predicted_variables %>%
  subset(grepl('clone',variable)) %>%
  arrange(variable_no)
individual_effect_inferred <- all_predicted_variables %>%
  subset(grepl('individual',variable)) %>%
  arrange(variable_no)

beta_val <- all_predicted_variables %>%
  subset(variable == 'beta_coefficient')

min_age <- min(Data$Gen / 13 * age_mult)

r_values <- data.frame(
  fitness = Data$fitness,
  fitness_factor = Data$fitness_factor,
  individual_factor = Data$individual_factor,
  genetic_effect = t(genetic_effect_inferred$mean %*% genetic_effect_indicator),
  genetic_effect_005 = t(genetic_effect_inferred$`0.05` %*% genetic_effect_indicator),
  genetic_effect_095 = t(genetic_effect_inferred$`0.95` %*% genetic_effect_indicator),
  clone_effect = clone_effect_inferred$mean[Data$individual_factor],
  clone_effect_005 = clone_effect_inferred$`0.05`[Data$individual_factor],
  clone_effect_095 = clone_effect_inferred$`0.95`[Data$individual_factor],
  individual_offset = individual_effect_inferred$mean[Data$individual_factor],
  individual_offset_005 = individual_effect_inferred$`0.05`[Data$individual_factor],
  individual_offset_095 = individual_effect_inferred$`0.95`[Data$individual_factor],
  true_clone_age = Data$GenAtCloneFoundation/13,
  true_count = Data$sample,
  coverage = Data$coverage,
  age = Age,
  Mutation = Data$Mutation,
  Individual = Data$Individual) %>%
  as_tibble() %>% 
  mutate(mu_val = inv.logit((genetic_effect + clone_effect)*age + individual_offset)/2) %>%
  mutate(
    tail_prob = extraDistr::pbbinom(
    q = true_count,
    size = coverage,
    alpha = (mu_val * beta_val$mean)/(1 - mu_val),
    beta = beta_val$mean)) %>% 
  group_by(individual_offset) %>%
  mutate(sum_stat = all(tail_prob > 0.025 & tail_prob < 0.975))

gof_plot_values <- r_values %>%
  na.omit() %>% 
  select(Individual,fitness,sum_stat) %>%
  distinct() %>%
  mutate(bad = sum_stat == F) %>%
  mutate(bad = factor(bad,levels = c(T,F),labels = c("No outliers","One or more\noutliers")))

explained_variants_plot <- gof_plot_values %>%
  mutate(fitness = cut(fitness,c(0.000,0.005,0.010,0.015,0.020,0.026),
                       c("0.001-0.005","0.005-0.010","0.010-0.015","0.015-0.020","0.020-0.025"))) %>% 
  group_by(fitness) %>%
  mutate(Total = length(unique(individual_offset))) %>% 
  group_by(fitness,bad) %>%
  summarise(N = length(unique(individual_offset)),
            Total = Total[1]) %>%
  ggplot(aes(x = fitness,y = N/Total,fill = bad)) + 
  geom_bar(alpha=1,
           stat = "identity") +
  theme_gerstung(base_size = 8) + 
  scale_x_discrete(expand = c(0,0)) + 
  scale_y_continuous(expand = c(0,0)) + 
  scale_fill_lancet(name=NULL) + 
  theme(legend.position = "bottom") + 
  ylab("Fraction") + 
  xlab("Fitness") 

explained_variants_plot_bar <- gof_plot_values %>% 
  group_by(bad) %>%
  summarise(N = length(bad)) %>%
  ungroup() %>%
  mutate(Freq = N / sum(N)) %>% 
  ggplot(aes(x = 1,
             y = Freq,
             fill = bad)) + 
  geom_bar(stat="identity") + 
  theme_gerstung() + 
  scale_x_discrete(expand = c(0,0)) + 
  scale_y_continuous(expand = c(0,0)) + 
  scale_fill_manual(values = rev(ggsci::pal_lancet()(2)),
                  breaks = c(F,T),
                  name = NULL,
                  #guide = F,
                  labels = c("Good fit","Poor fit")) +
  theme(legend.position = "bottom") + 
  xlab("") + 
  ylab("Proportion of trajectories")

explained_variants_plot
```

The graphic shown above shows that a fairly good amount of trajectories ($\sim87\%$) are explained by our model. We exclude the non-explained trajectories from the calculation of age at clone foundation.

### Genetic coefficients

```{r check_parameters,fig.height=2,fig.width=2}
predicted_variables <- variable_summaries(do.call(rbind,draws)) %>% 
  as_tibble() %>%
  subset(grepl('genetic',variable)) %>% 
  mutate(variable_no = as.numeric(str_match(variable,'[0-9]+(?=\\])'))) %>% 
  mutate(true_fitness = unique(simulated_samples$fitness)[variable_no]) %>% 
  mutate(true_fitness_per_year = true_fitness * 13 / age_mult) %>%
  spread(key = labels,value = values)
plot_min_max <- c(min(predicted_variables$HDPI_low,predicted_variables$true_fitness_per_year),
                  max(predicted_variables$HDPI_high,predicted_variables$true_fitness_per_year))
predicted_coefficients_plot <- predicted_variables %>%
  ggplot(aes(y = `0.50`,ymin = `HDPI_low`,ymax = `HDPI_high`,x = true_fitness_per_year)) + 
  geom_abline(slope=1,intercept = 0,linetype=2) +
  #geom_smooth(method = 'lm',formula = y ~ x,color = 'grey4',linetype=2,size=1,se=F) +
  geom_point(size = 0.5) + 
  geom_linerange(size = 0.2) +
  theme_gerstung(base_size = 6) + 
  theme(axis.title = element_text(size = 6)) + 
  xlab("Simulated genetic\nselection coefficient") + 
  ylab("Inferred genetic\nselection coefficient")

predicted_coefficients_plot
```

```{r correlation_inference}
cor_test_inference <- cor.test(predicted_variables$true_fitness_per_year,predicted_variables$`0.50`)
cor_test_inference
cat(paste("Rsquared =",cor_test_inference$estimate^2))
```

The coefficients closely agree with the data, with high correlation ($R^2>0.87$) between the inferred and the true coefficients.

### Age at clone foundation

The advantage of these simulations is that we have a very precise value for when the mutation first appeared. This allows us to assess how closely we can estimate the age at clone foundation.

```{r calculate_ages,fig.height=2.5,fig.width=5}
c_names <- colnames(draws$`11`)

offset_names <- grep('^individual_offset',c_names,perl = T)
genetic_c_names <- grep('^genetic_coef',c_names,perl = T)
clone_c_names <- grep('^clone_specific_coef',c_names,perl = T)

genetic_samples <- draws$`11`[,genetic_c_names]
clone_samples <- draws$`11`[,clone_c_names]
offset_samples <- draws$`11`[,offset_names]
beta_samples <- draws$`11`[,length(c_names)] %>%
  unlist()

sampling_idxs <- r_values %>%
  filter(sum_stat == T) %>% 
  group_by(
    fitness,fitness_factor,
    Individual,individual_factor,
    true_clone_age,
    Mutation
  ) %>% 
  summarise(
    minAge = min(age[which(round(true_count/coverage,4) > 0.005)]),
    maxAge = max(age),
    maxVAF = max(true_count/coverage),
    minVAF = min((true_count/coverage)[which(round(true_count/coverage,4) >= 0.005)])
  ) %>% 
  filter(!is.infinite(minAge)) %>%
  ungroup %>%
  mutate(true_clone_age_intervals = floor(true_clone_age/10) * 10)

age_data_list <- list()
i <- 0
for (y in list(c(2,5e4),c(13,5e4),c(2,2e5),c(13,2e5))) {
  i <- i + 1
  gen_time <- y[1]
  population <- y[2]
  age_estimates <- sampling_idxs %>% 
    apply(1,FUN = function(x){
      genetic_numeric <- as.numeric(x[2])
      clone_numeric <- as.numeric(x[4])
  
      b_genetic <- sample(genetic_samples[,genetic_numeric],1000,replace = F)
      b_clone <- sample(clone_samples[,clone_numeric],1000,replace = F)
      u <- sample(offset_samples[,clone_numeric],1000,replace = F)
      t1 <- as.numeric(x[7]) + min_age
      t1 <- t1 / age_mult
      age_distribution <- t0_adjusted(u,b_genetic+b_clone,gen_time / age_mult,population) + min_age
      age_distribution <- age_distribution / age_mult
      age_distribution <- ifelse(age_distribution < -1,-1,age_distribution)
      age_distribution <- ifelse(age_distribution > t1, t1, age_distribution)
      age_distribution <- age_distribution[!is.na(age_distribution)]
      D <- density(age_distribution)
      MAP <- D$x[which.max(D$y)]
      return(c(MAP,
               mean(age_distribution),
               quantile(age_distribution,c(0.05,0.10,0.16,0.5,0.84,0.90,0.95))))
    }) %>% 
    t %>% 
    as_tibble() %>%
    mutate(
      gen_time = gen_time,
      population = population
    )
  
  colnames(age_estimates) <- c("MAP","Mean","Q05","Q10","Q16","Q50","Q84","Q90","Q95","gen_time","population")
  age_data_list[[i]] <- age_estimates
}

age_data_full <- age_data_list %>% 
  lapply(function(x) cbind(x,sampling_idxs)) %>%
  do.call(what = rbind) %>%
  as_tibble()
age_data <- age_data_list[[4]] %>%
  cbind(sampling_idxs)

plot_min_max <- c(
  min(age_data$Q05,
      age_data$true_clone_age),
  max(age_data$Q05,
      age_data$true_clone_age)
)

estimated_ages_plot <- age_data %>% 
  ggplot(aes(x = true_clone_age,y = Q50)) + 
  geom_abline(slope=1,intercept=0,linetype=2,color = 'grey4') + 
  geom_point(aes(x = true_clone_age),
             colour = "grey",
             size = 0.5,alpha = 0.8) + 
  geom_linerange(aes(x = true_clone_age,
                     ymin = Q05,
                     ymax = Q95),
                 colour = "grey",
                 size = 0.2,
                 alpha = 0.8) + 
  geom_boxplot(
    aes(group = true_clone_age_intervals),
    alpha = 0.8,
    outlier.alpha = 0,
    size = 0.2
  ) + 
  scale_color_gradient(low = "pink",high = "red4",name = "Fitness") + 
  theme_gerstung(base_size = 6) + 
  ylab("Inferred age\nat clone foundation") +
  xlab("Simulated age\nat clone foundation") + 
  theme(legend.position = 'right') +
  coord_cartesian(ylim = plot_min_max) +
  scale_shape_discrete(name = "Generation time, population size") 

parameters_df <- r_values %>%
  select(genetic_effect,genetic_effect_005,genetic_effect_095,
         clone_effect,clone_effect_005,clone_effect_095,
         individual_offset,Mutation,Individual) %>%
  distinct
parameters_age_df <- merge(age_data,parameters_df,by = c("Mutation","Individual")) %>%
  mutate(minAge = minAge + min_age,
         maxAge = maxAge + min_age)
detection_threshold <- 0.002
N <- 2e5
age_threshold_accounting <- parameters_age_df %>%
  rowwise() %>%
  transmute(
    limit = onset_from_detection(clone_effect + genetic_effect,minVAF,mean(minAge,maxAge),g = 13,N = N),
    limit_005 = onset_from_detection(clone_effect_005 + genetic_effect_005,minVAF,mean(minAge,maxAge),13,N),
    limit_095 = onset_from_detection(clone_effect_095 + genetic_effect_095,minVAF,mean(minAge,maxAge),13,N),
    age_at_onset = Q50,
    coef = clone_effect + genetic_effect
  ) %>%
  subset(coef > 0) %>%
  mutate(plausible = limit >= 0,
         plausible_005 = limit_005 >= 0,
         plausible_095 = limit_095 >= 0)

table(age_threshold_accounting$plausible) / nrow(age_threshold_accounting)
```

```{r correlation_age}
cor_test_ages <- cor.test(age_data$true_clone_age,
                          age_data$Q50)

cat(paste("Mean absolute error =",mean(abs(age_data$Q50 - age_data$true_clone_age)),'\n'))
cat(paste("Rsquared =",cor_test_ages$estimate^2,'\n'))
cat(paste('p-value =',cor_test_ages$p.value))
```

```{r age_residuals,fig.height=3,fig.width=3}
age_data_post <- age_data
age_data_post$residuals_005 <- age_data_post$true_clone_age - age_data_post$Q05
age_data_post$residuals <- age_data_post$true_clone_age - age_data_post$Q50
age_data_post$residuals_095 <- age_data_post$true_clone_age - age_data_post$Q95
age_data_post <- age_data_post %>%
  mutate(residuals_min = ifelse(residuals_095 > residuals_005,residuals_005,residuals_095),
         residuals_max = ifelse(residuals_095 < residuals_005,residuals_005,residuals_095)) 
ggplot(age_data_post,aes(x = true_clone_age,y = residuals,color = fitness)) + 
  geom_point(size = 0.5) +
  theme_gerstung(base_size=6) + 
  theme(legend.position = "bottom",
        legend.key.height = unit(0.2,"cm")) + 
  scale_color_material()
```

The estimate for the ages, while not as good as the estimate for the coefficients, appears to be quite close to the true value. Additionally, on average, there is a difference of 6 years between our estimates and the true values. Residuals do not appear to be normally distributed, but a wider spread is observable for lower ages at clone foundation as per the figure above. Below we will observe that this is mostly caused by an unexpectedly high amount of time spent during the stochastic growth regime. 

Next, we verify how much changing population sizes and time per generation affects estimates. To assess this we have calculated the values for different population sizes - 50,000 and 200,000 HSC - and for different generation times - 2 and 13 generations per year. For this, we observe that the median difference between all 4 possible combinations, is $\sim19$ for estimates for the upper bound, $\sim17$ for the median value, and $\sim15$ for the lower bound.

```{r compare_population_sizes_gen_time,fig.width=4,fig.height=3}
age_data_full %>%
  group_by(Individual,true_clone_age) %>%
  summarise(m_005 = min(Q05),
            M_005 = max(Q05),
            m = min(Q50),
            M = max(Q50),
            m_095 = min(Q95),
            M_095 = max(Q95)) %>%
  na.omit() %>%
  mutate(Range_005 = M_005 - m_005,
         Range = M - m,
         Range_095 = M_095 - m_095) %>%
  ungroup() %>%
  summarise(Range_005Mean = mean(Range_005),
            RangeMean = mean(Range),
            Range_095Mean = mean(Range_095))

residual_plot <- age_data_full %>%
  group_by(Individual) %>%
  filter(length(unique(true_clone_age)) == 1) %>%
  ggplot(aes(x = reorder(Individual,true_clone_age),y = true_clone_age - Q50,
             color = paste(gen_time,population,sep='\n'))) + 
  geom_point(size = 0.5) +
  xlab("Index") +
  ylab("Residuals") +
  scale_colour_lancet(name = "Generations per year\nPopulation size") +
  theme_gerstung(base_size = 6) +
  theme(axis.text.x = element_blank(),
        legend.position = 'bottom',
        legend.key.size = unit(0,"cm"))

true_clone_age_plot <- age_data_full %>%
  group_by(Individual) %>%
  filter(length(unique(true_clone_age)) == 1) %>%
  ggplot(aes(x = reorder(Individual,true_clone_age),y = true_clone_age)) + 
  geom_point(size = 0.4) +
  theme_gerstung(base_size = 6) +
  theme(axis.text.x = element_blank()) + 
  xlab("") + 
  ylab("Age at onset")

age_data_full %>%
  ggplot(aes(x = true_clone_age,y = Q50,
             color = paste(gen_time,population,sep='\n'))) + 
  geom_abline(slope = 1) +
  geom_point(position = position_jitter(height = 0,width = 0),size = 0.4) +
  theme_gerstung(base_size = 6) +
  theme() + 
  xlab("Simulated age at onset") + 
  scale_colour_lancet(guide = F) +
  ylab("Inferred age at onset") + 
  theme(legend.position = "bottom",
        legend.key.size = unit(0,"cm")) + 
  geom_boxplot(aes(group = paste(true_clone_age,paste(gen_time,population,sep='\n')),
                   colour = paste(gen_time,population,sep='\n')),
               size = 0.25,outlier.size = 0.5) + 
  facet_wrap(~ sprintf("pop.size = %s, gen.time = %s",population,gen_time,sep=' '))

age_data_full %>%
  group_by(population,gen_time) %>%
  summarise(MAE = sum(abs(true_clone_age - Q50))/length(true_clone_age))

plot_grid(true_clone_age_plot,residual_plot,
          ncol = 1,align = "v",axis = "tlrb",
          rel_heights = c(0.4,1))
```

Below we inspect the clones for which the difference between predicted and true age at clone foundation differed more than by 20 years.

```{r inspecting_bad_ones, fig.width=10,fig.height=6.5}
root <- "/hps/nobackup/research/gerstung/josegcpa/projects/05VAF_DYNAMICS/hsc_output/hsc_%s_%s/r00%s.csv"

individuals_mutations <- age_data_post %>% 
  mutate(difference = abs(true_clone_age - Mean)) %>% 
  subset(difference > 20) %>% 
  arrange(difference)
file_ids <- str_split(individuals_mutations$Individual,pattern = '_') %>%
  do.call(what = rbind) %>%
  as.data.frame()
colnames(file_ids) <- c("fitness","mutation_rate","replicate")
file_ids <- file_ids %>% 
  mutate(mutation_rate = as.numeric(str_match(file_ids$mutation_rate,'[0-9.]+')),
         div = as.numeric(str_match(file_ids$mutation_rate,'[0-9.]+$')) - 9
  ) %>%
  mutate(mutation_rate = mutation_rate / (10^div)) %>% 
  select(-div)
file_ids$mutation <- individuals_mutations$Mutation %>%
  as.numeric
bad_predictions <- file_ids %>%
  apply(1,function(x) {
    x <- as.numeric(x)
    if (x[2] < 1) {
      mr <- as.character(format(x[2],nsmall=3))
    } else {
      mr <- as.character(paste0(0,format(x[2],nsmall=3)))
    }
    while (nchar(x[1]) < 6) {
      x[1] <- paste0(x[1],"0")
    }
    file_name <- sprintf(root,x[1],mr,x[3])
    tmp <- read_tsv(file_name,
                    col_names = c("Gen","Count","Clone","Mutation"),
                    col_types = list(col_number(),col_number(),
                                     col_number(),col_number())) 
    tmp <- tmp %>%
      mutate(bad = Mutation == x[4]) %>%
      mutate(fitness = x[1],
             mutation_rate = x[2],
             replicate = x[3])
    return(tmp)
  }) %>%
  do.call(what = rbind) %>% 
  mutate_all(as.numeric) %>% 
  mutate(bad = as.logical(bad)) %>% 
  filter(Mutation > 0) %>%
  group_by(Mutation,Gen,fitness,mutation_rate,replicate,bad) %>%
  summarise(Count = sum(Count)) %>%
  group_by(Mutation) %>%
  filter(length(Count) > 10) %>% 
  mutate(Driver = Mutation <= (N_DRIVERS)) %>% 
  mutate(Alpha = ifelse(Driver,0.9,0.05)) %>% 
  ungroup()

bad_predictions %>% 
  ggplot(aes(x = Gen,y = Count,color = Driver,alpha=Alpha,
             group = Mutation,
             size = as.numeric(Driver))) + 
  geom_hline(aes(yintercept = (1 - (1+fitness)^-1) / (1 - (1+fitness)^-pop_size) * pop_size),
             size = 1.0) +
  geom_line(aes(linetype = bad)) + 
  scale_y_continuous(trans = 'log',
                     breaks = c(10^seq(0,5),2e5),
                     limits = c(10,2e5)) + 
  scale_x_continuous(limits = c(0,1300)) + 
  scale_alpha(guide = F) + 
  theme_gerstung(base_size = 10) + 
  scale_color_d3(name = NULL,labels = c("Passenger","Driver")) + 
  theme(legend.position = "bottom") + 
  scale_size(labels = c(FALSE,TRUE),
             breaks = c(FALSE,TRUE),
             limits = c(FALSE,TRUE),
             range = c(0.5,1.5),
             guide = F) + 
  ylab("Count") + 
  facet_wrap(~ fitness + mutation_rate + replicate,nrow = 2,scales = 'free') + 
  ggtitle("Trajectories with large errors for the inferred age at clone foundation",
          "Mean absolute error > 20 years")
```

## Generating a final summary figure

```{r example_trajectory,fig.height=6,fig.width = 7}
fixation_probility <- function(s,N) {
  return(
    1 / (N * s)
  )
}

counts <- Data %>%
  group_by(fitness,mutation_rate,Replicate) %>%
  summarise(L = length(unique(Mutation_C)),
            M = max(VAF),
            minGen = abs(GenAtCloneFoundation - 600)[1]) %>%
  filter(L > 0) %>%
  filter(fitness >= 0.005)
chosen_one_full <- counts %>% 
  arrange(minGen,M,L)

good_examples <- chosen_one_full[c(58,201,212),]

plot_list <- list()
for (i in 1:nrow(good_examples[2,])) {
  try({
    chosen_one <- good_examples[2,] # 4
    params <- c(f = chosen_one$fitness,
                m = chosen_one$mutation_rate/(10^-9),
                r = chosen_one$Replicate)
  
    if (params[2] < 1) {
          params[2] <- as.character(format(params[2],nsmall=3))
        } else {
          params[2] <- as.character(paste0(0,format(params[2],nsmall=3)))
        }
    
    while (nchar(params[1]) < 6) {
      params[1] <- paste0(params[1],"0")
    }
    
    file <- sprintf(root,params[1],params[2],as.numeric(params[3]))

    df <- read_tsv(file,
                   col_names = c("Gen","Count","Clone","Mutation"),
                   col_types = list(col_number(),col_number(),
                                    col_number(),col_number()))
  
    fit_adv <- as.numeric(str_match_all(file,"[0-9.]+")[[1]][2])
    population_threshold <- fixation_probility(fit_adv,pop_size)
    time_stochastic <- 1 / fit_adv
    
    example_trajectory <- df %>% 
      filter(Mutation > 0) %>%
      group_by(Mutation,Gen) %>%
      summarise(Count = sum(Count)) %>%
      group_by(Mutation) %>%
      filter(length(Count) > 1) %>% 
      mutate(Driver = Mutation <= (N_DRIVERS)) %>% 
      mutate(Alpha = ifelse(Driver,0.9,0.02)) %>% 
      mutate(Break = Gen - c(0,Gen[1:max(0,(length(Gen)-1))]) > 20) %>% 
      mutate(Component = cumsum(Break)) %>%
      ungroup() %>%
      mutate(Mutation = paste(Mutation,Component,sep = '_')) %>%
      group_by(Mutation) %>%
      mutate(Count = Count/2e5) %>%
      mutate(Deterministic = ifelse(Count > population_threshold & Driver == T,
                                    Count,
                                    NA)) 
    
    tmp <- example_trajectory %>%
      subset(Driver == T) %>% 
      group_by(Mutation) %>% 
      summarise(minGen = min(Gen),
                C = Count[which.min(Gen)]) %>%
      filter(C > 5e-6)
    
    example_trajectory <- rbind(
      example_trajectory,
      tmp %>% 
        apply(1,function(x) example_trajectory %>% 
                subset(Gen == x[2] & Mutation == x[1]) %>% 
                mutate(Gen = Gen - 20,Count = 5e-6)) %>% 
        do.call(what = rbind))
    
    dead_clones <- example_trajectory %>% 
      group_by(Mutation) %>%
      filter(min(Count) == 5e-6,
             max(Count) < population_threshold,
             #max(Count) > population_threshold / 10,
             max(Gen) - min(Gen) > 80) %>%
      # filter(min(Count[
      #   which(Gen > Gen[min(which(Count > population_threshold / 6))])
      # ]) < 5e-4) %>%
      mutate(GenDiff = max(Gen) - min(Gen)) %>%
      ungroup %>% 
      filter(Driver == T) %>% 
      filter(GenDiff == max(GenDiff)) %>% 
      mutate(Count = ifelse(Gen < Gen[min(which(Count == 5e-6))],5e-6,Count))
    
    #repurposed_trajectory <- dead_clones[1:5,] %>%
    #  mutate(Count = c(5e-4,rep(5e-6,1000,4)),
    #         Gen = seq(1020,1100,20))
    #dead_clones <- dead_clones[dead_clones$Gen < 1000,]
    #dead_clones <- rbind(dead_clones,repurposed_trajectory)
    
    big_clone <- example_trajectory$Mutation[which.max(example_trajectory$Count)]
    big_clone_fixation <- example_trajectory$Gen[which.min(which(example_trajectory$Count > population_threshold))]
    
    simulated_gen_at_foundation <- example_trajectory %>%
      subset(Driver == T) %>%
      summarise(m = min(Gen),
                M = max(Count)) %>%
      ungroup() %>% 
      subset(M == max(M))
    inferred_gen_at_foundation <- example_trajectory %>%
      subset(is.na(Deterministic) & Driver == T & Mutation == simulated_gen_at_foundation$Mutation) %>% 
      summarise(m = max(Gen),
                M = max(Count)) %>%
      ungroup() %>% 
      subset(M == max(M)) %>% 
      mutate(InferredAgeAtCloneFoundation = m - time_stochastic) 

    age_at_clone_foundation_info <- data.frame(
      x = c(inferred_gen_at_foundation$InferredAgeAtCloneFoundation,
            simulated_gen_at_foundation$m),
      label = c("Inferred age at clone foundation",
                "Simulated age at clone foundation"),
      y_label = c(1e-5,1e-4),
      y_points = c(6.5e-6)
    )
    example_trajectory_neutral <- example_trajectory %>% 
      subset(Driver == F)
    example_trajectory_driver <- example_trajectory %>%
      subset(Driver == T)
    
    tmp <- example_trajectory_driver %>% 
      subset(!is.na(Deterministic) & Mutation == inferred_gen_at_foundation$Mutation) 
    driver_fit_coefficient <- lsfit(tmp$Gen / 13,logit(tmp$Count*0.999999))$coefficients
    points_meeting_threshold <- c(
      x=(logit(population_threshold) - driver_fit_coefficient[1])/driver_fit_coefficient[2],
      y=population_threshold
    )

    points_fit <- data.frame(
      x = c(points_meeting_threshold[1]-time_stochastic/13,points_meeting_threshold[1],1300/13),
      y = c(5e-6,points_meeting_threshold[2],inv.logit(100 * driver_fit_coefficient[2] + driver_fit_coefficient[1])),
      colour = "Fit"
    )
    points_fit_plot <- rbind(
      data.frame(
        x = seq(points_fit$x[1],points_fit$x[2],length.out = 100),
        y = seq(points_fit$y[1],points_fit$y[2],length.out = 100),
        colour = "Fit"
      ),
      data.frame(
        x = seq(points_fit$x[2],100,by = 0.1),
        y = inv.logit(seq(points_fit$x[2],100,by = 0.1) * driver_fit_coefficient[2] + driver_fit_coefficient[1]),
        colour = "Fit"
      )
    )

    example_trajectory_plot <- example_trajectory_neutral %>% 
      ggplot(aes(x = Gen / 13,
                 y = Count,
                 alpha=Alpha,
                 color = Driver,
                 group = Mutation)) + 
      geom_line(aes(size = as.numeric(Driver)),alpha = 1,colour = "#ffe08c") + 
      geom_hline(yintercept = population_threshold,size = 0.5) +
      geom_hline(yintercept = 5e-6,linetype="dotted",size=0.5) +
      # geom_line(data = dead_clones,
      #           size = 1.5,
      #           colour = 'orange',
      #           alpha = 1) +
      geom_line(
        data = points_fit_plot,
        inherit.aes = F,
        aes(x = x,y = y,colour = colour),
        #color = '#d68b82',
        size = 1
      ) +
      geom_point(
        data = example_trajectory_driver,
        alpha = 1
      ) + 
      geom_line(
        data = example_trajectory_driver,
        size = 0.4,
        alpha = 1,
        linetype = 5
      ) + 
      annotate(geom = 'label',
               x = 20 / 13,
               y = 10^((log10(population_threshold)+log10(1/2e5))/2),
               label = "Stochastic\ngrowth",
               hjust = 0,
               size = 1.9,
               label.size = unit(0,"cm"),
               label.r = unit(0,"cm"),
               alpha = 0.8) +
      annotate(geom = 'label',
               x = 100 / 13,
               size = 1.9,
               y = sqrt(population_threshold),
               label = "Deterministic\ngrowth",
               hjust = 0,
               label.size = unit(0,"cm"),
               label.r = unit(0,"cm"),
               alpha = 0.8) +
      annotate(geom = 'label',
               x = 1300 / 13,
               y = population_threshold,
               label = "drift threshold = 1/(Nf)",
               size = 1.9,
               hjust = 1.1,
               vjust = -0.2,
               label.size = unit(0,"cm"),
               label.r = unit(0,"cm"),
               label.padding = unit(0,"cm"),
               alpha = 0.8) +
      annotate(geom = 'label',
             x = 1300 / 13,
             size=1.9,
             y = population_threshold/100,
             label = "clones that do not\ncross the drift\nthreshold tend to\ndisappear",
             hjust = 1.1,
             vjust = -0.2,
             label.size = unit(0,"cm"),
             label.r = unit(0,"cm"),
             label.padding = unit(0,"cm"),
             alpha = 0.8) +
      annotate(
        geom = 'label',
        x = 600 / 13,
        size = 1.9,
        y = 0.03,
        label=(paste(expression("counts ~ e"^"ft"*""))),
        parse = TRUE,
        hjust = 1,
        vjust = -0.2,
        label.size = unit(0,"cm"),
        label.r = unit(0,"cm"),
        fontface="italic",
        alpha = 0.8) +
      geom_errorbar(
        inherit.aes = F,
        data = data.frame(
          ymin = c(min(example_trajectory$Count),population_threshold),
          ymax = c(population_threshold,1.0),
          x = c(20 / 13,100 / 13)
        ),
        aes(ymin = ymin, 
            ymax = ymax,
            x = x,
            group = x),
        color = 'black',
        size = 0.3) + 
      scale_y_continuous(trans = 'log10',
                         breaks = c(5e-6,c(10^c(-5:0))),
                         limits = c(5e-6,1),
                         labels = c("Single cell",10^c(-5:0)),
                         expand = c(0.01,0.0,0.0,0.0)) + 
      scale_x_continuous(expand = c(0,1)) +
      scale_alpha(guide = F) + 
      theme_gerstung(base_size = 6) + 
      scale_color_manual(name = NULL,
                         breaks = c(T,F,"Fit"),
                         drop = F,
                         values = c("#ab2503","#00468BFF","#d68b82"),
                         labels = c("Driver","Neutral","Fit")) + 
      theme(legend.position = "bottom",
            axis.title = element_text(size = 6)) + 
      scale_size(breaks = c(FALSE,T),
                 limits = c(FALSE,T),
                 labels = c("Neutral","Driver"),
                 range = c(0.4,0.45),
                 name = NULL) +
      ylab("VAF") +
      xlab("t (years)")
    
      plot_list[[file]] <- example_trajectory_plot
  })
}
```

```{r big_picture,fig.height=3,fig.width=6}
main_results <- plot_grid(
  example_trajectory_plot + theme(legend.position = c(0.5,0.9),legend.direction = "horizontal",
                                  legend.key.height = unit(0.2,"cm"),
                                  axis.text = element_text(size = 6)),
  plot_grid(
    predicted_coefficients_plot + 
      xlab("Simulated driver effect") + 
      ylab("Inferred driver effect") + 
      theme(axis.text = element_text(size = 6)),
    estimated_ages_plot + 
      xlab("Simulated age at onset") + 
      ylab("Inferred age at onset") + 
      theme(axis.text = element_text(size = 6)),
    ncol = 1
  ),
  nrow = 1,
  rel_widths = c(7,3.5)
)

main_results
ggsave(main_results,filename = "figures/simulations/simulation_plot.pdf",height = 3,width = 6,
       useDingbats=FALSE)
ggsave(example_trajectory_plot + theme(legend.position = "top"),
       filename = "figures/simulations/example_trajectory_plot.pdf",height = 3,width = 3)
ggsave(predicted_coefficients_plot,
       filename = "figures/simulations/predicted_coefficients_plot.pdf",height = 2,width = 2)
ggsave(explained_variants_plot + theme(legend.position = "top"),
       filename = "figures/simulations/explained_variants_plot.pdf",height = 2,width = 2)
ggsave(estimated_ages_plot + theme(legend.position = "right"),
       filename = "figures/simulations/estimated_ages_plot.pdf",height = 2,width = 2)
```

# Session details

```{r session_details}
cat(system("git log -n1", intern=TRUE), sep="\n")
```

```{r session_objects}
l <- ls()
data.frame(variable=l, Reduce("rbind",lapply(l, function(x) data.frame(classes=paste(class(get(x)),collapse = ','), size=format(object.size(get(x)), units="auto")))))
```

```{r session_packages}
sessionInfo()
```