---
title: "The Natural History and Fitness Landscape of Clonal Haematopoiesis"
subtitle: "Post-hoc analysis"
author: 
    - Jos√© Guilherme de Almeida
    - Moritz Gerstung
output: 
  html_document:
    theme: lumen
    highlight: tango
    code_folding: hide
    df_print: paged
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 4
    number_sections: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r import_functions, include=FALSE}
source("Scripts/vaf_dynamics_functions.R")
model_file_name <- "models/model_D.RDS"
set.seed(42)
```

# Post-hoc analysis

```{r making_dirs_loading_data}
model_id <- gsub('_$','',str_match(model_file_name,'model_[A-Z]+'))
load(file = sprintf("vaf_modelling_coefficients_%s.Rdata",model_id))
values_model <- readRDS(file = model_file_name)
c('prevalence','dnds',
  'phenotype_associations',
  'survival_analysis') %>%
  sapply(
    function(x) dir.create(sprintf("figures/%s/%s",model_id,x),showWarnings = F)
  )
source("Scripts/prepare_data.R")
```

## What if - recreating mutation prevalence

An interesting scenario is to use our growth coefficients and defined dynamics to see how this would affect mutation prevalence through different age brackets. 

```{r observing incidence, fig.width=4,fig.height=7}
INTERVAL <- 5
DETECTION_LIMIT <- 0.005
incidence <- r_values %>%
  subset(coverage > 0) %>%
  mutate(AgeGroup = floor(age/INTERVAL) * INTERVAL) %>%
  mutate(AgeGroupLabel = paste(AgeGroup,AgeGroup+INTERVAL,sep = '-')) %>%
  mutate(genes = str_match(genes,'[A-Z0-9]+')) %>%
  group_by(AgeGroup,AgeGroupLabel) %>%
  mutate(TotalIndividuals = length(unique(individual))) %>%
  ungroup() %>%
  group_by(AgeGroup,AgeGroupLabel,genes,individual) %>%
  summarise(MutPresence = any(true / coverage > DETECTION_LIMIT),
            TotalIndividuals = TotalIndividuals[1]) %>%
  group_by(AgeGroup,AgeGroupLabel,genes) %>%
  summarise(MutCount = sum(MutPresence),
            TotalIndividuals = TotalIndividuals[1]) %>%
  mutate(y = (MutCount) / (TotalIndividuals),
         ymax = qbeta(0.99,MutCount,TotalIndividuals - MutCount + 1),
         ymin = qbeta(0.01,MutCount,TotalIndividuals - MutCount + 1))

true_incidence <- incidence %>%
    subset(AgeGroup <= 90 & AgeGroup > 50) %>% 
    ggplot(aes(x = AgeGroupLabel,
               fill = genes,label = genes,
               y = y)) +
  theme_gerstung(base_size = 12) +
  scale_fill_manual(values = gene_colours,
                    name = NULL) + 
  geom_bar(stat = "identity",color = 'black') +
  scale_x_discrete() + 
  rotate_x_text() +
  xlab("Age ranges") + 
  ylab("Cummulative frequency") + 
  ggsave(useDingbats=FALSE,sprintf("figures/%s/prevalence/Prevalence.pdf",model_id),height = 7,width = 4)
```

```{r modelling incidence,warning=F}
n_points <- load_data() %>%
  select(SardID,Phase) %>%
  distinct() %>%
  group_by(SardID) %>%
  summarise(n = length(SardID)) %>% 
  group_by(n) %>%
  summarise(N = length(n)) %>%
  mutate(N = N / sum(N))
coverage <- r_values$coverage
coverage <- coverage[coverage != 0]
coverage_distribution <- fitdistrplus::fitdist(coverage,"gamma",method = "mme")
min_coverage <- min(coverage)
BETA <- values_model$beta_values[[1]][1,1]

# sample from age distribution and not uniform 

unique_combinations <- r_values %>%
  group_by(individual) %>%
  mutate(min_age = min(age),n_timepoints = length(unique(age)),
         max_age = max(age)) %>%
  ungroup() %>%
  select(individual,site,coefficient,b_clone,u_values,genes,min_age,n_timepoints,max_age,age) %>% 
  distinct %>%
  group_by(individual) %>% 
  mutate(Phase = paste("Phase",findInterval(age,unique(age)),sep = '_')) %>% 
  spread(key = "Phase",value = "age") 

unique_individuals <- r_values %>%
  select(individual) %>%
  unlist %>%
  unique

age_diff <- full_data %>% 
  select(SardID,Age) %>% 
  distinct() %>% 
  arrange(SardID,Age) %>% 
  group_by(SardID) %>% 
  mutate(Age = c(Age) - c(Age[2:length(Age)],Age[length(Age)])) %>% 
  subset(Age != 0) %>% 
  ungroup() %>% 
  summarise(AgeDiff = mean(Age)) %>% 
  unlist %>% 
  abs

samples_through_time <- list()

for (ind in unique_individuals) {
  tmp <- unique_combinations %>%
    subset(individual == ind)
  n_timepoints <- tmp$n_timepoints[1]

  ages <- tmp[,c(10:14)] - values_model$min_age
  ages <- ages[!is.na(ages)]
  inferred_counts <- ages %>%
    sapply(
      function(x) {
        tmp %>%
          as_tibble() %>%
          mutate(p = inv.logit((coefficient + b_clone) * x + u_values) * 0.5,
                 age = x,
                 coverage = round(rgamma(length(coefficient),
                                         shape = coverage_distribution$estimate[1],
                                         rate = coverage_distribution$estimate[2]))) %>%
          mutate(coverage = ifelse(coverage < min_coverage,
                                   min_coverage,
                                   coverage)) %>%
          mutate(counts = rbbinom(1,coverage,(BETA * p)/(1 - p),BETA))
      },
      simplify = F
    ) %>%
    do.call(what = rbind) %>%
    as_tibble() %>%
    mutate(sample = counts / coverage) %>%
    select(p,individual,site,sample,age,genes,counts,coverage)
  samples_through_time[[as.character(ind)]] <- inferred_counts
}

incidence_interpolation <- samples_through_time %>%
  do.call(what = rbind) %>%
  mutate(age = age + values_model$min_age) %>%
  mutate(AgeGroup = floor(age/INTERVAL) * INTERVAL) %>%
  mutate(AgeGroupTrue = AgeGroup) %>%
  mutate(AgeGroup = factor(paste(AgeGroup,AgeGroup+INTERVAL,sep = '-'),
                           levels = paste(sort(unique(AgeGroup)),sort(unique(AgeGroup)) + INTERVAL,sep = '-'))) %>%
  mutate(genes = str_match(genes,'[A-Z0-9]+')) %>%
  group_by(AgeGroup,AgeGroupTrue) %>%
  mutate(TotalIndividuals = length(unique(individual))) %>%
  group_by(AgeGroup,AgeGroupTrue,genes,individual) %>%
  summarise(MutPresence = any(sample > DETECTION_LIMIT),
            TotalIndividuals = TotalIndividuals[1]) %>%
  group_by(AgeGroup,AgeGroupTrue,genes) %>%
  summarise(MutCount = sum(MutPresence),
            TotalIndividuals = TotalIndividuals[1]) %>%  
  mutate(y = (MutCount) / (TotalIndividuals),
         ymax = qbeta(0.99,MutCount,TotalIndividuals - MutCount + 1),
         ymin = qbeta(0.01,MutCount,TotalIndividuals - MutCount + 1))

simulated_incidence <- incidence_interpolation %>%
  subset(AgeGroupTrue <= 90 & AgeGroupTrue > 50) %>% 
  ggplot(aes(x = AgeGroup,
               fill = genes,label = genes,
               y = y)) +
  theme_gerstung(base_size = 12) +
  scale_fill_manual(values = gene_colours,
                    name = NULL) + 
  geom_bar(stat = "identity",color = 'black') +
  scale_x_discrete() + 
  rotate_x_text() + 
  xlab("Age ranges") + 
  ylab("Cummulative frequency") + 
  ggsave(useDingbats=FALSE,sprintf("figures/%s/prevalence/PrevalenceSimulated.pdf",model_id),height = 7,width = 4)
```

```{r plotting_incidences,fig.width=8,fig.height=7}
plot_grid(true_incidence + ggtitle("True incidence") + scale_y_continuous(limits = c(0,2)),
          simulated_incidence + ggtitle("Simulated incidence") + scale_y_continuous(limits = c(0,2)),
          nrow = 1) + 
  ggsave(useDingbats=FALSE,sprintf("figures/%s/prevalence/PrevalenceBoth.pdf",model_id),width = 8,height = 7) 
```

```{r distance_incidences}
cosine_dist <- function(p1,p2) {
  return(sum(p1*p2) / (sqrt(sum(p1^2))*sqrt(sum(p2^2))))
}
tmp <- incidence_interpolation %>%
  select(AgeGroupLabel=AgeGroup,AgeGroup=AgeGroupTrue,genes,MutCount,TotalIndividuals,y,ymax,ymin)

tmp <- rbind(data.frame(tmp,Type = c("Simulated")),
             data.frame(incidence,Type = c("True"))) %>% 
  subset(AgeGroup <= 90) %>%
  subset(AgeGroup > 50)

COS_D <- list()
for (ag in unique(tmp$AgeGroupLabel)) { 
  tmp_2 <- tmp %>% 
    subset(AgeGroupLabel == ag) %>%
    select(-ymax,-ymin,-MutCount) %>%
    spread(value = "y",key = "Type") %>% 
    mutate(Simulated = ifelse(is.na(Simulated),0,Simulated),
           True = ifelse(is.na(True),0,True))
  COS_D[[ag]] <- c(ag,cosine_dist(tmp_2$Simulated,tmp_2$True))
}
COS_D <- do.call(rbind,COS_D)
tibble(`Age Group` = COS_D[,1],`Cosine Similarity` = COS_D[,2])
```

Using the trajectories we can replicate the observed prevalence to some extent. We overestimate the prevalence of a few mutations early on life (SF3B1, MYD88, GNB1, TP53). However, it is reassuring to see that the broad picture is till maintained, with high cosine distances between the proportions of individuals with each mutation remaining quite high (at least 0.94).

## dN/dS associations

A popular measure of evolutionary selection in cancer is the dN/dS ratio as per Martincorena, which uses the ratio of non-synonymous mutations to synonymous mutations normalised by nucleotide context as a proxy for evolutionary selection. However, what is observable is that there is no correlation between the expected values for dN/dS and growth coefficients; indeed, a few genes appear to have higher than expected growth coefficients when considering their relative prevalence in our cohort (i.e. U2AF1-Q157R and SRSF2-P95H). 

```{r dnds_comparison, fig.height=4,fig.width=8}
unique_gene <- values_model$sub_data %>%
  select(Gene,gene_numeric) %>% 
  distinct %>%
  arrange(gene_numeric) %>%
  select(Gene) %>%
  unlist
unique_site <- values_model$sub_data %>%
  select(amino_acid_change,site_numeric) %>% 
  distinct %>%
  arrange(site_numeric) %>%
  select(amino_acid_change) %>%
  unlist
unique_site_multiple <- values_model$sub_data %>%
  select(amino_acid_change,site_numeric_multiple) %>% 
  subset(!is.na(site_numeric_multiple)) %>%
  distinct %>%
  arrange(site_numeric_multiple) %>%
  select(amino_acid_change) %>%
  unlist %>% 
  as.character() %>% as.factor

values_b <- r_values %>%
  select(
    mean = coefficient,
    HDPI_low = coefficient_005,
    HDPI_high = coefficient_095,
    coefficient_var,
    site_name = site,
    gene = genes
  ) %>%
  distinct

dnds_data <- load_dnds()
dnds_site <- dnds_data$site %>%
  mutate(site_merge = paste(gene,aachange,sep = '-'))
coefficients_dnds <- values_b %>%
  mutate(genes_corrected = str_match(gene,"[A-Z0-9]+")) %>%
  mutate(site_merge = paste(genes_corrected,str_match(site_name,'[a-zA-Z0-9:>;\\.\\+]+$'),sep = '-')) %>% 
  merge(dnds_site,by = "site_merge") %>%
  subset(sig == T) %>%
  select(mean,HDPI_low,HDPI_high,
         "dnds","genes_corrected","site_merge","aachange") %>%
  distinct()

dnds_cor_test <- cor.test(coefficients_dnds$mean,log10(coefficients_dnds$dnds))
dnds_site_plot <- coefficients_dnds %>%
  ggplot(aes(x = dnds,
             y = mean,
             ymax = HDPI_low,
             ymin = HDPI_high)) +
  geom_smooth(method = 'lm',linetype = 2,color = 'black',
              alpha = 0.05,
              formula = y ~ x - 1) + 
  geom_point(aes(color = genes_corrected)) + 
  geom_errorbar(aes(color = genes_corrected),
                alpha=0.50) + 
  theme_gerstung() + 
  # annotate(geom = 'text',x = 1000,y = min(coefficients_dnds$`0.025`,na.rm = T),
  #          label = sprintf("Rsquared = %s\np-value = %s",
  #                          round(dnds_cor_test$estimate^2,2),
  #                          round(dnds_cor_test$p.value,2))) +
  scale_color_manual(values = gene_colours,name = NULL) + 
  theme(legend.position = "bottom") +
  xlab('log(Site-specific dN/dS)') + 
  ylab("Genetic clone growth coefficient") + 
  scale_y_continuous() + 
  scale_x_continuous(trans='log10') + 
  #ggtitle("dN/dS associations with genetic growth coefficients") + 
  ggsave(useDingbats=FALSE,filename = sprintf("figures/%s/dnds/dnds.pdf",model_id),
         height = 5,width = 5)

dnds_gene <- dnds_data$genes %>%
  subset(sig == T) %>% 
  select(gene_name,
         dNdS_NT=wmis_cv,
         dNdS_NT_0025=mis_95CIlow,
         dNdS_NT_0975=mis_95CIhigh,
         dNdS_T=wnon_cv,
         dNdS_T_0025=trunc_95CIlow,
         dNdS_T_0975=trunc_95CIhigh)
dnds_gene <- rbind(
  data.frame(Gene=dnds_gene$gene_name,
             Value=dnds_gene$dNdS_NT,
             Value_0025 = dnds_gene$dNdS_NT_0025,
             Value_0975 = dnds_gene$dNdS_NT_0975,
             Truncating=F),
  data.frame(Gene=dnds_gene$gene_name,
             Value=dnds_gene$dNdS_T,
             Value_0025 = dnds_gene$dNdS_T_0025,
             Value_0975 = dnds_gene$dNdS_T_0975,
             Truncating=T)
) %>%
  mutate(Gene=as.character(Gene))

values_b_gene <- r_values %>%
  select(
    mean = b_gene,
    HDPI_low = b_gene_005,
    HDPI_high = b_gene_095,
    b_gene_var,
    gene = genes
  ) %>%
  distinct %>%
  mutate(Truncating = grepl(pattern = '[A-Z0-9]t',x = gene)) %>% 
  mutate(Gene = str_match(gene,'[A-Z0-9]+'))
  
coefficients_dnds_gene <- values_b_gene %>%
  merge(dnds_gene,by = c("Gene","Truncating"))
dnds_cor_test_gene <- cor.test(coefficients_dnds_gene$mean,log10(coefficients_dnds_gene$Value))
dnds_gene_plot <- coefficients_dnds_gene %>%
  ggplot(aes(y=mean,ymin=HDPI_low,ymax=HDPI_high,
             x=Value,xmin=Value_0025,xmax=Value_0975,
             color=Gene,shape = Truncating)) + 
  geom_smooth(method = 'glm',inherit.aes=F,
              mapping = aes(y=mean,x=Value),
              color='black',linetype=2,
              alpha=0.05,
              formula = y ~ x - 1) + 
  geom_point() + 
  geom_errorbar(alpha=0.5) + 
  geom_errorbarh(alpha=0.5) + 
  theme_gerstung() + 
  # annotate(geom = 'text',x = 1,y = min(coefficients_dnds_gene$`0.025`),
  #          label = sprintf("Rsquared = %s\np-value = %s",
  #                          round(dnds_cor_test_gene$estimate^2,2),
  #                          round(dnds_cor_test_gene$p.value,2))) +
  scale_shape(labels = c("Non-truncating","Truncating"),name = NULL) +
  scale_color_manual(values = gene_colours,name = NULL) + 
  theme(legend.position = "bottom") +
  xlab('log(Gene dN/dS)') + 
  ylab("Gene-only clone growth coefficient") + 
  #ggtitle("dN/dS association with gene growth coefficients") + 
  scale_x_continuous(trans='log10') + 
  scale_y_continuous(limits = c(-0.2,0.35)) +
  ggsave(useDingbats=FALSE,filename = sprintf("figures/%s/dnds/dnds_gene.pdf",model_id),height = 5,width = 5) 

legend_dnds <- get_legend(dnds_gene_plot)
plot_grid(dnds_gene_plot + theme(legend.position = "none"),
          dnds_site_plot + theme(legend.position = "none")) %>%
  plot_grid(legend_dnds,rel_heights = c(1,0.30),ncol=1)
```

## Associations between VAF and blood counts and blood biochemistry

Here we calculate if/how changes in VAF are affected by changes in blood count and blood chemistry data.

```{r load data and merge}
stat_sign_individuals <- statistics_data %>%
  group_by(individual) %>%
  summarise(sig = sum(sum_stat < 0.7) > 0) %>%
  subset(sig == F)

blood_count_data <- load_blood_count_data()
fuller_data <- merge(load_data(),blood_count_data,by = c("SardID","Phase")) %>%
  subset(!(SardID %in% c(load_excluded_individuals(),load_excluded_individuals_lymph()))) %>%
  subset(!(SardID %in% stat_sign_individuals$individual))
```

```{r data preprocessing}
complete_blood_counts <- fuller_data %>%
  group_by(SardID) %>%
  mutate(relative_timepoint = Phase - min(Phase) + 1) %>% 
  select(SardID,Phase,Age,relative_timepoint,Gender,
         RBC,HB,MCV,MCH,WBC,NEUT_PERC,LYMPH_PERC,MONO_PERC,EOS_PERC,BASO_PERC,
         PLT,Glucose,SerumInsulin,BLoodUreaNitrogen,Creatinine,ALT = ALT.y,AST,GGT,Fibrinogen,
         TotalCholesterol,HDLcholesterol,Triglycerides,Iron,Transferrin,UricAcid,ESR,Phase,Age) %>%
  group_by(SardID) %>%
  mutate(complete = sum(is.na(RBC)) == 0) %>%
  ungroup() %>%
  subset(complete == T) %>%
  mutate(
    NEUT = WBC * NEUT_PERC/100,
    MONO = WBC * MONO_PERC/100,
    LYMPH = WBC * LYMPH_PERC/100,
    BASO = WBC * BASO_PERC/100,
    EOS = WBC * EOS_PERC/100
  ) %>%
  mutate(GRAN = NEUT + BASO + EOS) %>%
  distinct %>% 
  subset(complete == T) %>% 
  gather(key = "par",value = "val",-SardID,-Phase,-Age,-relative_timepoint,-complete,-Gender) 

slopes_and_means <- complete_blood_counts %>% 
  subset(!is.na(val)) %>% 
  group_by(SardID,par,Gender) %>%
  summarise(
    mean = mean(val),
    slope = ifelse(length(Phase) > 2,
                   sum((Age - mean(Age)) * (val - mean(val))) / sum((Age-mean(Age))^2),
                   NA),
    meanAge = mean(Age)
  )
```

```{r }
blood_and_clonal_data <- slopes_and_means %>%
  merge(Parameters_Age,by.x = "SardID",by.y = "individual") %>%
  group_by(SardID) %>%
  filter(b_clone_mean + b_gene_mean + b_site_mean == max(b_clone_mean + b_gene_mean + b_site_mean))

associations <- unique(blood_and_clonal_data$par) %>%
  lapply(
    function(x) {
      tmp_data <- subset(blood_and_clonal_data,par == x)
      lm(b_clone_mean ~ mean + slope + meanAge + Gender,tmp_data) %>%
        return
    }
  )

associations %>%
  lapply(summary) 
```

## Sex, smoking and other factors

```{r differences_sex, fig.width=15,fig.height=5}
fisher_exact_factor <- function(x,totals) {
  pres <- table(x)
  levels_pres <- names(pres)
  levels_totals <- names(totals)
  DF <- data.frame(
    level = c(levels_pres,levels_totals),
    counts = c(pres,totals),
    P = c(rep(T,length(levels_pres)),rep(F,length(totals)))
  ) 
  DF <- DF %>%
    spread(key = 'level',value = 'counts')
  pres <- gtools::na.replace(DF[2,2:ncol(DF)],0)
  not_pres <- gtools::na.replace(DF[1,2:ncol(DF)],0) - gtools::na.replace(DF[2,2:ncol(DF)],0)
  out <- fisher.test(t(rbind(pres,not_pres)))
  return(out)
}

totals <- table(distinct(select(full_data,SardID,Gender))$Gender)

p_val_comparisons <- full_data %>%
  select(individual = SardID,Gender,Gene) %>% 
  distinct %>%
  #mutate(Gender = ifelse(Gender == 'F','Female','Male')) %>%
  mutate(Gene = str_match(Gene,'[A-Z0-9]+')) %>%
  group_by(Gene,Gender) %>%
  mutate(size = length(Gene)) %>%
  group_by(Gene) %>%
  summarise(p.val = fisher_exact_factor(Gender,totals)$p.value,
            max_total = max(size),
            min_total = min(size)) %>%
  mutate(p.val.label = ifelse(p.val < 0.05 / length(unique(full_data$Gene)),'*',NA))

effect_distribution_gender <- full_data %>%
  select(individual = SardID,Gender) %>% 
  distinct %>%
  merge(r_values %>% select(individual,site,coefficient,b_clone,genes),by = c('individual')) %>%
  group_by(individual) %>%
  mutate(effect = coefficient) %>% 
  distinct() %>% 
  filter(effect == max(effect)) %>%
  mutate(Gender = ifelse(Gender == 'F','Female','Male')) %>%
  mutate(Truncating = !grepl("nt",genes)) %>%
  mutate(Truncating = ifelse(is.na(Truncating),F,T)) %>%
  mutate(genes = str_match(genes,'[A-Z0-9]+')) %>%
  ggplot(aes(x = as.factor(Gender),y = effect)) + 
  geom_boxplot(alpha = 0.8) + 
  theme_gerstung(base_size = 15) + 
  geom_signif(comparisons = list(c("Male","Female")),
              map_signif_level = function(x) return(paste("p-value =",format(x,scientifier = T,digits = 4))),
              test = wilcox.test) + 
  ylab("Driver gene effect") + 
  xlab("Sex") 

gene_mutation_distribution_gender <- full_data %>%
  select(individual = SardID,Gender,Gene) %>% 
  distinct %>%
  mutate(Gender = ifelse(Gender == 'F','Female','Male')) %>%
  mutate(Gene = str_match(Gene,'[A-Z0-9]+')) %>%
  ggplot(aes(x = Gene,fill = Gender)) + 
  geom_bar(stat = "count",position = 'dodge') + 
  geom_text(data = p_val_comparisons,
            inherit.aes = F,
            aes(x = Gene,label = p.val.label,y = max_total+1),
            size = 5) +
  theme_gerstung(base_size = 15) +
  scale_fill_lancet(name = NULL) + 
  rotate_x_text() + 
  theme(legend.position = "bottom") + 
  ylab("Count") + 
  xlab("Gene")

age_distribution_gender <- full_data %>%
  select(individual = SardID,Gender) %>% 
  distinct %>%
  merge(r_values %>% select(age,individual,site,coefficient,b_clone,genes),by = c('individual')) %>%
  mutate(full_effect = coefficient + b_clone) %>% 
  group_by(individual) %>%
  filter(coefficient == max(coefficient)) %>%
  mutate(Gender = ifelse(Gender == 'F','Female','Male')) %>%
  mutate(Truncating = !grepl("nt",genes)) %>%
  mutate(Truncating = ifelse(is.na(Truncating),F,T)) %>%
  mutate(genes = str_match(genes,'[A-Z0-9]+')) %>%
  ggplot(aes(x = Gender,y = age,fill = Gender)) + 
  geom_boxplot() + 
  theme_gerstung(base_size = 15) +
  scale_fill_lancet(name = NULL) + 
  rotate_x_text() + 
  theme(legend.position = "bottom") + 
  ylab("Mutation count") + 
  xlab("Gene")

effect_distribution_gender %>% 
  ggsave(useDingbats=FALSE,filename = sprintf("figures/%s/phenotype_associations/FigureGender.pdf",model_id),
         width = 3.5,height = 4)
effect_distribution_gender %>% 
  ggsave(filename = sprintf("figures/%s/phenotype_associations/FigureGender.svg",model_id),
         width = 3.5,height = 4)

D <- full_data %>% 
  group_by(SardID,Gene,Gender) %>% 
  summarise(Age = min(Age)) %>% 
  distinct %>% 
  group_by(SardID,Gender,Age) %>% 
  summarise(HasU2AF1 = "U2AF1" %in% Gene)

glm(HasU2AF1 ~ Gender + Age, data = D,
    family = stats::binomial(link = "logit")) %>% summary

plot_grid(gene_mutation_distribution_gender,effect_distribution_gender,nrow = 1,rel_widths = c(1,0.3)) + 
  ggsave(useDingbats=FALSE,filename = sprintf("figures/%s/phenotype_associations/FigureGenderWithBarplot.pdf",model_id),
         width = 15,height = 4.5)
```

```{r untangling_mf_differences,fig.width=14,fig.height=5}
p_values_gender <- merge(
  r_values %>% 
    select(individual,site,coefficient,genes,b_clone) %>% 
    distinct %>% 
    merge(distinct(select(full_data,individual=SardID,Gender)),by = "individual") %>% 
    mutate(genes = str_match(genes,'[A-Z0-9]+')) %>%
    filter(!is.na(coefficient)) %>%
    group_by(genes) %>%
    summarise(
      pval.coefficient = ifelse(
        all(c("M","F") %in% Gender),
        wilcox.test(coefficient[Gender == 'M'],coefficient[Gender == 'F'],exact = F)$p.val,
        NA)),
  r_values %>% 
    select(individual,site,coefficient,genes,b_clone) %>% 
    distinct %>% 
    merge(distinct(select(full_data,individual=SardID,Gender)),by = "individual") %>% 
    mutate(genes = str_match(genes,'[A-Z0-9]+')) %>%
    group_by(genes) %>%
    summarise(
      pval.b_clone = ifelse(
        all(c("M","F") %in% Gender),
        wilcox.test(b_clone[Gender == 'M'],b_clone[Gender == 'F'],exact = F)$p.val,
        NA)),
  by = "genes"
)

clone_specific_gender_plot <- r_values %>% 
  select(individual,site,coefficient,genes,b_clone) %>% 
  distinct %>% 
  merge(distinct(select(full_data,individual=SardID,Gender)),by = "individual") %>% 
  mutate(genes = str_match(genes,'[A-Z0-9]+')) %>% 
  ggplot(aes(x = genes,y = b_clone,fill = Gender)) + 
  geom_boxplot(lwd=0.2,outlier.size = 0.5) + 
  theme_gerstung() + 
  rotate_x_text() + 
  scale_fill_lancet(name = NULL,labels = c("Female","Male")) +
  theme(legend.position = "bottom") + 
  xlab("Genes") + 
  ylab("Mutation-independent\ngrowth effect")

vaf_gender_plot <- r_values %>% 
  select(individual,site,coefficient,genes,b_clone,true,coverage,age) %>% 
  group_by(individual,site) %>%
  filter(age == max(age)) %>% 
  merge(distinct(select(full_data,individual=SardID,Gender)),by = "individual") %>% 
  mutate(genes = str_match(genes,'[A-Z0-9]+')) %>% 
  ggplot(aes(x = genes,y = true/coverage,fill = Gender)) + 
  geom_boxplot(lwd=0.2,outlier.size = 0.5) + 
  theme_gerstung() + 
  rotate_x_text() + 
  scale_fill_lancet(name = NULL,labels = c("Female","Male")) +
  theme(legend.position = "bottom") + 
  xlab("Genes") + 
  ylab("VAF") + scale_y_continuous(
    trans = 'log10',
    minor_breaks = c(sapply(10^c(-4:-1),function(x) seq(x*2,x*10-x,length.out = 8))))

clone_ages_gender_plot <- clone_ages %>% 
  select(site,individual,genes,CloneAgeHenryLower_adj,CloneAgeHenryUpper_adj) %>%
  distinct() %>%
  subset(abs(CloneAgeHenryLower_adj) < 120 & CloneAgeHenryUpper_adj > -10) %>%
  merge(distinct(select(full_data,individual=SardID,Gender)),by = "individual") %>%
  mutate(genes = str_match(genes,'[A-Z0-9]+')) %>% 
  ggplot(aes(x = genes,y = (CloneAgeHenryLower_adj+CloneAgeHenryUpper_adj)/2,fill = Gender)) + 
  geom_boxplot(lwd=0.2,outlier.size = 0.5) + 
  theme_gerstung() + 
  rotate_x_text() + 
  scale_fill_lancet(name = NULL,labels = c("Female","Male")) +
  theme(legend.position = "bottom") + 
  xlab("Genes") + 
  ylab("Clone age")

full_data %>%
  subset(grepl('DNMT3A',Gene)) %>%
  ggplot(aes(x = Age,y = VAF,group = paste(SardID,mutation_identifier),color = Gender)) + 
  geom_line() + 
  theme_gerstung() +
  scale_color_lancet() +
  scale_y_continuous(trans = 'log10') + 
  facet_wrap(~ Gender,nrow = 1)

plot_grid(clone_specific_gender_plot,vaf_gender_plot,clone_ages_gender_plot,nrow = 1)
```

```{r smoking_gene_composition}
load_smoking_data() %>% 
  mutate(status = ifelse(QuitSmoking == 'N' & CurrentSmoker == 'N',"Never smoked",
                         ifelse(QuitSmoking == 'N' & CurrentSmoker == 'Y',"Smoker",
                                "Previous smoker"))) %>% 
  subset(!is.na(CurrentSmoker)) %>% 
  group_by(SardID) %>%
  summarise(
    status = ifelse(
      sum(status == 'Smoker') > 0,'Smoked during study',
      ifelse(sum(status == 'Previous smoker') > 0, 'Smoked before study',
             'Never smoked')
    )
  ) %>% 
  mutate(status_binary = ifelse(status == 'Never smoked','Never smoked','Smoked')) %>% 
  merge(clone_ages,by.x = "SardID",by.y = "individual") %>% 
  select(genes,SardID,status,status_binary) %>%
  distinct() %>%
  mutate(genes = str_match(genes,'[A-Z0-9]+')) %>%
  distinct() %>%
  ggplot(aes(x = genes,fill = status_binary)) +
  geom_bar(position = 'dodge') + 
  #geom_text(aes(label = total)) +
  theme_gerstung() + 
  theme(legend.position = 'bottom') + 
  rotate_x_text() + 
  xlab("Gene") + 
  ylab("Enrichment") +
  facet_wrap(~ status_binary,scales = "free_y") +
  scale_fill_lancet(name = NULL)
```

```{r differences_smoking_coefficient, fig.height=7,fig.width=6}
load_smoking_data() %>% 
  mutate(status = ifelse(QuitSmoking == 'N' & CurrentSmoker == 'N',"Never smoked",
                         ifelse(QuitSmoking == 'N' & CurrentSmoker == 'Y',"Smoker",
                                "Previous smoker"))) %>% 
  subset(!is.na(CurrentSmoker)) %>% 
  group_by(SardID) %>%
  summarise(
    status = ifelse(
      sum(status == 'Smoker') > 0,'During',
      ifelse(sum(status == 'Previous smoker') > 0, 'Before',
             'Never')
    )
  ) %>% 
  mutate(status = factor(status,levels = c("Never","Before","During"))) %>%
  merge(r_values,by.x = "SardID",by.y = "individual") %>%
  mutate(individual = SardID) %>%
  merge(statistics_data,by = c("individual","site")) %>%
  subset(sum_stat < 0.7) %>% 
  mutate(genes = str_match(genes,'[A-Z0-9]+')) %>%
  group_by(SardID) %>%
  mutate(effect = coefficient) %>%
  filter(effect == max(effect)) %>% 
  select(SardID,effect,status) %>%
  distinct() %>% 
  ggplot(aes(x = status,y = effect)) + 
  #geom_point(alpha = 0.8) +
  geom_boxplot(alpha = 0.8) +
  theme_gerstung(base_size = 15) +
  geom_signif(comparisons = list(c("Never","Before")),
              y_position = 0.45,
              #map_signif_level = function(x) return(format(x,scientifier = T,digits = 4)),
              test = wilcox.test) + 
  geom_signif(comparisons = list(c("Never","During")),
              #map_signif_level = function(x) return(format(x,scientifier = T,digits = 4)),
              test = wilcox.test) +
  geom_signif(comparisons = list(c("Before","During")),
              y_position = 0.45,
              #map_signif_level = function(x) return(format(x,scientifier = T,digits = 4)),
              test = wilcox.test) + 
  theme(legend.position = 'bottom') + 
  scale_color_manual(values = gene_colours,name = NULL) + 
  xlab("Smoked") + 
  ylab("Driver gene effect") + 
  ggsave(useDingbats=FALSE,filename = sprintf("figures/%s/phenotype_associations/FigureSmoking.pdf",model_id),
         width = 2.8,height = 4) +
  ggsave(filename = sprintf("figures/%s/phenotype_associations/FigureSmoking.svg",model_id),
         width = 2.8,height = 4)

load_smoking_data() %>% 
  mutate(status = ifelse(QuitSmoking == 'N' & CurrentSmoker == 'N',"Never smoked",
                         ifelse(QuitSmoking == 'N' & CurrentSmoker == 'Y',"Smoker",
                                "Previous smoker"))) %>% 
  subset(!is.na(CurrentSmoker)) %>% 
  group_by(SardID) %>%
  summarise(
    status = ifelse(
      sum(status == 'Smoker') > 0,'During',
      ifelse(sum(status == 'Previous smoker') > 0, 'Before',
             'Never')
    )
  ) %>% 
  mutate(status = factor(status,levels = c("Never","Before","During"))) %>%
  merge(r_values,by.x = "SardID",by.y = "individual") %>%
  mutate(individual = SardID) %>%
  merge(statistics_data,by = c("individual","site")) %>%
  subset(sum_stat < 0.7) %>% 
  mutate(genes = str_match(genes,'[A-Z0-9]+')) %>%
  group_by(SardID) %>%
  mutate(effect = coefficient) %>%
  filter(effect == max(effect)) %>% 
  select(SardID,effect,status,age) %>%
  mutate(status = ifelse(status == 'Never','No','Yes')) %>%
  distinct() %>% 
  ggplot(aes(x = status,y = effect)) + 
  #geom_point(alpha = 0.8) +
  geom_boxplot(alpha = 0.8) +
  theme_gerstung(base_size = 15) +
  geom_signif(comparisons = list(c("Yes","No")),
              map_signif_level = function(x) return(paste("p-value =",format(x,scientifier = T,digits = 4))),
              test = wilcox.test) + 
  theme(legend.position = 'bottom') + 
  scale_color_manual(values = gene_colours,name = NULL) + 
  xlab("Smoked") + 
  ylab("Driver gene effect") + 
  ggsave(filename = sprintf("figures/%s/phenotype_associations/FigureSmoking_2levels.svg",model_id),
         width = 3.5,height = 4) +
  ggsave(useDingbats=FALSE,filename = sprintf("figures/%s/phenotype_associations/FigureSmoking_2levels.pdf",model_id),
         width = 3.5,height = 4)
```

```{r smoking_clone_age, fig.width=8, fig.height=3}
INTERVAL <- 5
smoking_clone_ages <- load_smoking_data() %>% 
  mutate(status = ifelse(QuitSmoking == 'N' & CurrentSmoker == 'N',"Never smoked",
                         ifelse(QuitSmoking == 'N' & CurrentSmoker == 'Y',"Smoker",
                                "Previous smoker"))) %>% 
  subset(!is.na(CurrentSmoker)) %>% 
  group_by(SardID) %>%
  filter(phase == max(phase)) %>%
  ungroup() %>%
  merge(Parameters_Age,by.x = "SardID",by.y = "individual",all.x = T) %>%
  merge(distinct(select(subset(load_data_keep(),Gene %in% load_included_genes() | is.na(Gene)),Gender,SardID,age=Age)),
        by = "SardID",all.y = T) %>%
  mutate(individual = SardID) %>%
  mutate(genes = str_match(gene,'[A-Z0-9]+')) %>%
  group_by(SardID) %>%
  mutate(effect = Mean) %>%
  filter(age == max(age)) %>%
  select(SardID,effect,site,CurrentSmoker, QuitYears,status,age,genes,Gender) %>% 
  #subset(effect > 18 & effect < 100) %>% 
  mutate(effect = age - effect) %>% 
  mutate(QuitYears = ifelse(CurrentSmoker == 'Y',0,QuitYears)) %>% 
  mutate(CurrentlySmoking = effect > QuitYears) %>%
  mutate(CurrentlySmoking = ifelse(status == "Never smoked",FALSE,CurrentlySmoking)) %>%
  mutate(AgeGroups = floor((age - effect) / INTERVAL) * INTERVAL)

smoking_clone_ages %>% 
  ungroup() %>%
  subset(AgeGroups <= 60 | is.na(AgeGroups)) %>% 
  subset(AgeGroups >= 15) %>% 
  mutate(AgeGroups = paste(AgeGroups,AgeGroups + INTERVAL,sep='-')) %>% 
  mutate(AgeGroups = ifelse(AgeGroups == '15-20','18-20',AgeGroups)) %>%
  group_by(SardID,AgeGroups) %>%
  distinct() %>%
  summarise(NClones = length(site),
            CurrentlySmoking = CurrentlySmoking[1]) %>%
  ggplot(aes(x = AgeGroups,
             y = NClones,
             fill = CurrentlySmoking,
             colour = CurrentlySmoking,
             group = paste(AgeGroups,CurrentlySmoking))) +
  stat_summary(geom = 'line',
               inherit.aes = F,
               alpha = 0.7,
               mapping = aes(x = AgeGroups,
                             y = NClones,
                             colour = CurrentlySmoking,
                             group = CurrentlySmoking),
               fun.y = mean,
               size = 1) +
  # geom_point(position = position_jitterdodge(jitter.width = 0.4,
  #                                            dodge.width = 0.8,
  #                                            jitter.height = 0.02),
  #            alpha = 0.6) +
  
  theme_gerstung(base_size = 15) +
  #theme(legend.position = 'bottom') +
  scale_colour_d3(labels = c("No","Yes"),name = "Smoking") +
  scale_fill_d3(labels = c("No","Yes"),name = "Smoking") + 
  #rotate_x_text() +
  xlab("Age groups") + 
  ylab("No. of clones\nappearing") + 
  #scale_y_continuous(trans = 'log10') +
  theme(axis.text.x = element_text(size = 11.2)) + 
  ggsave(useDingbats=FALSE,sprintf("figures/%s/phenotype_associations/FigureCloneSmoking.pdf",model_id),height = 4,width = 7) + 
  ggsave(sprintf("figures/%s/phenotype_associations/FigureCloneSmoking.svg",model_id),height = 4,width = 7)
```

```{r smoking_clone_age_boxplot}
load_smoking_data() %>% 
  mutate(status = ifelse(QuitSmoking == 'N' & CurrentSmoker == 'N',"Never smoked",
                         ifelse(QuitSmoking == 'N' & CurrentSmoker == 'Y',"Smoker",
                                "Previous smoker"))) %>% 
  subset(!is.na(CurrentSmoker)) %>% 
  group_by(SardID) %>%
  filter(phase == max(phase)) %>%
  ungroup() %>%
  merge(clone_ages,by.x = "SardID",by.y = "individual") %>%
  mutate(individual = SardID) %>%
  merge(statistics_data,by = c("individual","site")) %>%
  mutate(genes = str_match(genes,'[A-Z0-9]+')) %>% 
  group_by(SardID) %>%
  mutate(effect = (CloneAgeHenryLower_adj + CloneAgeHenryUpper_adj)/2) %>%
  filter(age == max(age)) %>%
  select(SardID,effect,site,CurrentSmoker, QuitYears,status,age) %>% 
  subset(effect > 18 & effect < 100) %>% 
  mutate(effect = age - effect) %>% 
  mutate(QuitYears = ifelse(CurrentSmoker == 'Y',0,QuitYears)) %>% 
  mutate(CurrentlySmoking = effect > QuitYears) %>%
  mutate(CurrentlySmoking = ifelse(status == "Never smoked",FALSE,CurrentlySmoking)) %>%
  ggplot(aes(x = 0,y = effect,colour = CurrentlySmoking)) + 
  geom_boxplot() + 
  theme_gerstung()
```

## Is the unknown cause effect in JAK2 provoked by specific genotypes?

We have access to gene dosage data for different mutations which relate with JAK2 oncogenesis. We can, as such, test if they are in any way or form correlated with the unknown cause effect in JAK2, a gene whose unknown cause effects appear to be quite varied.

```{r}
jak2_karyotype <- load_jak2_karyotype() %>%
  na.omit()
jak2_karyotype %>%
  na.omit() %>% 
  ggplot(aes(x = Genotype)) +
  geom_bar() + 
  facet_wrap(~ sprintf('%s (chr. %s)',Site,Chromosome),scales = 'free_x') + 
  theme_gerstung(base_size = 15) + 
  scale_y_continuous(breaks = seq(0,20,by=2)) + 
  ylab("Number of individuals")

jak2_karyotype_and_effect <- r_values %>%
  select(b_clone,SardID=individual,genes,age) %>%
  group_by(SardID) %>%
  filter(age == min(age)) %>% 
  distinct() %>% 
  subset(genes == 'JAK2') %>%
  merge(jak2_karyotype,by = "SardID") %>%
  select(-Dose) %>% 
  mutate(Chromosome = paste('Chr',Chromosome,sep='')) %>% 
  pivot_wider(names_from = c(Chromosome,Site),values_from = Genotype,names_sep = 'Site')

jak2_karyotype_and_effect %>% 
  gather("key","value",-SardID,-b_clone,-genes,-age) %>% 
  ggplot(aes(x = value,y = b_clone)) + 
  geom_point() +
  theme_gerstung(base_size = 15) + 
  facet_wrap(~ key,scales = 'free_x') + 
  xlab("Genotype") + 
  ylab("Unknown cause effect")

iou_genotype <- apply(t(combn(colnames(jak2_karyotype_and_effect)[5:16],2)),1,
      function(x) {
        tmp <- data.frame(A = jak2_karyotype_and_effect[,x[1]],
                          B = jak2_karyotype_and_effect[,x[2]]) %>% 
          table
        I <- sum(diag(tmp))
        U <- sum(tmp)
        return(c(A=x[1],B=x[2],IoU=I/U))
      }) %>% 
  t %>%
  as.data.frame(stringsAsFactors=F) %>%
  mutate(IoU = as.numeric(as.character(IoU)))

all_chr_site <- colnames(jak2_karyotype_and_effect)[c(5:16)]
all_chr_site <- all_chr_site[!(all_chr_site %in% iou_genotype$A[iou_genotype$IoU == 1])]

lm_formula <- formula(
  paste('b_clone',paste(c(all_chr_site),collapse = ' + '),sep=' ~ ')
)

lm(formula = lm_formula,data = jak2_karyotype_and_effect) %>%
  summary
```

From the data we have it is hard to conclude anything regarding the effect of different genotypes in differences in JAK2 dynamics. Looking at dot plots, there appears to be different distributions between the three different genotypes in site 1287194 in chromosome 5, but the very small sample size does not allow us to draw any conclusions from this.

## Is the U2AF1 prevalence in blood cells we observe in our cohort normal or explained by the old age?

To assess this we need to get the prevalence of U2AF1 in blood cells, which we obtain by marginalising over AML, CML, MDS and CH

```{r checking_U2AF1_prevalence}
u2af1_CH_list <- list(
  "Jaiswal 2014" = c(n_mutations = 0,total_individuals = 17182),
  "Genovese 2014" = c(n_mutations = 0,total_individuals = 12380),
  "McKerrell 2015" = c(n_mutations = NA,total_individuals = 4219),
  "Zink 2017" = c(n_mutations = 0, total_individuals = 11262),
  "Acuna-Hidalgo" = c(n_mutations = 1, total_individuals = 9350),
  "Coombs 2017" = c(n_mutations = 1,total_individuals = 6898), # a few other U2AF1 mutations, but only in individuals which developed a malignant haematological condition
  "Young 2016" = c(n_mutations = 0,total_individuals = 20),
  "Young 2019" = c(n_mutations = 1,75), # 3 more U2AF1 mutations in AML
  "Desai 2018" = c(n_mutations = 0,total_individuals = 181) # 6 U2AF1 mutations in MAL
)
p_u2af1_c_ch <- sum(sapply(u2af1_CH_list,function(x) x[1]),na.rm = T) / sum(sapply(u2af1_CH_list,function(x) x[2]))
p_u2af1_c_mds <- 8.73/100
p_u2af1_c_cml <- 6.67/100
p_u2af1_c_aml <- 4.63/100
p_ch <- c(0.9/100,20/100)
p_mds <- 2 / 1e5
p_cml <- 1.6 / 1e5
p_aml <- 4.3 / 1e5

# marginalise P(U2AF1)
p_u2af1 <- p_u2af1_c_ch * p_ch + p_u2af1_c_mds * p_mds + p_u2af1_c_aml * p_aml + p_u2af1_c_cml * p_cml

load_data() %>% 
  select(Gene,SardID) %>% 
  distinct %>% 
  summarise(`U2AF1 incidence in our cohort` = length(unique(SardID[Gene == 'U2AF1'])) / length(unique(SardID)),
            `Expected U2AF1 incidence for individuals over 70` = p_u2af1[2]) 
```

The incidence of U2AF1 mutations in our cohort is quite higher than what would be expected in a sample of individuals over 70. As such, we can, to some extent, validate our findings regarding the late appearance of most U2AF1 clones.

## Survival analysis 

```{r survival_analysis,fig.height=6.5,fig.width=12,dpi=300}
library(survival)
library(survminer)
# quick_plot_survival <- function(s,) {
#   s <- survfit(s)
#   summary_s <- summary(s)
#   data_plot <- data.frame(
#     y = summary_s$surv,
#     ymin = summary_s$lower,
#     ymax = summary_s$upper,
#     x = summary_s$time
#   )
# }

comorbidity_data <- load_comorbidity_data()
smoking_data <- load_smoking_data() %>% 
  mutate(status = ifelse(QuitSmoking == 'N' & CurrentSmoker == 'N',"Never smoked",
                         ifelse(QuitSmoking == 'N' & CurrentSmoker == 'Y',"Smoker",
                                "Previous smoker"))) %>% 
  subset(!is.na(CurrentSmoker)) %>% 
  group_by(SardID) %>%
  summarise(
    status = ifelse(
      sum(status == 'Smoker') > 0,'During',
      ifelse(sum(status == 'Previous smoker') > 0, 'Before',
             'Never')
    )
  ) %>% 
  mutate(status = factor(status,levels = c("Never","Before","During"))) %>%
  mutate(EverSmoked = ifelse(status == "Never","No","Yes"))

blood_parameter_ltp <- load_blood_count_data() %>%
  group_by(SardID) %>% 
  filter(Phase == max(Phase))

grouping_data_size <- load_data() %>% 
  group_by(SardID) %>% 
  filter(Phase == max(Phase)) %>% 
  group_by(SardID) %>% 
  summarise(
    LargestCloneVAF = max(VAF),
    NClones = length(unique(AAChange.refGene[!is.na(AAChange.refGene)])),
    LastTimePoint = min(Age),
    Gender = unique(Gender)) %>%
  mutate(HasLargeClone=LargestCloneVAF>0.025) %>%
  distinct()
grouping_data_dynamics <- r_values %>% 
  select(SardID=individual,site,b_clone,coefficient) %>%
  group_by(SardID) %>% 
  summarise(FastestClone=max(b_clone+coefficient),
            FastestGene=max(coefficient)) %>% 
  distinct() 
 
survival_data <- load_survival_data() %>%
  as_tibble() %>%
  subset(DEAD != "U") %>%
  subset(gtools::na.replace(as.character(AGE_of_death) != 'Unknown',T)) %>%
  mutate(Age_General = as.numeric(gtools::na.replace(AGE_of_death,0))+as.numeric(gtools::na.replace(AGE_atStudyEnd,0))) %>%
  mutate(DEAD = ifelse(DEAD=='Y',1,0))
full_survival_data <- merge(grouping_data_size,survival_data,by = "SardID",all=T) %>%
  merge(grouping_data_dynamics,by = "SardID",all=T) %>%
  merge(smoking_data,by = "SardID") %>%
  merge(comorbidity_data,by = "SardID") %>% 
  mutate(Diabetes2Status = ifelse(DIABETES_2,"Yes","No"),
         HadCancer = ifelse(CANCER,"Yes","No")) %>% 
  mutate(DEAD=as.numeric(DEAD),
         Age_General=as.numeric(Age_General),
         FastestClone=ifelse(is.na(FastestClone),0,FastestClone),
         FastestGene=ifelse(is.na(FastestGene),0,FastestGene),
         HasLargeClone=ifelse(is.na(HasLargeClone),F,HasLargeClone),
         NClones=ifelse(is.na(NClones),0,NClones),
         LargestCloneVAF=ifelse(is.na(LargestCloneVAF),0,LargestCloneVAF)) %>% 
  mutate(NClones_Factor = ifelse(
    NClones >= 3,
    "3+",
    ifelse(NClones > 0,"1-2",NClones))) %>% 
  mutate(NoClones=NClones_Factor=='0',
         OneTwoClones=NClones_Factor=='1-2',
         ThreeMoreClones=NClones_Factor=='3+') %>% 
  mutate(FastClone = ifelse(FastestClone >= 0.1,">=0.1","<0.1")) %>%
  filter(!is.na(Age_General)) %>%
  mutate(Age_General = Age_General - LastTimePoint) %>% 
  merge(blood_parameter_ltp,by = 'SardID') %>% 
  subset(!(SardID %in% c(load_excluded_individuals(),load_excluded_individuals_lymph())))

model.base <- coxph(
  Surv(Age_General,DEAD) ~ LastTimePoint + Gender + FastestClone + NClones + LargestCloneVAF + Diabetes2Status + EverSmoked,
  x=T,y=T,model=T,
  data=full_survival_data) 

forest_plot <- ggforest(model.base,
                        full_survival_data,
                        main = "",refLabel = "")

plot_size <- ggsurvplot(
  coxph(
    Surv(Age_General,DEAD) ~ LastTimePoint + Gender + FastestClone + NClones + LargestCloneVAF + strata(HasLargeClone),
    x=T,y=T,model=T,
    data=full_survival_data) %>% survfit,
  data = full_survival_data
)$plot + 
  scale_color_lancet(name="Clone size",labels=c("<2.5%",">=2.5%"))

plot_number <- ggsurvplot(
  coxph(
    Surv(Age_General,DEAD) ~ LastTimePoint + Gender + FastestClone + NClones + LargestCloneVAF + strata(NClones_Factor),
    x=T,y=T,model=T,
    data=full_survival_data) %>% survfit,
  data = full_survival_data
)$plot + 
  scale_color_uchicago(name="Number of clones")

plot_dynamics <- ggsurvplot(
  coxph(
    Surv(Age_General,DEAD) ~ LastTimePoint + Gender + FastestClone + NClones + LargestCloneVAF + strata(FastClone),
    x=T,y=T,model=T,
    data=full_survival_data) %>% survfit,
  data = full_survival_data
)$plot + 
  scale_color_d3(name = "Mutation-independent growth coefficient")

plot_grid(
  plot_grid(plot_size,plot_number,plot_dynamics,nrow = 1),
  plot_grid(ggplot()+theme_void(),forest_plot+theme_gerstung(),ggplot()+theme_void(),
            rel_widths = c(0.3,0.8,0.3),nrow = 1),
  ncol = 1,
  rel_heights = c(1,1.2)
) + 
  ggsave(useDingbats=FALSE,sprintf("figures/%s/survival_analysis/general_figure.pdf",model_id),
         height = 5,width = 12)

```

This survival analysis shows fairly clearly that there does not seem to be any strong association with dynamic parameters. This may be due to the small sample size, but we are still not able to replicate results by others regarding the number of clones and clone size. We suspect this may be due to the fact that our cohort is fairly older than those previously used (the median age during the first time point was of 70.7 years). 

# Session details

```{r session_details}
cat(system("git log -n1", intern=TRUE), sep="\n")
```

```{r session_objects}
l <- ls()
data.frame(variable=l, Reduce("rbind",lapply(l, function(x) data.frame(classes=paste(class(get(x)),collapse = ','), size=format(object.size(get(x)), units="auto")))))
```

```{r session_packages}
sessionInfo()
```