---
title: "The Natural History and Fitness Landscape of Clonal Haematopoiesis"
subtitle: "Growth coefficient inference and ages at clone foundation"
author: 
    - Jos√© Guilherme de Almeida
    - Moritz Gerstung
output: 
  html_document:
    theme: lumen
    highlight: tango
    code_folding: hide
    df_print: paged
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 4
    number_sections: true
--- 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,dpi = 300)
```

```{r import_functions, include=FALSE}
source("scripts/vaf_dynamics_functions.R")
set.seed(42)
```

# Problem setting 

We have a cohort with roughly $400$ individuals from Sardinia. These individuals were target-sequenced at a high depth (1000X), which enables us to detect clonal haematopoiesis with a high precision, and at different time points, which in theory enables the modelling of how VAF progress. Ultimately, this enables us to infer the future trajectory of VAF values and determine the effect size underlying this. To model VAF, hierarchical Bayes modelling is used and, in general terms, each VAF is modelled as a binomial distribution whose size is the coverage and probability best described as the linear combination of one or more effects and an offset per individual.

```{r map sardinia,fig.height=8,fig.width=8,fig.align="center",results='hide',fig.keep='all',warning=FALSE,message=FALSE}
m_s <- map_sardinia()
m_s
```

# Data preparation

Data preparation here is concerned mostly with the following aspects:

* Selecting genes or domains based on their significance for our analysis 
* Excluding individuals who have (had) haematological malignancies
* Determining mutations which occur across more than 1 individual
* Defining subsets within our data 

```{r data_preparation}
source("scripts/prepare_data.R")
```

# Data exploration

An initial inspection of the trajectories does not allow us to make any ambitious claims on their dynamics considering age as the only determinant of dynamics. 

Some conclusions can be drawn at this point:

* No apparent single dynamics underlies most VAF trajectories when considering the age of individuals, but this becomes neater when using relative timepoints (relative time instead of absolute time) - this points towards the necessity of an individual-based offset. This also makes the likeliness of other factors (i.e. other mutations) affecting the dynamics of VAF higher
* There is some apparently very similar dynamics for different VAFs within the same individual, which may be a good indicator of the existence of a slope offset on a *per* individual basis for all clones. This may also helps us to decipher any possible interaction there may exist between clones

```{r plot_vaf_trajectories_per_individual, fig.width=14, fig.height=10}
full_data %>% subset(single_occurring == F) %>% 
  mutate(q005=qbeta(p=0.05,MUTcount_Xadj+1,TOTALcount+1-MUTcount_Xadj),
         q095=qbeta(p=0.95,MUTcount_Xadj+1,TOTALcount+1-MUTcount_Xadj)) %>% 
  select(TOTALcount,MUTcount_Xadj,SardID,Age,VAF,q005,q095,Phase,amino_acid_change) %>% 
  
  ggplot(aes(x = Age,y = MUTcount_Xadj/(TOTALcount),ymin = q005,ymax = q095,color = as.factor(SardID))) + 
  geom_linerange(alpha = 0.7) + 
  geom_line() + 
  geom_point() + 
  facet_wrap(~ amino_acid_change,scales = 'free') + 
  theme_minimal() + 
  scale_color_discrete(guide=FALSE) +
  ylab("VAF") + 
  ggtitle("Trajectory view for each site with more than two mutations (one color per individual)") + 
  ggsave("figures/data_exploration/vaf_trajectories_per_individual.pdf",width=14, height=10)
```

```{r plot_vaf_trajectories_per_site, fig.width=28, fig.height=30}
full_data %>% #subset(single_occurring == F) %>% 
  mutate(q005=qbeta(p=0.05,MUTcount_Xadj+1,TOTALcount+1-MUTcount_Xadj),
         q095=qbeta(p=0.95,MUTcount_Xadj+1,TOTALcount+1-MUTcount_Xadj)) %>% 
  select(TOTALcount,MUTcount_Xadj,SardID,Age,VAF,q005,q095,Phase,amino_acid_change) %>% 
  ggplot(aes(x = Age,y = MUTcount_Xadj/(TOTALcount),ymin = q005,ymax = q095,color = amino_acid_change)) + 
  geom_linerange(alpha = 0.7) + 
  geom_line() + 
  geom_point() + 
  facet_wrap(~ SardID,scales = 'free') + 
  theme_minimal(base_size=15) + 
  scale_color_discrete(guide=FALSE) +
  ylab("VAF") + 
  ggtitle("Trajectory view for each individual (one color per mutation)") + 
  scale_x_continuous(n.breaks = 4,
                     breaks = function(x) floor(x[1]) + seq(1,x[2],by = round((x[2] - x[1])/5))) + 
  ggsave("figures/data_exploration/vaf_trajectories_per_site.pdf",width=28, height=30)
```

# Modelling 

```{r defining_site_domain_gene_lists}
site_list <- formatted_data_train_1$unique_site_multiple
domain_list <- formatted_data_train_1$unique_domain
gene_list <- formatted_data_train_1$unique_gene
recurrent_sites <- formatted_data_train_1$unique_site_multiple
```

```{r creating_result_lists}
r_values_list <- list()
statistics_lists <- list()
```

We model the counts of mutation $i$ in individual $j$ - clone $c_{ji}$ - at age $t$ as the product of a beta-binomial distribution such that $counts_{c_{ji}} \sim BB(cov_{c_{ji}}(t),\alpha_{c_{ji}}(t),\beta)$, where $cov_{c_{ji}}$ is the coverage for mutation $c_{ji}$ at timepoint $t$, $\beta$ is the technical overdispersion and $\alpha = \frac{\beta p}{1 - p(t)}$, where $p(t)$ is the expected value for the counts at age $t$. 

Here we work with a relatively simple formulation of clone growth - for a given clone $i$ in individual $j$ we consider there to be a *genetic* growth effect ($b_{genetic_{i}} = b_{gene_i} + b_{domain_i} + b_{site_i}$), a *clone-specific* growth effect ($b_{c_{ji}}$) and a *clone-specific* offset ($u_{ji}$). Only the *growth effects* mentioned here depend on age. These two growth effects and offset are linearly combined and inverse-logit transformed to yield a probability such that $p_{ji}(t) = ilogit(b_{genetic_{i}} = (b_{genetic_{i}} + b_{c_{ji}}) * t + u_{ji})$. Further model specifications are specified below:

$$
b_{gene_i} \sim N(0,1)
\\
b_{domain_i} \sim N(0,0.1)
\\
b_{site_i} \sim N(0,0.1)
\\
u_{ji} \sim Uniform(-100,0)
\\
b_{clone_{ji}} \sim N(0,0.05)
\\
p_i = ilogit((b_{gene_i} + b_{domain_i} + b_{site_i} + b_{clone_{ji}}) * age + u_{ji})
\\
\beta \sim N(\mu=\mu_{\beta},\sigma=\sigma_{\beta})
\\
\alpha_{SR} = \frac{\beta p}{1 - p}
\\
counts \sim BB(coverage,alpha,beta)
$$

## Modelling the technical overdispersion

Technical replicate data shows some level of technical overdispersion. This can be modelled in our current model by changing our binomial modelling of the counts to a beta binomial, with a fixed $\beta$ parameter and an $\alpha$ parameter that depends on $\beta$ and on the probability that is inferred by modelling, particularly by using the formulation for the expected value of the beta distribution $E[X] = \frac{\alpha}{\alpha+\beta}$. 

Considering the above, we can model the counts as $counts \sim BB(n =  coverage,alpha = \frac{\beta p}{1 - p},beta = \beta)$

```{r visualising overdispersion,fig.height=14,fig.width=14}
one_hot_encode <- function(factor_vector) {
  factor_levels <- sort(as.character(unique(factor_vector)))
  print(factor_levels)
  output <- sapply(
    factor_vector,
    function(x) {
      as.numeric(factor_levels == x)
    }
  )
  return(t(output))
}

tru_q_data <- read.table('data/TruQ.txt',header = T)
replicate_data <- read.table('data/Replicates_all.txt',header = T) %>%
  mutate(VAF = ifelse(CHR == "X", VAF / 2,VAF),
         MUTcount = ifelse(CHR == "X", MUTcount / 2,MUTcount))
ages <- merge(full_data,replicate_data,by = c("SardID","Phase")) %>%
  select(Age, SardID, Phase) %>%
  unique
replicated_data_multiple <- replicate_data %>%
  group_by(SardID,variant) %>%
  summarise(count = length(unique(Phase))) %>%
  subset(count > 1)
replicate_data <- merge(
  replicate_data,
  ages,
  by = c("SardID","Phase"),
  all = T
) %>% 
  subset(SardID %in% replicated_data_multiple$SardID & variant %in% replicated_data_multiple$variant)

tru_q_data %>% 
  ggplot(aes(x = VAF_expected,y = VAF_observed)) + 
  geom_point(aes(colour = as.factor(Replicate))) + 
  geom_errorbar(
    aes(
      ymin = qbeta(0.05,MUTcount + 1,TOTALcount - MUTcount + 1),
      ymax = qbeta(0.95,MUTcount + 1,TOTALcount - MUTcount + 1)
  )) + 
  theme_minimal(base_size = 15) +
  facet_wrap(~ Gene) + 
  ggsci::scale_color_lancet(name = "Replicate") +
  theme(legend.position = 'bottom') + 
  xlab("VAF expected") +
  ylab("VAF observed") + 
  ggtitle("TruQ data") + 
  ggsave("figures/overdispersion/overdispersion_vis_truq.pdf",height=14,width=14)

replicate_data %>% 
  ggplot(aes(x = as.factor(Phase),y = VAF)) + 
  geom_point(aes(colour = as.factor(Replicate)),
             size = 2) + 
  geom_linerange(
    aes(
      ymin = qbeta(0.05,MUTcount + 1,TOTALcount - MUTcount + 1),
      ymax = qbeta(0.95,MUTcount + 1,TOTALcount - MUTcount + 1)
    )) + 
  theme_minimal(base_size = 15) +
  facet_wrap(~ paste(Gene,variant,SardID,sep = '-'),scale = 'free') + 
  ggsci::scale_color_lancet(name = "Replicate") +
  theme(legend.position = 'bottom') + 
  xlab("Phase") +
  ylab("VAF") + 
  ggtitle("Replicate data from Sardinian samples") + 
  ggsave("figures/overdispersion/overdispersion_vis_sard.pdf",height=14,width=14)
```

The invervals here are estimated using a beta distribution ($Beta(alpha = count+1,beta = coverage - count + 1)$).

### Estimating $\beta$ with technical replicates

Using technical replicates for known DNA quantities ($truQ$) and for actual patient data ($SR$), we can posit the following model:

$$
p_{truQ} = observed\ VAF_{truQ} + 1/coverage_{truQ}
\\
\beta ~ \sim exponential(0.001)
\\
\alpha_{truQ} = \frac{\beta p_{truQ}}{1 - p_{truQ}}
\\
expected\ VAF_{truQ} + 1/coverage_{truQ} \sim Beta(alpha,beta)
$$

We need to adjust the expected and observed VAFs with the equivalent of having one count due to the fact that beta distributions only permit $\alpha>0$ and $\beta>0$.

$$
p_{SR} = growth\ coefficient = effect_{gene} + individual_{offset}
\\
\alpha_{SR} = \frac{\beta p_{SR}}{1 - p_{SR}}
\\
observed\ counts_{SR} \sim BB(coverage + 1,alpha,beta)
$$

Here, $p_{SR}$ is parametrized as $p_{SR} = ilogit((b_{gene_{i}} + b_{c_{ji}}) * t + u_{ji})$. 

For this subsubsection, the model is specified in `scripts/estimating_overdispersion.R`.

We used this model to infer the $\mu_{\beta}$ and $\sigma_{\beta}$ mentioned in the model specification above.

### Modelling clone growth as a product of genetic and clone-specific effects

The model specified in `scripts/F2_clone_effect.R` was executed using `scripts/run_model_f2.R`. The model parameters are stored in `models/model_F2_full.RDS` and will be used for the remainder of this section.

```{r model_f2_load_values}
values_model_f2 <- readRDS("models/model_F2_full.RDS")
```

``` {r model_f2_dynamic_visualization_preparation, fig.width = 30,fig.height = 30}
val_subset <- full_formatted_data

X_val <- val_subset$counts[values_model_f2$gene_idxs,]
gene_idxs <- values_model_f2$gene_idxs

interference_idxs <- values_model_f2$interference_idxs
u_idx <- values_model_f2$u_idx

b_gene_values_val <- values_model_f2$b_gene_values[[1]]$values[values_model_f2$b_gene_values[[1]]$labels == '0.50'] %>% 
  matrix(ncol=1) 
b_domain_values_val <- values_model_f2$b_domain_values[[1]]$values[values_model_f2$b_domain_values[[1]]$labels == '0.50'] %>% 
  matrix(ncol=1) 
b_site_values_val <- values_model_f2$b_site_values[[1]]$values[values_model_f2$b_site_values[[1]]$labels == '0.50'] %>% 
  matrix(ncol=1) 
beta_val <- values_model_f2$beta_values[[1]]$values[values_model_f2$beta_values[[1]]$labels == '0.50'] %>% 
  matrix(ncol=1) 
b_clone_values_val <- values_model_f2$b_clone[[1]]$values[values_model_f2$b_clone[[1]]$labels == '0.50'] %>% 
  matrix(ncol=1)

u_values_val <- values_model_f2$u_values[[1]]$values[values_model_f2$u_values[[1]]$labels == '0.50'] %>%
  matrix(nrow=1) 

ae_gene <- val_subset$gene_to_site_indicator[values_model_f2$gene_idxs,] %*% b_gene_values_val
ae_domain <- val_subset$domain_to_site_indicator[values_model_f2$gene_idxs,] %*% b_domain_values_val
ae_site <- val_subset$site_multiple_to_site_indicator[values_model_f2$gene_idxs,] %*% b_site_values_val

age_effect_coef <- ae_gene + ae_domain + ae_site
age_effect <- age_effect_coef[interference_idxs$site] + b_clone_values_val[interference_idxs$ind_site]
age_effect <- age_effect * (val_subset$ages[interference_idxs$ind_age] - values_model_f2$min_age)

offset_per_individual <- u_values_val[interference_idxs$ind_site]
r <- age_effect + offset_per_individual
mu_val <- inv.logit(r)

r_values <- data.frame(
  mu_val = mu_val,
  true = val_subset$counts[gene_idxs,][cbind(interference_idxs$site,interference_idxs$ind_age)],
  age = val_subset$ages[interference_idxs$ind_age],
  coverage = val_subset$coverage[gene_idxs,][cbind(interference_idxs$site,interference_idxs$ind_age)],
  individual = val_subset$unique_individual_true[interference_idxs$ind],
  coefficient_095 = values_model_f2$b_values[[1]]$values[values_model_f2$b_values[[1]]$labels == '0.95'][interference_idxs$site],
  coefficient = values_model_f2$b_values[[1]]$values[values_model_f2$b_values[[1]]$labels == '0.50'][interference_idxs$site],
  coefficient_005 = values_model_f2$b_values[[1]]$values[values_model_f2$b_values[[1]]$labels == '0.05'][interference_idxs$site],
  site = val_subset$unique_site[values_model_f2$gene_idxs][interference_idxs$site],
  b_clone_005 = c(values_model_f2$b_clone[[1]]$values[values_model_f2$b_clone[[1]]$labels == '0.05'][interference_idxs$ind_site]),
  b_clone = c(values_model_f2$b_clone[[1]]$values[values_model_f2$b_clone[[1]]$labels == '0.50'][interference_idxs$ind_site]),
  b_clone_095 = c(values_model_f2$b_clone[[1]]$values[values_model_f2$b_clone[[1]]$labels == '0.95'][interference_idxs$ind_site]),
  u_values_005 = c(values_model_f2$u_values[[1]]$values[values_model_f2$u_values[[1]]$labels == '0.05'][interference_idxs$ind_site]),
  u_values = c(values_model_f2$u_values[[1]]$values[values_model_f2$u_values[[1]]$labels == '0.50'][interference_idxs$ind_site]),
  u_values_095 = c(values_model_f2$u_values[[1]]$values[values_model_f2$u_values[[1]]$labels == '0.95'][interference_idxs$ind_site])
) %>%
  as.data.frame() %>%
  apply(1,FUN = function(x){
    mu_val <- as.numeric(x[1])
    cov <- as.numeric(x[4])
    alpha_value <- (beta_val * mu_val) / (1 - mu_val)
    Q <- extraDistr::rbbinom(n = 1000,
                             size = cov,
                             alpha = alpha_value,
                             beta = beta_val) %>%
      quantile(c(0.05,0.50,0.95))
    names(Q) <- c("pred_005","pred","pred_095")
    return(c(x,Q))
  }) %>% 
  t %>%
  as.data.frame() %>% 
  mutate(genes = sapply(as.character(site),function(x) unlist(strsplit(x,'-'))[[1]])) %>%
  group_by(site) %>%
  mutate(n = length(pred)) %>%
  ungroup() %>%
  mutate(
    mu_val = as.numeric(as.character(mu_val)),
    true = as.numeric(as.character(true)),
    age = as.numeric(as.character(age)),
    coverage = as.numeric(as.character(coverage)),
    individual = as.numeric(as.character(individual)),
    coefficient_005 = as.numeric(as.character(coefficient_005)),
    coefficient = as.numeric(as.character(coefficient)),
    coefficient_095 = as.numeric(as.character(coefficient_095)),
    pred_005 = as.numeric(as.character(pred_005)),
    pred = as.numeric(as.character(pred)),
    pred_095 = as.numeric(as.character(pred_095)),
    b_clone_005 = as.numeric(as.character(b_clone_005)),
    b_clone = as.numeric(as.character(b_clone)),
    b_clone_095 = as.numeric(as.character(b_clone_095)),
    u_values_005 = as.numeric(as.character(u_values_005)),
    u_values = as.numeric(as.character(u_values)),
    u_values_095 = as.numeric(as.character(u_values_095))
  )

r_values_ <- r_values %>% 
  mutate(VAFpred = pred / coverage,
         VAFtrue = true / coverage) %>% 
  subset(coverage > 0) %>%
  group_by(individual,site) %>% 
  mutate(normalizedVAFpred = (VAFpred + 1) / (VAFpred[which(age == min(age))] + 1),
         normalizedVAFtrue = (VAFtrue + 1) / (VAFtrue[which(age == min(age))] + 1)) %>%
  ungroup() %>%
  mutate(individual = as.character(individual)) %>% 
  gather(key = 'key',value = 'value',VAFpred,VAFtrue) %>%
  mutate(
    pred_005 = ifelse(key == 'VAFpred',
                      pred_005 / coverage,
                      NA),
    pred_095 = ifelse(key == 'VAFpred',
                      pred_095 / coverage,
                      NA)
  ) 

r_values_list$model_f2 <- r_values

values_model_f2$b_site_values[[1]] %>% mutate(variable = rep(values_model_f2$training_subset$unique_site_multiple,each=9)) %>% 
  write.csv("site_coefficients.csv",quote = F,row.names = F)
values_model_f2$b_gene_values[[1]] %>% mutate(variable = rep(values_model_f2$training_subset$unique_gene,each=9)) %>% 
  write.csv("gene_coefficients.csv",quote = F,row.names = F)
```

#### Assessing the goodness of fit with the residual effect

We here define the residual effect (RE) as $RE = 1 - s$, with $s=CDF_{\chi^2}(\sum_{i=1}^{n}Q_{N(0,1)}(CDF_{BBinom(coverage,\alpha_{inferred},\beta{inferred})}(counts))^2)$. We draw a cutoff at 0.7, considering trajectories with $RE < 0.7$ as good trajectories and those with $RE \get 0.7$ as bad trajectories. Bad trajectories are excluded from any downstream analysis.

```{r model_f2_summary_statistic, fig.height=5,fig.width=6}
order_same_rank <- function(v) {
  u <- unique(v)
  o <- rep(0,length(v))
  for (i in 1:length(u)) {
    o[v == u[i]] <- i
  }
  return(o)
}

beta_value <- values_model_f2$beta_values[[1]][1,1]

statistic_full <- r_values_ %>%
  subset(key == 'VAFpred') %>%
  mutate(tail_prob = extraDistr::pbbinom(
    q = true,
    size = coverage,
    alpha = (mu_val * beta_value)/(1 - mu_val),
    beta = beta_value)) %>%
  group_by(site,individual) %>% 
  summarise(sum_stat = pchisq(sum(qnorm(tail_prob)^2),length(tail_prob)),
            gene = genes[1]) %>%
  mutate(split = 'full') %>%
  select(split,site,sum_stat,gene,individual) %>%
  ungroup() %>%
  mutate(recurring = site %in% values_model_f2$training_subset$unique_site_multiple)

total_variants_explained <- statistic_full %>%
  group_by(site) %>%
  summarise(TotalExplainedVariants = sum(sum_stat < 0.7),
            TotalVariants = length(sum_stat))

statistic_aggregated <- r_values_ %>%
  subset(key == 'VAFpred') %>%
  mutate(
    tail_prob = extraDistr::pbbinom(
    q = true,
    size = coverage,
    alpha = (mu_val * beta_value)/(1 - mu_val),
    beta = beta_value)) %>%
  group_by(site) %>% 
  summarise(sum_stat = pchisq(sum(qnorm(tail_prob)^2),length(tail_prob)),
            gene = genes[1],
            MinimumAge = min(age),
            MaximumAge = max(age),
            MaxValue = max(c(pred_095,value,true/coverage),na.rm = T)) %>%
  mutate(split = 'full') %>%
  ungroup() %>%
  merge(total_variants_explained,by = c("site"))

p <- statistic_full %>%
  ggplot(aes(x = gene,y = sum_stat)) +
  geom_point(aes(color = recurring),
             stat = 'identity',
             position = position_jitter(width = 0.1),
             alpha = 0.3,
             size = 0.4) + 
  rotate_x_text() +
  theme_minimal() +
  theme(legend.position = 'bottom') + 
  xlab("") +
  ylab("Summary statistic") + 
  scale_y_continuous() + 
  geom_hline(yintercept = 0.7,alpha = 0.5) +
  scale_color_lancet() + 
  rotate_x_text()

p

statistic_full %>% 
  select(sum_stat) %>%
  mutate(sum_stat = sum_stat < .7) %>%
  table

statistics_lists$model_f2 <- list(
  individual_site = statistic_full,
  site = statistic_aggregated
)
```

#### Visualising example trajectories

```{r model_f2_dynamic_visualization_all,fig.width = 30,fig.height = 30}
r_values_merged <- merge(statistic_full,
                         r_values_,
                         all_y = F,
                         all.x = F,
                         by.x = c("site","individual"),
                         by.y = c("site","individual")) %>%
  mutate(site_plot = sapply(site,function(x) str_split(x,';')[[1]][1])) %>%
  mutate(site_plot = gsub('-','\n',site_plot))

average_statistic <- r_values_merged %>% 
  group_by(site) %>%
  summarise(AverageStatistic = mean(sum_stat),
            TotalExplainedVariants = sum(sum_stat < 0.7),
            MinimumAge = min(age),
            MaximumAge = max(age),
            MaxValue = max(c(pred_095,value),na.rm = T)) 

r_values_merged %>% 
  ggplot(aes(x = age,y = value)) +
  geom_ribbon(alpha = 0.1,
              data = r_values_merged %>%
                subset(key == 'VAFpred'),
              aes(ymin = pred_005,
                  ymax = pred_095,
                  group = as.factor(paste0(as.character(individual),key))),
              na.rm = T) +
  geom_hline(aes(yintercept = (1/2e5) * (10/coefficient))) +
  geom_line(aes(colour = key,group = as.factor(paste0(as.character(individual),key)))) +
  facet_wrap(~ site_plot,scales = "free") + 
  scale_color_lancet(name = NULL) +
  theme_minimal(base_size = 10) + 
  theme(legend.position = 'bottom',
        axis.text.x = element_text(size=7)) +
  xlab("Age") + 
  ylab("VAF") + 
  scale_x_continuous(n.breaks = 4,
                     breaks = function(x) floor(x[1]) + seq(1,x[2],by = round((x[2] - x[1])/4))) +
  ggtitle("All inferred (blue) and observed (red) trajectories") +
  ggsave("figures/inferred_trajectories/plot_F2.pdf",width = 30,height = 30)
```

```{r model_f2_dynamic_visualization,warning=F,fig.width = 8,fig.height = 6}
ss <- statistic_aggregated %>%
  subset(sum_stat <= 0.05) %>%
  select(site) %>%
  unlist
site_subset <- sort(val_subset$unique_site_multiple)[c(1,10,21,26,30,31,32,43,65,68)][c(1,3,5,6,7,8,9,10)]

r_values_merged %>% 
  subset(site %in% site_subset) %>%
  mutate(key = ifelse(key == "VAFpred",
                      "Inferred",
                      "Observed")) %>%
  ggplot(aes(x = age,y = value)) +
  geom_ribbon(data = r_values_merged %>%
                subset(key == 'VAFpred') %>%
                subset(site %in% site_subset),
              alpha = 0.1,
              aes(ymin = pred_005,
                  ymax = pred_095,
                  group = as.factor(paste0(as.character(individual),key))),
              na.rm = T) +
  geom_line(aes(colour = key,
                group = as.factor(paste0(as.character(individual),key)),
                alpha = 1-sum_stat),
            size = 0.7) +
  geom_label(
    label.r = unit(0,"cm"),
    label.size = unit(0,"cm"),
    label.padding = unit(0,"cm"),
    alpha = 0.7,
    data = subset(statistic_aggregated,site %in% site_subset),
    size = 3,
    aes(#group = site,
        x = MinimumAge + (MaximumAge - MinimumAge) * 0.35,
        y = MaxValue - MaxValue * 0.1,
        label = paste0('Explained\nvariants = ',format(100 * TotalExplainedVariants/TotalVariants,digits = 2),'%')),
    fontface = "italic") +
  facet_wrap(~ site,scales = "free",ncol = 4) + 
  scale_color_lancet(name = NULL) +
  theme_minimal(base_size = 15) + 
  theme(legend.position = 'bottom',
        strip.text = element_text(size = 9),
        legend.key.height = unit(0,"cm")) +
  scale_y_continuous(limits = c(0, NA)) + 
  xlab("Age") + 
  ylab("VAF") +
  scale_alpha(guide = F,range = c(0.4,1.0)) +
  ggsave("figures/inferred_trajectories/Figure4_smaller_F2.pdf",width = 8,height = 6) + 
  ggsave("figures/inferred_trajectories/Figure4_smaller_F2.svg",width = 8,height = 6)
```

Observing the above trajectories gives a better insight into the effectiveness of our model in capturing trajectories with very distinct clonal dynamics.

#### Inspecting coefficients

In this subsection we inspect coefficients for all genomic resolutions (gene, domain and site) individually and then observe the general picture of their combined effects. The most relevant aspects are the wide range of dynamics observable for all genes and the lack of evidence for a strong domain signal. Additionally, sites in the same gene do not appear to have in most cases a more or less pronounced signal - a clear exception of this is P95H in SRSF2, which grows much faster than mutations on the site site for different amino acids (P95A and P95L). The frameshift in T483fs in PPM1D appears to produce clones which grow faster than other mutations on PPM1d but its large credible interval prevents us from claiming anything with a higher degree o certainty. Mutations in U2AF1 and IDH1 are unusually fast growers when compared with other genes. Another striking aspect of this analysis is how the prevalence of mutations is quite a poor predictor of their clonal dynamics - JAK2 and SF3B1 mutations, while being quite prevalent, do not produce clones which grow particularly fast.

```{r model_f2_gene_coefficients, fig.width=14, fig.height=10}
values_b_gene <- values_model_f2$b_gene_values[[1]]
values_b_gene$site_name <- rep(val_subset$unique_gene,
                               each = length(unique(values_model_f2$b_values[[1]]$labels)))
values_b_gene %>% spread(key = labels,value = values) %>%
  mutate(Truncating = grepl("[A-Z0-9]+t",site_name),
         site_name = str_match(site_name,"[A-Z0-9]+")) %>%
  ggplot(aes(y = `0.50`,x = site_name,color = site_name,shape = Truncating)) +
  geom_point(size = 2,
             position = position_dodge(width = 1.0)) +
  geom_linerange(aes(ymin = HDPI_low,ymax = HDPI_high),
                 position = position_dodge(width = 1.0)) + 
  geom_hline(yintercept = 0,alpha = 0.2) +
  theme_minimal(base_size = 15) +
  rotate_x_text() + 
  xlab("Gene") + 
  ylab("Effect size") + 
  facet_grid(~ site_name,scales = "free_x",space = "free_x") +
  scale_color_manual(values = gene_colours,
                     guide = F) +
  theme(legend.position = "bottom",
        legend.text = element_text(size = 10),
        legend.title = element_text(size = 10),
        axis.title = element_text(size = 15),
        panel.grid.major.x = element_blank(),
        panel.grid.minor.x = element_blank(),
        strip.text = element_blank(),
        legend.key.height = unit(0,"cm")
        ) + 
  scale_shape_discrete(breaks = c(TRUE,FALSE),
                       labels = c("Truncating","Non-truncating"),
                       name = NULL) +
  ggsave("figures/dynamic_coefficients/FigureGene_F2.pdf",height = 3.5,width = 8)
```

```{r model_f2_domain_coefficients, fig.width=14, fig.height=10}
values_b_domain <- values_model_f2$b_domain_values[[1]]
values_b_domain$site_name <- rep(val_subset$unique_domain,
                                 each = length(unique(values_model_f2$b_values[[1]]$labels)))
values_b_domain$gene <-  sapply(values_b_domain$site_name,function(x) (strsplit(x,'-') %>% unlist)[1])
values_b_domain %>% spread(key = labels,value = values) %>%
  ggplot(aes(y = `0.50`,x = substr(site_name,1,30),color = gene)) +
  geom_point() +
  geom_linerange(aes(ymin = `0.05`,ymax = `0.95`)) + 
  geom_hline(yintercept = 0,alpha = 0.2) +
  theme_minimal() +
  rotate_x_text() + 
  xlab("") + 
  ylab("Effect size") + 
  facet_grid(~ gene,scales = "free_x",space = "free_x") +
  scale_color_manual(values = gene_colours,
                     guide = F) +
    theme(legend.position = "bottom",
        panel.grid.major.x = element_blank(),
        panel.grid.minor.x = element_blank(),
        strip.text = element_text(size = 5.7,face = "bold")) + 
  ggsave("figures/dynamic_coefficients/FigureDomain_F2.pdf",height = 3.5,width = 8)
```

```{r model_f2_site_coefficients, fig.width=14, fig.height=10}
values_b_site <- values_model_f2$b_site_values[[1]]
values_b_site$site_name <- rep(val_subset$unique_site_multiple,
                               each = length(unique(values_model_f2$b_values[[1]]$labels)))
values_b_site$gene <-  sapply(values_b_site$site_name,function(x) (strsplit(x,'-') %>% unlist)[1])

values_b_site %>% spread(key = labels,value = values) %>%
  ggplot(aes(y = `0.50`,x = substr(site_name,1,30),color = gene)) +
  geom_point() +
  geom_linerange(aes(ymin = HDPI_low,ymax = HDPI_high)) + 
  geom_hline(yintercept = 0,alpha = 0.2) +
  theme_minimal() +
  rotate_x_text() + 
  xlab("") + 
  ylab("Effect size") + 
  facet_grid(~ gene,scales = "free_x",space = "free_x") +
  scale_color_manual(values = gene_colours,
                     guide = F) +
    theme(legend.position = "bottom",
        panel.grid.major.x = element_blank(),
        panel.grid.minor.x = element_blank(),
        strip.text = element_text(size = 5.7,face = "bold")) +
  ggsave("figures/dynamic_coefficients/FigureSite_F2.pdf",height = 3.5,width = 8)
```

```{r model_f2_full_effects, fig.width=14, fig.height=6.5,warning=F}
values_b <- values_model_f2$b_values[[1]]
values_b$site_name <- rep(val_subset$unique_site[values_model_f2$gene_idxs],
                          each = length(unique(values_model_f2$b_values[[1]]$labels)))
values_b$gene <- values_b$site_name %>% 
  sapply(function(x) unlist(str_split(x,pattern = '-'))[1])

gene_order <- values_b %>%
  mutate(gene = str_match(gene,"[A-Z0-9]+")) %>%
  subset(labels == "0.50") %>% 
  subset(site_name %in% val_subset$unique_site_multiple) %>%
  group_by(gene) %>%
  summarise(average = mean(values)) %>%
  arrange(average) %>%
  select(gene) %>%
  unlist

Number_Of_Mutations <- r_values %>%
  select(individual,site) %>%
  distinct() %>%
  group_by(site) %>%
  summarise(n = length(site)) %>%
  ungroup() %>%
  mutate(site_name = site) %>%
  select(-site)

values_b %>% 
  merge(Number_Of_Mutations,by = 'site_name') %>%
  mutate(Truncating = grepl("[A-Z0-9]+t",gene),
         gene = str_match(gene,"[A-Z0-9]+")) %>%
  spread(key = labels,value = values) %>%
  subset(site_name %in% val_subset$unique_site_multiple) %>%
  group_by(gene) %>%
  mutate(gene_average = mean(`0.50`)) %>%
  mutate(gene_average =  ifelse(`0.50` == max(`0.50`),gene_average,NA)) %>% # avoid having more than one geom_hline per gene
  ungroup() %>%
  mutate(site_name = sapply(site_name,function(x) unlist(strsplit(x,'-'))[2])) %>%
  mutate(site_name = sapply(site_name,function(x) unlist(strsplit(x,';'))[1])) %>%
  mutate(site_name = sapply(site_name,function(x) {
    tmp <- unlist(strsplit(x,':'))
    return(tmp[length(tmp)])})) %>%
  mutate(gene = factor(as.character(gene),levels = gene_order,
                       ordered = TRUE)) %>%
  ggplot(aes(y = exp(`0.50`),
             x = reorder(substr(site_name,1,30),`0.50` + Truncating),
             color = gene)) +
  geom_hline(aes(colour = gene,yintercept = exp(gene_average)),
             size = 2,alpha = 0.5) +
  geom_point(aes(shape = Truncating,
                 size = n)) +
  geom_linerange(aes(ymin = exp(HDPI_low),ymax = exp(HDPI_high))) + 
  geom_hline(yintercept = 0,alpha = 0.2) +
  theme_minimal(base_size = 15) +
  rotate_x_text() + 
  xlab("Site") + 
  ylab("Growth per year") + 
  facet_grid(~ gene,scales = "free_x",space = "free_x") +
  scale_color_manual(values = gene_colours,
                     guide = F) +
  theme(legend.position = "bottom",
        panel.grid.major.x = element_blank(),
        panel.grid.minor.x = element_blank(),
        strip.text = element_text(size = 10,face = "bold",angle = 90),
        legend.key.height = unit(0,"cm")) +
  coord_cartesian(ylim = c(exp(min(values_b$values)) * 0.99,exp(max(values_b$values)) * 1.01)) + 
  scale_shape_discrete(breaks = c(TRUE,FALSE),
                       labels = c("Truncating","Non-truncating"),
                       name = NULL) +
  scale_size(range = c(2,6),
             breaks = c(2,5,10,25,50)) +
  ggsave("figures/dynamic_coefficients/FigureFull_F2.pdf",width = 14,height = 6.5) + 
  ggsave("figures/dynamic_coefficients/FigureFull_F2.svg",width = 14,height = 6.5)
```

#### Inferring ages at clone foundation

The ages at clone foundation are determined using the inferred median values for each parameter for each individual. To do this, we solve $t_0 = \frac{log(\frac{1}{n}) - \alpha + log(\frac{g}{\beta}) - 1}{\beta}$, with $n = 200,000$, $g = 13$, $\beta=growth\ coefficient$ and $alpha = individual\ offset$. We exclude from analysis trajectories with $RE \ge 0.7$ and negative growth coefficients (this leads to no real solution). We also remove impossible values for ages at clone foundation from further analysis ($>100$ and $<0$). 

```{r model_f2_age_at_clone_growth, fig.width=14,fig.height=5.5}
clone_ages <- r_values %>%
  merge(statistic_full,by = c("site","individual"),all = F) %>%
  filter(sum_stat < 0.7) %>% 
  filter(coefficient + b_clone > 0) %>%
  mutate(CloneAgeHenryLower_adj = t0_adjusted(u_values,coefficient + b_clone,13,5e4) + values_model_f2$min_age,
         CloneAgeHenryUpper_adj = t0_adjusted(u_values,coefficient + b_clone,13,2e5) + values_model_f2$min_age
         )
  
gene_order <- clone_ages %>%
  mutate(Truncating = grepl("[A-Z0-9]+t",genes),
         Gene = str_match(genes,"[A-Z0-9]+")) %>%
  subset(abs(CloneAgeHenryLower_adj) < 120 & CloneAgeHenryUpper_adj > -10) %>%
  group_by(Gene) %>%
  summarise(average = mean(CloneAgeHenryUpper_adj + CloneAgeHenryLower_adj)/2) %>%
  arrange(average) %>%
  select(Gene) %>%
  unlist() 

data_plot_clone_ages <- clone_ages %>%
  mutate(Truncating = grepl("[A-Z0-9]+t",genes),
         Gene = str_match(genes,"[A-Z0-9]+"),
         amino_acid_change = as.character(site)) %>%
  subset(abs(CloneAgeHenryLower_adj) < 120 & CloneAgeHenryUpper_adj > -10) %>%
  subset(amino_acid_change %in% values_model_f2$training_subset$unique_site_multiple) %>%
  mutate(amino_acid_change = sapply(amino_acid_change,function(x) unlist(strsplit(x,'-'))[2])) %>%
  mutate(amino_acid_change = sapply(amino_acid_change,function(x) unlist(strsplit(x,';'))[1])) %>%
  mutate(amino_acid_change = sapply(amino_acid_change,function(x) {
    tmp <- unlist(strsplit(x,':'))
    return(tmp[length(tmp)])})) %>%  mutate(Gene = factor(as.character(Gene),levels = gene_order)) %>%
  mutate(Truncating_numeric = as.numeric(Truncating)) %>% 
  select(individual,CloneAgeHenryLower_adj,CloneAgeHenryUpper_adj,Gene,amino_acid_change,Truncating,Truncating_numeric) %>%
  distinct() %>%
  mutate(dummy_individual = reorder(1:length(individual), CloneAgeHenryLower_adj)) 

data_plot_clone_ages_summary <- data_plot_clone_ages %>%
  group_by(amino_acid_change,Gene) %>%
  summarise(max_age = max(CloneAgeHenryLower_adj),
            min_age = min(CloneAgeHenryUpper_adj),
            n = length(CloneAgeHenryUpper_adj)) %>%
  ungroup() %>%
  mutate(width = n / max(n) * 1)

data_plot_clone_ages %>%
  ggplot(aes(group = dummy_individual)) +
  geom_hline(yintercept = 0) +
  geom_tile(
    data = data_plot_clone_ages_summary,
    inherit.aes = F,
    alpha = 0.2,
    aes(
      height = max_age - min_age,
      y = (max_age + min_age)/2,
      x = amino_acid_change,
      width = 0.85,
      fill = Gene
    )
  ) + 
  geom_linerange(
    size = 0.3,
    alpha = 0.8,
    aes(x = reorder(amino_acid_change,
                    (CloneAgeHenryLower_adj + CloneAgeHenryUpper_adj) / 2 + Truncating_numeric * 1000),
        ymax = CloneAgeHenryLower_adj,
        ymin = CloneAgeHenryUpper_adj,
        color = Gene),
    position = position_dodge(width=1.0)) +
  geom_point(
    size = 1.2,
    alpha = 0.9,
    position = position_dodge(width=1.0),
    aes(x = reorder(amino_acid_change,
                    (CloneAgeHenryLower_adj + CloneAgeHenryUpper_adj) / 2 + Truncating_numeric * 1000),
        shape = Truncating,
        y = (CloneAgeHenryLower_adj + CloneAgeHenryUpper_adj) / 2,
        color = Gene)) +
  facet_grid(~ Gene,scales = "free",space = "free") + 
  theme_minimal(base_size = 15) + 
  theme(strip.text = element_text(size = 10,face = 'bold',angle = 90),
        legend.position = 'bottom',
        legend.key.height = unit(0,"cm")) + 
  ylab("Age at beginning of clone growth") +
  xlab("Site") +
  rotate_x_text() + 
  scale_color_manual(values = gene_colours,
                     guide = F) +
  scale_fill_manual(values = gene_colours,
                    guide = F) +
  scale_y_continuous(
    expand = c(0,0),
    limits = c(-10,100),
    oob = scales::squish,
    breaks = c(-500,-400,-300,-200,-100,0,10,20,30,40,50,60,70,80,90,100,1000,2000,3000,5000)
    ) + 
  scale_shape_discrete(breaks = c(TRUE,FALSE),
                       labels = c("Truncating","Non-truncating"),
                       name = NULL) +
  ggsave("figures/age_at_clone_foundation/Figure6_F2.pdf",width = 14,height = 5.5) + 
  ggsave("figures/age_at_clone_foundation/Figure6_F2.svg",width = 14,height = 5.5)
```

```{r ages density,fig.height=4,fig.width=5}
data_plot_clone_ages %>%
  #subset(abs(CloneAgeHenryLower_adj) < 120 & CloneAgeHenryUpper_adj > -10) %>%
  mutate(AgeMidPoint = (CloneAgeHenryLower_adj + CloneAgeHenryUpper_adj)/2) %>%
  #subset(AgeMidPoint < 100) %>% 
  group_by(Gene) %>%
  mutate(MoreThanOne = length(Gene) >= 2) %>%
  ungroup() %>%
  subset(MoreThanOne) %>%
  ggplot(aes(y = AgeMidPoint,fill = Gene,
             x = Gene,group = Gene)) +
  geom_boxplot(alpha = 0.9) +
  theme_minimal(base_size = 15) + 
  theme(strip.text = element_text(size = 8,face = 'bold'),
        legend.position = 'bottom',
        legend.key.height = unit(0,"cm")) + 
  rotate_x_text() + 
  scale_fill_manual(values = gene_colours,
                    guide = F) +
  scale_shape_discrete(breaks = c(TRUE,FALSE),
                       labels = c("Truncating","Non-truncating"),
                       name = NULL) +
  scale_y_continuous(breaks = seq(0,100,10)) +
  ylab("Age at clone foundation") + 
  ggsave("figures/age_at_clone_foundation/Figure6_F2_Boxplots.pdf",width = 5,height = 4)
```

A fairly evident aspect of this analysis is how DNMT3A and TET2 clones appear before 60 years of age and appear to be fairly uniform throughout life. However, other genes such as PPM1D and U2AF1 show up much later in life. 

#### Global picture

Using what we have so far we can replicate the general picture derived by [McKerrel et al.](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4542313/).

```{r global_picture,fig.width=9,fig.height=6}
average_age <- data_plot_clone_ages %>%
  mutate(MidPoint = (CloneAgeHenryLower_adj + CloneAgeHenryUpper_adj)/2) %>%
  group_by(amino_acid_change) %>%
  summarise(MedianAge = median(MidPoint),
            Gene = Gene[1]) %>% 
  transmute(site = paste(str_match(Gene,"[A-Z0-9]+"),
                         amino_acid_change,sep = '-'),
            Gene = Gene,
            MedianAge = MedianAge)

most_recurring_coefficients_gene <- r_values_list$model_f2 %>% 
  mutate(Gene = str_match(genes,'[A-Z0-9]+')) %>%
  group_by(Gene) %>% 
  filter(site == names(sort(table(site),decreasing = T)[1]) | grepl('SRSF2-P95[AHL]',site)) %>%
  select(Gene,site,coefficient) %>%
  distinct() %>%
  mutate(site = paste(Gene,str_match(site,'[0-9A-Za-z]+$'),sep = '-'))

global_picture_parameters <- merge(average_age,most_recurring_coefficients_gene,
                                   by = c("Gene","site"),
                                   all = F) %>%
  arrange(Gene)

TOT_READS <- 1e3
global_picture_out <- list()
for (S in unique(global_picture_parameters$site)) {
  subset_df <- global_picture_parameters[global_picture_parameters$site == S,]
  u_one_cell <- logit(1/TOT_READS) - subset_df$MedianAge * subset_df$coefficient
  time_draws <- inv.logit(u_one_cell + seq(floor(subset_df$MedianAge),90) * subset_df$coefficient) / 2
  global_picture_out[[S]] <- data.frame(
    time = seq(floor(subset_df$MedianAge),90),
    draws = time_draws,
    site = S,
    Gene = subset_df$Gene
  )
}

text_global_picture_out <- global_picture_out %>% 
  do.call(what = rbind) %>%
  subset(Gene %in% c("DNMT3A","U2AF1","SRSF2","JAK2","IDH1","SF3B1")) %>%
  subset(site != 'SRSF2-P95L') %>% 
  group_by(site) %>% 
  summarise(x = max(time),
            y = max(draws),
            label = site[1])

global_picture_out %>% 
  do.call(what = rbind) %>%
  subset(Gene %in% c("DNMT3A","U2AF1","SRSF2","JAK2","IDH1","SF3B1")) %>% 
  subset(site != 'SRSF2-P95L') %>% 
  ggplot(aes(x = time,y = draws,colour = Gene,group = site)) + 
  geom_line(size = 2,
            alpha = 0.5) + 
  geom_text_repel(inherit.aes = F,
            size = 4,
            data = text_global_picture_out,
            hjust = 0,
            vjust = 1.0,
            fontface = "bold",min.segment.length = 1,force = 0.1,
            direction = "y",
            aes(x = x,y = y,label = label)) +
  scale_y_continuous(trans = 'log10',breaks = c(#0,5e-6,1e-5,1e-4,
                                                0.001,0.010,0.10,0.20,0.3,0.5),
                     minor_breaks = NULL,
                     expand = c(0,0.1)) + 
  scale_x_continuous(limits = c(30,110),
                     breaks = c(40,60,80),
                     minor_breaks = NULL,
                     expand = c(0,0.05)) +
  scale_color_lancet(guide=F) + 
  theme_minimal(base_size = 20) +
  theme(legend.position = 'bottom',
        panel.grid = element_blank(),axis.line = element_line()) +
  xlab("Age") + 
  ylab("logVAF") +
  ggsave("figures/inferred_trajectories/GlobalPicture.pdf",height = 6,width = 7) +
  ggsave("figures/inferred_trajectories/GlobalPicture.svg",height = 6,width = 7)
```

## Statistical analysis 

```{r comparing_statistics,fig.height=5,fig.width=18}
statistics_data <- statistics_lists$model_f2$individual_site %>% 
  select(site,sum_stat,gene,individual) %>% 
  mutate(model = "") %>%
  mutate(recurrent = ifelse(site %in% recurrent_sites,TRUE,FALSE)) %>%
  group_by(model) %>%
  mutate(gene = str_match(gene,"[A-Z0-9]+"),
         truncating = str_match(gene,"[nt]+")) %>%
  mutate(model_label = sprintf("%s\np(s-value < 0.7) = %s",model,round(sum(sum_stat < 0.7)/length(sum_stat),3))) %>%
  mutate(gene_numeric = as.numeric(factor(gene,levels = rev(unique(gene)))))

gene_numeric_correspondence <- statistics_data %>%
  ungroup() %>%
  select(gene,gene_numeric) %>%
  distinct()

individuals_with_no_interactions <- interference_idxs %>% 
  select(ind,inter_site) %>%
  distinct() %>%
  transmute(individual = val_subset$unique_individual_true[ind],
            has_interaction = !is.na(inter_site))
```

```{r statistics_model_d,fig.height=7,fig.width=7}
statistics_data %>%
  ungroup() %>%
  ggplot(aes(x = gene_numeric,y = sum_stat,colour = recurrent)) + 
  geom_hline(yintercept = 0.7,alpha = 0.5) + 
  geom_density2d(size = 0.5,
                 alpha = 0.2) +
  geom_point(position = position_jitter(width = 0.3),
             size = 1,
             alpha = 0.4) + 
  theme_minimal(base_size = 15) +
  facet_grid(~ model_label) + 
  rotate_x_text() + 
  stat_summary(
    aes(x = gene_numeric),
    geom = 'label',
    color = 'black',
    label.r = unit(0, "lines"),
    label.size = 0,
    size = 4,
    alpha = 0.85,
    label.padding = unit(0.05,'cm'),
    hjust = 1,
    fun.data = function(x) {
      return(data.frame(y = 0.699,label = sprintf("%s",round(sum(x < 0.7)/length(x),2))))
    }) +
  scale_x_continuous(
    breaks = gene_numeric_correspondence$gene_numeric,
    labels = gene_numeric_correspondence$gene,
    sec.axis = dup_axis(name = NULL),
    expand = c(0.02,0.02)) +
  scale_y_continuous(
    breaks = c(0.00,0.25,0.5,0.7,0.75,1.00),
    expand = c(0.01,0.01)
  ) +
  scale_color_lancet(name = "Present in more\nthan 1 individiual") +
  theme(legend.position = "bottom",
        panel.background = element_rect(fill = NULL,colour = 'black'),
        panel.grid.major.x = element_blank(),
        panel.grid.minor.x = element_blank(),
        panel.grid.major.y = element_line(colour = 'grey',size = 0.2)) + 
  coord_flip() + 
  xlab("Gene") + 
  ylab("s-value") + 
  ggsave("figures/statistical_analysis_of_coefficients/StatisticD2.pdf",height = 6.5,width = 6)

statistics_data %>%
  ungroup() %>%
  mutate(model_label = paste("Proportion of explained trajectories =",round(sum(sum_stat < 0.7)/length(sum_stat),3))) %>% 
  mutate(good = sum_stat < 0.7) %>% 
  mutate(recurrent = as.factor(ifelse(recurrent,"Yes","No"))) %>%
  ggplot(aes(x = gene_numeric,
             y = sum_stat)) + 
  geom_boxplot(aes(color = recurrent,group = paste(gene,recurrent)),
               position = position_dodge(0.8),
               alpha=0.8) +
  geom_hline(yintercept = 0.7,alpha = 0.5,linetype=2) + 
  theme_minimal(base_size = 15) +
  facet_grid(~ model_label) + 
  rotate_x_text() + 
  stat_summary(
    aes(x = gene_numeric),
    geom = 'label',
    color = 'black',
    label.r = unit(0, "lines"),
    label.size = 0,
    size = 4,
    alpha = 0.85,
    label.padding = unit(0.05,'cm'),
    hjust = 1,
    fun.data = function(x) {
      return(data.frame(y = 0.699,label = sprintf("%s",round(sum(x < 0.7)/length(x),2))))
    }) +
  scale_x_continuous(
    breaks = gene_numeric_correspondence$gene_numeric,
    labels = gene_numeric_correspondence$gene,
    sec.axis = dup_axis(name = NULL),
    expand = c(0.02,0.02)) +
  scale_y_continuous(
    breaks = c(0.00,0.25,0.5,0.7,0.75,1.00),
    expand = c(0.01,0.01)
  ) +
  scale_color_manual(
    values = c("#00468BFF","#ED0000FF"),
    label = c("Yes","No"),
    name = "Present in more\nthan 1 individiual") +
  theme(legend.position = "bottom",
        panel.background = element_rect(fill = NULL,colour = 'black'),
        panel.grid.major.x = element_blank(),
        panel.grid.minor.x = element_blank(),
        panel.grid.major.y = element_line(colour = 'grey',size = 0.2)) + 
  coord_flip() + 
  xlab("Gene") + 
  ylab("Residual effect") + 
  ggsave("figures/statistical_analysis_of_coefficients/StatisticD2Boxplot.svg",height = 6.5,width = 6) +
  ggsave("figures/statistical_analysis_of_coefficients/StatisticD2Boxplot.pdf",height = 6.5,width = 6) 
```

```{r saving_everything}
r_values <- r_values_list$model_f2

save(r_values,statistics_data,clone_ages,file = "vaf_modelling_coefficients.Rdata")
```

# Session details

```{r session_details}
cat(system("git log -n1", intern=TRUE), sep="\n")
```

```{r session_objects}
l <- ls()
data.frame(variable=l, Reduce("rbind",lapply(l, function(x) data.frame(classes=paste(class(get(x)),collapse = ','), size=format(object.size(get(x)), units="auto")))))
```

```{r session_packages}
sessionInfo()
```