---
title: "The Natural History and Fitness Landscape of Clonal Haematopoiesis"
subtitle: "Post-hoc analysis"
author: 
    - Jos√© Guilherme de Almeida
    - Moritz Gerstung
output: 
  html_document:
    theme: lumen
    highlight: tango
    code_folding: hide
    df_print: paged
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 4
    number_sections: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r import_functions, include=FALSE}
source("scripts/vaf_dynamics_functions.R")
model_file_name <- "models/model_F5_full.RDS"
set.seed(42)
```

# Post-hoc analysis

```{r making_dirs_loading_data}
model_id <- gsub('_$','',str_match(model_file_name,'model_.*_'))
load(file = sprintf("vaf_modelling_coefficients_%s.Rdata",model_id))
values_model <- readRDS(file = model_file_name)
c('prevalence','dnds',
  'phenotype_associations',
  'survival_analysis') %>%
  sapply(
    function(x) dir.create(sprintf("figures/%s/%s",model_id,x),showWarnings = F)
  )
source("scripts/prepare_data.R")
```

## What if - recreating mutation prevalence

An interesting scenario is to use our growth coefficients and defined dynamics to see how this would affect mutation prevalence through different age brackets. 

```{r observing incidence, fig.width=4,fig.height=7}
INTERVAL <- 5
DETECTION_LIMIT <- 0.005
incidence <- r_values %>%
  subset(coverage > 0) %>%
  mutate(AgeGroup = floor(age/INTERVAL) * INTERVAL) %>%
  mutate(AgeGroupLabel = paste(AgeGroup,AgeGroup+INTERVAL,sep = '-')) %>%
  mutate(genes = str_match(genes,'[A-Z0-9]+')) %>%
  group_by(AgeGroup,AgeGroupLabel) %>%
  mutate(TotalIndividuals = length(unique(individual))) %>%
  ungroup() %>%
  group_by(AgeGroup,AgeGroupLabel,genes,individual) %>%
  summarise(MutPresence = any(true / coverage > DETECTION_LIMIT),
            TotalIndividuals = TotalIndividuals[1]) %>%
  group_by(AgeGroup,AgeGroupLabel,genes) %>%
  summarise(MutCount = sum(MutPresence),
            TotalIndividuals = TotalIndividuals[1]) %>%
  mutate(y = (MutCount) / (TotalIndividuals),
         ymax = qbeta(0.99,MutCount,TotalIndividuals - MutCount + 1),
         ymin = qbeta(0.01,MutCount,TotalIndividuals - MutCount + 1))

true_incidence <- incidence %>%
    subset(AgeGroup <= 90 & AgeGroup > 50) %>% 
    ggplot(aes(x = AgeGroupLabel,
               fill = genes,label = genes,
               y = y)) +
  theme_gerstung(base_size = 12) +
  scale_fill_manual(values = gene_colours,
                    name = NULL) + 
  geom_bar(stat = "identity",color = 'black') +
  scale_x_discrete() + 
  rotate_x_text() +
  xlab("Age ranges") + 
  ylab("Cummulative frequency") + 
  ggsave(useDingbats=FALSE,sprintf("figures/%s/prevalence/Prevalence.pdf",model_id),height = 7,width = 4)
```

```{r modelling incidence,warning=F}
n_points <- load_data() %>%
  select(SardID,Phase) %>%
  distinct() %>%
  group_by(SardID) %>%
  summarise(n = length(SardID)) %>% 
  group_by(n) %>%
  summarise(N = length(n)) %>%
  mutate(N = N / sum(N))
coverage <- r_values$coverage
coverage <- coverage[coverage != 0]
coverage_distribution <- fitdistrplus::fitdist(coverage,"gamma",method = "mme")
min_coverage <- min(coverage)
BETA <- values_model$beta_values[[1]][1,1]

# sample from age distribution and not uniform 

unique_combinations <- r_values %>%
  group_by(individual) %>%
  mutate(min_age = min(age),n_timepoints = length(unique(age)),
         max_age = max(age)) %>%
  ungroup() %>%
  select(individual,site,coefficient,b_clone,u_values,genes,min_age,n_timepoints,max_age,age) %>% 
  distinct %>%
  group_by(individual) %>% 
  mutate(Phase = paste("Phase",findInterval(age,unique(age)),sep = '_')) %>% 
  spread(key = "Phase",value = "age") 

unique_individuals <- r_values %>%
  select(individual) %>%
  unlist %>%
  unique

age_diff <- full_data %>% 
  select(SardID,Age) %>% 
  distinct() %>% 
  arrange(SardID,Age) %>% 
  group_by(SardID) %>% 
  mutate(Age = c(Age) - c(Age[2:length(Age)],Age[length(Age)])) %>% 
  subset(Age != 0) %>% 
  ungroup() %>% 
  summarise(AgeDiff = mean(Age)) %>% 
  unlist %>% 
  abs

samples_through_time <- list()

for (ind in unique_individuals) {
  tmp <- unique_combinations %>%
    subset(individual == ind)
  n_timepoints <- tmp$n_timepoints[1]

  ages <- tmp[,c(10:14)] - values_model$min_age
  ages <- ages[!is.na(ages)]
  inferred_counts <- ages %>%
    sapply(
      function(x) {
        tmp %>%
          as_tibble() %>%
          mutate(p = inv.logit((coefficient + b_clone) * x + u_values) * 0.5,
                 age = x,
                 coverage = round(rgamma(length(coefficient),
                                         shape = coverage_distribution$estimate[1],
                                         rate = coverage_distribution$estimate[2]))) %>%
          mutate(coverage = ifelse(coverage < min_coverage,
                                   min_coverage,
                                   coverage)) %>%
          mutate(counts = rbbinom(1,coverage,(BETA * p)/(1 - p),BETA))
      },
      simplify = F
    ) %>%
    do.call(what = rbind) %>%
    as_tibble() %>%
    mutate(sample = counts / coverage) %>%
    select(p,individual,site,sample,age,genes,counts,coverage)
  samples_through_time[[as.character(ind)]] <- inferred_counts
}

incidence_interpolation <- samples_through_time %>%
  do.call(what = rbind) %>%
  mutate(age = age + values_model$min_age) %>%
  mutate(AgeGroup = floor(age/INTERVAL) * INTERVAL) %>%
  mutate(AgeGroupTrue = AgeGroup) %>%
  mutate(AgeGroup = factor(paste(AgeGroup,AgeGroup+INTERVAL,sep = '-'),
                           levels = paste(sort(unique(AgeGroup)),sort(unique(AgeGroup)) + INTERVAL,sep = '-'))) %>%
  mutate(genes = str_match(genes,'[A-Z0-9]+')) %>%
  group_by(AgeGroup,AgeGroupTrue) %>%
  mutate(TotalIndividuals = length(unique(individual))) %>%
  group_by(AgeGroup,AgeGroupTrue,genes,individual) %>%
  summarise(MutPresence = any(sample > DETECTION_LIMIT),
            TotalIndividuals = TotalIndividuals[1]) %>%
  group_by(AgeGroup,AgeGroupTrue,genes) %>%
  summarise(MutCount = sum(MutPresence),
            TotalIndividuals = TotalIndividuals[1]) %>%  
  mutate(y = (MutCount) / (TotalIndividuals),
         ymax = qbeta(0.99,MutCount,TotalIndividuals - MutCount + 1),
         ymin = qbeta(0.01,MutCount,TotalIndividuals - MutCount + 1))

simulated_incidence <- incidence_interpolation %>%
  subset(AgeGroupTrue <= 90 & AgeGroupTrue > 50) %>% 
  ggplot(aes(x = AgeGroup,
               fill = genes,label = genes,
               y = y)) +
  theme_gerstung(base_size = 12) +
  scale_fill_manual(values = gene_colours,
                    name = NULL) + 
  geom_bar(stat = "identity",color = 'black') +
  scale_x_discrete() + 
  rotate_x_text() + 
  xlab("Age ranges") + 
  ylab("Cummulative frequency") + 
  ggsave(useDingbats=FALSE,sprintf("figures/%s/prevalence/PrevalenceSimulated.pdf",model_id),height = 7,width = 4)
```

```{r plotting_incidences,fig.width=8,fig.height=7}
plot_grid(true_incidence + ggtitle("True incidence") + scale_y_continuous(limits = c(0,2)),
          simulated_incidence + ggtitle("Simulated incidence") + scale_y_continuous(limits = c(0,2)),
          nrow = 1) + 
  ggsave(useDingbats=FALSE,sprintf("figures/%s/prevalence/PrevalenceBoth.pdf",model_id),width = 8,height = 7) 
```

```{r distance_incidences}
cosine_dist <- function(p1,p2) {
  return(sum(p1*p2) / (sqrt(sum(p1^2))*sqrt(sum(p2^2))))
}
tmp <- incidence_interpolation %>%
  select(AgeGroupLabel=AgeGroup,AgeGroup=AgeGroupTrue,genes,MutCount,TotalIndividuals,y,ymax,ymin)

tmp <- rbind(data.frame(tmp,Type = c("Simulated")),
             data.frame(incidence,Type = c("True"))) %>% 
  subset(AgeGroup <= 90) %>%
  subset(AgeGroup > 50)

COS_D <- list()
for (ag in unique(tmp$AgeGroupLabel)) { 
  tmp_2 <- tmp %>% 
    subset(AgeGroupLabel == ag) %>%
    select(-ymax,-ymin,-MutCount) %>%
    spread(value = "y",key = "Type") %>% 
    mutate(Simulated = ifelse(is.na(Simulated),0,Simulated),
           True = ifelse(is.na(True),0,True))
  COS_D[[ag]] <- c(ag,cosine_dist(tmp_2$Simulated,tmp_2$True))
}
COS_D <- do.call(rbind,COS_D)
tibble(`Age Group` = COS_D[,1],`Cosine Similarity` = COS_D[,2])
```

Using the trajectories we can replicate the observed prevalence to some extent. We overestimate the prevalence of a few mutations early on life (SF3B1, MYD88, GNB1, TP53). However, it is reassuring to see that the broad picture is till maintained, with high cosine distances between the proportions of individuals with each mutation remaining quite high (at least 0.94).

## dN/dS associations

A popular measure of evolutionary selection in cancer is the dN/dS ratio as per Martincorena, which uses the ratio of non-synonymous mutations to synonymous mutations normalised by nucleotide context as a proxy for evolutionary selection. However, what is observable is that there is no correlation between the expected values for dN/dS and growth coefficients; indeed, a few genes appear to have higher than expected growth coefficients when considering their relative prevalence in our cohort (i.e. U2AF1-Q157R and SRSF2-P95H). 

```{r dnds_comparison, fig.height=4,fig.width=8}
values_b <- values_model$b_values[[1]]
values_b$site_name <- rep(values_model$training_subset$unique_site[values_model$gene_idxs],
                          each = length(unique(values_model$b_values[[1]]$labels)))
values_b$gene <- values_b$site_name %>% 
  sapply(function(x) unlist(str_split(x,pattern = '-'))[1])

dnds_data <- load_dnds()
dnds_site <- dnds_data$site %>%
  mutate(site_merge = paste(gene,aachange,sep = '-'))
coefficients_dnds <- values_b %>%
  spread(key = "labels",value = "values") %>%
  mutate(genes_corrected = str_match(gene,"[A-Z0-9]+")) %>%
  mutate(site_merge = paste(genes_corrected,str_match(site_name,'[a-zA-Z0-9:>;\\.\\+]+$'),sep = '-')) %>% 
  merge(dnds_site,by = "site_merge") %>%
  subset(sig == T) %>%
  select(HDPI_low,`0.025`,`0.50`,`0.975`,HDPI_high,
         "dnds","genes_corrected","site_merge","aachange") %>%
  distinct()

dnds_cor_test <- cor.test(coefficients_dnds$`0.50`,log10(coefficients_dnds$dnds))
dnds_site_plot <- coefficients_dnds %>%
  ggplot(aes(x = dnds,
             y = `0.50`,
             ymax = `0.025`,
             ymin = `0.975`)) +
  geom_smooth(method = 'lm',linetype = 2,color = 'black',
              alpha = 0.05,
              formula = y ~ x - 1) + 
  geom_point(aes(color = genes_corrected)) + 
  geom_errorbar(aes(color = genes_corrected),
                alpha=0.50) + 
  theme_gerstung() + 
  # annotate(geom = 'text',x = 1000,y = min(coefficients_dnds$`0.025`,na.rm = T),
  #          label = sprintf("Rsquared = %s\np-value = %s",
  #                          round(dnds_cor_test$estimate^2,2),
  #                          round(dnds_cor_test$p.value,2))) +
  scale_color_manual(values = gene_colours,name = NULL) + 
  theme(legend.position = "bottom") +
  xlab('log(Site-specific dN/dS)') + 
  ylab("Genetic clone growth coefficient") + 
  scale_y_continuous(limits = c(-0.2,0.5)) + 
  scale_x_continuous(trans='log10') + 
  #ggtitle("dN/dS associations with genetic growth coefficients") + 
  ggsave(useDingbats=FALSE,filename = sprintf("figures/%s/dnds/dnds.pdf",model_id),
         height = 5,width = 5)

dnds_gene <- dnds_data$genes %>%
  subset(sig == T) %>% 
  select(gene_name,
         dNdS_NT=wmis_cv,
         dNdS_NT_0025=mis_95CIlow,
         dNdS_NT_0975=mis_95CIhigh,
         dNdS_T=wnon_cv,
         dNdS_T_0025=trunc_95CIlow,
         dNdS_T_0975=trunc_95CIhigh)
dnds_gene <- rbind(
  data.frame(Gene=dnds_gene$gene_name,
             Value=dnds_gene$dNdS_NT,
             Value_0025 = dnds_gene$dNdS_NT_0025,
             Value_0975 = dnds_gene$dNdS_NT_0975,
             Truncating=F),
  data.frame(Gene=dnds_gene$gene_name,
             Value=dnds_gene$dNdS_T,
             Value_0025 = dnds_gene$dNdS_T_0025,
             Value_0975 = dnds_gene$dNdS_T_0975,
             Truncating=T)
) %>%
  mutate(Gene=as.character(Gene))

values_b_gene <- values_model$b_gene_values[[1]] 
values_b_gene$Gene <- rep(values_model$training_subset$unique_gene,
                          each = length(unique(values_model$b_gene_values[[1]]$labels)))
values_b_gene <- values_b_gene %>% 
  spread(key = "labels",value = "values") %>% 
  mutate(Truncating = grepl('[0-9A-Z]t',Gene)) %>%
  mutate(Gene = as.character(str_match(Gene,'[0-9A-Z]+'))) %>%
  select(-variable) 
coefficients_dnds_gene <- values_b_gene %>%
  merge(dnds_gene,by = c("Gene","Truncating"))
dnds_cor_test_gene <- cor.test(coefficients_dnds_gene$`0.50`,log10(coefficients_dnds_gene$Value))
dnds_gene_plot <- coefficients_dnds_gene %>%
  ggplot(aes(y=`0.50`,ymin=`0.025`,ymax=`0.975`,
             x=Value,xmin=Value_0025,xmax=Value_0975,
             color=Gene,shape = Truncating)) + 
  geom_smooth(method = 'glm',inherit.aes=F,
              mapping = aes(y=`0.50`,x=Value),
              color='black',linetype=2,
              alpha=0.05,
              formula = y ~ x - 1) + 
  geom_point() + 
  geom_errorbar(alpha=0.5) + 
  geom_errorbarh(alpha=0.5) + 
  theme_gerstung() + 
  # annotate(geom = 'text',x = 1,y = min(coefficients_dnds_gene$`0.025`),
  #          label = sprintf("Rsquared = %s\np-value = %s",
  #                          round(dnds_cor_test_gene$estimate^2,2),
  #                          round(dnds_cor_test_gene$p.value,2))) +
  scale_shape(labels = c("Non-truncating","Truncating"),name = NULL) +
  scale_color_manual(values = gene_colours,name = NULL) + 
  theme(legend.position = "bottom") +
  xlab('log(Gene dN/dS)') + 
  ylab("Gene-only clone growth coefficient") + 
  #ggtitle("dN/dS association with gene growth coefficients") + 
  scale_x_continuous(trans='log10') + 
  scale_y_continuous(limits = c(-0.2,0.35)) +
  ggsave(useDingbats=FALSE,filename = sprintf("figures/%s/dnds/dnds_gene.pdf",model_id),height = 5,width = 5) 

legend_dnds <- get_legend(dnds_gene_plot)
plot_grid(dnds_gene_plot + theme(legend.position = "none"),
          dnds_site_plot + theme(legend.position = "none")) %>%
  plot_grid(legend_dnds,rel_heights = c(1,0.30),ncol=1)
```

## Associations between VAF and blood counts and blood biochemistry

Here we calculate if/how changes in VAF are affected by changes in blood count and blood chemistry data.

```{r load data and merge}
stat_sign_individuals <- statistics_data %>%
  group_by(individual) %>%
  summarise(sig = sum(sum_stat < 0.7) > 0) %>%
  subset(sig == F)

blood_count_data <- load_blood_count_data()
fuller_data <- merge(load_data(),blood_count_data,by = c("SardID","Phase")) %>%
  subset(!(SardID %in% c(load_excluded_individuals(),load_excluded_individuals_lymph()))) %>%
  subset(!(SardID %in% stat_sign_individuals$individual))
```

```{r data preprocessing}
complete_blood_counts <- fuller_data %>%
  group_by(SardID) %>%
  mutate(relative_timepoint = Phase - min(Phase) + 1) %>% 
  select(SardID,Phase,VAF,Age,amino_acid_change,relative_timepoint,
         RBC,HB,MCV,MCH,WBC,NEUT_PERC,LYMPH_PERC,MONO_PERC,EOS_PERC,BASO_PERC,
         PLT,Glucose,SerumInsulin,BLoodUreaNitrogen,Creatinine,ALT = ALT.y,AST,GGT,Fibrinogen,
         TotalCholesterol,HDLcholesterol,Triglycerides,Iron,Transferrin,UricAcid,ESR,Phase,Age) %>%
  group_by(SardID) %>%
  mutate(complete = sum(is.na(RBC)) == 0) %>%
  ungroup() %>%
  subset(complete == T) %>%
  mutate(
    NEUT = WBC * NEUT_PERC/100,
    MONO = WBC * MONO_PERC/100,
    LYMPH = WBC * LYMPH_PERC/100,
    BASO = WBC * BASO_PERC/100,
    EOS = WBC * EOS_PERC/100
  ) %>%
  mutate(GRAN = NEUT + BASO + EOS)
```

Latent class mixture models allow us to define different classes for time series. For our case, the time series we consider here are those described by blood count and chemistry parameters through time, allowing us to stratify individuals based on their growth. This was inspired by [this work](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6819541/pdf/12889_2019_Article_7752.pdf).

```{r prepare_lcmm}
complete_blood_counts_lcmm <- complete_blood_counts %>% 
  subset(!(SardID %in% c(load_excluded_individuals(),load_excluded_individuals_lymph()))) %>%
  merge(distinct(select(full_data,SardID,Gender)),by = "SardID",all = F) %>%
  group_by(SardID) %>%
  mutate(MinAge = min(Age),
         Age = Age - min(Age),
         subject = as.numeric(SardID)) %>%
  distinct()

LINK <- "linear"
MIXTURE_FORMULA <- formula(" ~ Age + Age:factor(Gender) + factor(Gender)")

LCMM_MODELS <- list()
```

```{r lcmm_esr, fig.height=10,fig.width=5}
LCMM_MODELS$ESR <- list()

lcmm_fit <- lcmm(
  fixed = ESR ~ Age +  Age:factor(Gender) + factor(Gender),
  random = ~ 1,
  subject = "subject",
  mixture = MIXTURE_FORMULA,
  data = as.data.frame(complete_blood_counts_lcmm),
  link = LINK,
  idiag = T,
  ng = 2)

LCMM_MODELS$ESR$MODEL_FULL <- lcmm(
  fixed = ESR ~ Age +  Age:factor(Gender) + factor(Gender),
  random = ~ 1,
  subject = "subject",
  data = as.data.frame(complete_blood_counts_lcmm),
  link = LINK,
  idiag = 1)

summary(lcmm_fit)

tmp <- which.max(table(lcmm_fit$pprob$class))

if (tmp == 1) {
  lcmm_fit$pprob <- data.frame(
    subject = lcmm_fit$pprob$subject,
    class = lcmm_fit$pprob$class,
    prob1 = lcmm_fit$pprob$prob1,
    prob2 = lcmm_fit$pprob$prob2
  )
} else {
  lcmm_fit$pprob <- data.frame(
    subject = lcmm_fit$pprob$subject,
    class = 3 - lcmm_fit$pprob$class,
    prob1 = lcmm_fit$pprob$prob2,
    prob2 = lcmm_fit$pprob$prob1
  )
}

LCMM_MODELS$ESR$MODEL <- lcmm_fit
LCMM_MODELS$ESR$TRAJECTORIES <- complete_blood_counts_lcmm %>%
  merge(lcmm_fit$pprob,by = "subject") %>%
  mutate(class = factor(ifelse(class == 1,'A','B'),levels=c("A","B"))) %>%
  ggplot(aes(x = Age,y = ESR)) + 
  geom_line(aes(group = SardID),alpha = 0.5,colour = 'grey') +
  facet_wrap( ~ class) +
  geom_smooth(method = 'loess',formula = y ~ x,span = 0.8,method.args = list(degree = 1),
              colour = 'black',fill = 'grey20') + 
  theme_gerstung(base_size = 15) + 
  theme(legend.position = 'bottom') + 
  xlab("Years since study entry")

LCMM_MODELS$ESR$BOXPLOT <- r_values %>%
  select(individual,coefficient,b_clone) %>%
  group_by(individual) %>%
  filter((b_clone) == max(b_clone)) %>%
  distinct() %>%
  merge(lcmm_fit$pprob,by.x = "individual",by.y = "subject",all.y = T) %>%
  mutate(class = factor(ifelse(class == 1,'A','B'),levels=c("A","B"))) %>%
  mutate(b_clone = ifelse(is.na(b_clone),0,b_clone)) %>% 
  ggplot(aes(x = as.factor(class),y = b_clone)) + 
  geom_boxplot(alpha = 0.8) + 
  geom_signif(comparisons = list(c("A","B")),
                map_signif_level = function(x) return(paste("p-value =",round(x,4))),
              test = wilcox.test) + 
  stat_summary(geom = 'text',
               fontface = 'italic',
               fun.data = function(x) return(list(y = 0.22,label = sprintf('n = %s',length(x))))) +
  xlab("Trajectory type") + 
  ylab("Growth effect") + 
  theme_gerstung(base_size = 15) 
```

```{r lcmm_wbc, fig.height=10,fig.width=5}
LCMM_MODELS$WBC <- list()

lcmm_fit <- lcmm(
  fixed = WBC ~ Age + Age:factor(Gender) + factor(Gender),
  random = ~ 1,
  subject = "subject",
  mixture = MIXTURE_FORMULA,
  data = as.data.frame(complete_blood_counts_lcmm),
  link = LINK,
  idiag = T,
  ng = 2)

LCMM_MODELS$WBC$MODEL_FULL <- lcmm(
  fixed = WBC ~ Age +  Age:factor(Gender) + factor(Gender),
  random = ~ 1,
  subject = "subject",
  data = as.data.frame(complete_blood_counts_lcmm),
  link = LINK,
  idiag = 1)

summary(lcmm_fit)

tmp <- which.max(table(lcmm_fit$pprob$class))

if (tmp == 1) {
  lcmm_fit$pprob <- data.frame(
    subject = lcmm_fit$pprob$subject,
    class = lcmm_fit$pprob$class,
    prob1 = lcmm_fit$pprob$prob1,
    prob2 = lcmm_fit$pprob$prob2
  )
} else {
  lcmm_fit$pprob <- data.frame(
    subject = lcmm_fit$pprob$subject,
    class = 3 - lcmm_fit$pprob$class,
    prob1 = lcmm_fit$pprob$prob2,
    prob2 = lcmm_fit$pprob$prob1
  )
}

LCMM_MODELS$WBC$MODEL <- lcmm_fit
LCMM_MODELS$WBC$TRAJECTORIES <- complete_blood_counts_lcmm %>%
  merge(lcmm_fit$pprob,by = "subject") %>%
  mutate(class = factor(ifelse(class == 1,'A','B'),levels=c("A","B"))) %>%
  ggplot(aes(x = Age,y = WBC)) + 
  geom_line(aes(group = SardID),alpha = 0.5,colour = 'grey') +
  facet_wrap( ~ class) +
  geom_smooth(method = 'loess',formula = y ~ x,span = 0.8,method.args = list(degree = 1),
              colour = 'black',fill = 'grey20') + 
  theme_gerstung(base_size = 15) + 
  theme(legend.position = 'bottom') + 
  xlab("Years since study entry")

LCMM_MODELS$WBC$BOXPLOT <- r_values %>%
  select(individual,coefficient,b_clone) %>%
  group_by(individual) %>%
  filter((b_clone) == max(b_clone)) %>%
  distinct() %>%
  merge(lcmm_fit$pprob,by.x = "individual",by.y = "subject",all.y = T) %>%
  mutate(class = factor(ifelse(class == 1,'A','B'),levels=c("A","B"))) %>%
  mutate(b_clone = ifelse(is.na(b_clone),0,b_clone)) %>% 
  ggplot(aes(x = as.factor(class),y = b_clone)) + 
  geom_boxplot(alpha = 0.8) + 
  geom_signif(comparisons = list(c("A","B")),
                map_signif_level = function(x) return(paste("p-value =",round(x,4))),
              test = wilcox.test) + 
  stat_summary(geom = 'text',
               fontface = 'italic',
               fun.data = function(x) return(list(y = 0.22,label = sprintf('n = %s',length(x))))) +
  xlab("Trajectory type") + 
  ylab("Growth effect") + 
  theme_gerstung(base_size = 15) 
```

```{r lcmm_hb, fig.height=10,fig.width=5}
LCMM_MODELS$HB <- list()

lcmm_fit <- lcmm(
  fixed = HB ~ Age + Age:factor(Gender) + factor(Gender),
  random = ~ 1,
  subject = "subject",
  mixture = MIXTURE_FORMULA,
  data = as.data.frame(complete_blood_counts_lcmm),
  link = LINK,
  idiag = T,
  ng = 2)

LCMM_MODELS$HB$MODEL_FULL <- lcmm(
  fixed = HB ~ Age +  Age:factor(Gender) + factor(Gender),
  random = ~ 1,
  subject = "subject",
  data = as.data.frame(complete_blood_counts_lcmm),
  link = LINK,
  idiag = T)

summary(lcmm_fit)

tmp <- which.max(table(lcmm_fit$pprob$class))

if (tmp == 1) {
  lcmm_fit$pprob <- data.frame(
    subject = lcmm_fit$pprob$subject,
    class = lcmm_fit$pprob$class,
    prob1 = lcmm_fit$pprob$prob1,
    prob2 = lcmm_fit$pprob$prob2
  )
} else {
  lcmm_fit$pprob <- data.frame(
    subject = lcmm_fit$pprob$subject,
    class = 3 - lcmm_fit$pprob$class,
    prob1 = lcmm_fit$pprob$prob2,
    prob2 = lcmm_fit$pprob$prob1
  )
}

LCMM_MODELS$HB$MODEL <- lcmm_fit
LCMM_MODELS$HB$TRAJECTORIES <- complete_blood_counts_lcmm %>%
  merge(lcmm_fit$pprob,by = "subject") %>%
  mutate(class = factor(ifelse(class == 1,'A','B'),levels=c("A","B"))) %>%
  ggplot(aes(x = Age,y = HB)) + 
  geom_line(aes(group = SardID),alpha = 0.5,colour = 'grey') +
  facet_wrap( ~ class) +
  geom_smooth(method = 'loess',formula = y ~ x,span = 0.8,method.args = list(degree = 1),
              colour = 'black',fill = 'grey20') + 
  theme_gerstung(base_size = 15) + 
  theme(legend.position = 'bottom') + 
  xlab("Years since study entry")

LCMM_MODELS$HB$BOXPLOT <- r_values %>%
  select(individual,coefficient,b_clone) %>%
  group_by(individual) %>%
  filter((b_clone) == max(b_clone)) %>%
  distinct() %>%
  merge(lcmm_fit$pprob,by.x = "individual",by.y = "subject",all.y = T) %>%
  mutate(class = factor(ifelse(class == 1,'A','B'),levels=c("A","B"))) %>%
  mutate(b_clone = ifelse(is.na(b_clone),0,b_clone)) %>% 
  ggplot(aes(x = as.factor(class),y = b_clone)) + 
  geom_boxplot(alpha = 0.8) + 
  geom_signif(comparisons = list(c("A","B")),
                map_signif_level = function(x) return(paste("p-value =",round(x,4))),
              test = wilcox.test) + 
  stat_summary(geom = 'text',
               fontface = 'italic',
               fun.data = function(x) return(list(y = 0.22,label = sprintf('n = %s',length(x))))) +
  xlab("Trajectory type") + 
  ylab("Growth effect") + 
  theme_gerstung(base_size = 15) 
```

```{r lcmm_plt, fig.height=10,fig.width=5}
LCMM_MODELS$PLT <- list()

lcmm_fit <- lcmm(
  fixed = PLT ~ Age + Age:factor(Gender) + factor(Gender),
  random = ~ 1,
  subject = "subject",
  mixture = MIXTURE_FORMULA,
  data = as.data.frame(complete_blood_counts_lcmm),
  link = LINK,
  idiag = T,
  ng = 2)

LCMM_MODELS$PLT$MODEL_FULL <- lcmm(
  fixed = PLT ~ Age +  Age:factor(Gender) + factor(Gender),
  random = ~ 1,
  subject = "subject",
  data = as.data.frame(complete_blood_counts_lcmm),
  link = LINK,
  idiag = 1)

summary(lcmm_fit)

tmp <- which.max(table(lcmm_fit$pprob$class))

if (tmp == 1) {
  lcmm_fit$pprob <- data.frame(
    subject = lcmm_fit$pprob$subject,
    class = lcmm_fit$pprob$class,
    prob1 = lcmm_fit$pprob$prob1,
    prob2 = lcmm_fit$pprob$prob2
  )
} else {
  lcmm_fit$pprob <- data.frame(
    subject = lcmm_fit$pprob$subject,
    class = 3 - lcmm_fit$pprob$class,
    prob1 = lcmm_fit$pprob$prob2,
    prob2 = lcmm_fit$pprob$prob1
  )
}

LCMM_MODELS$PLT$MODEL <- lcmm_fit
LCMM_MODELS$PLT$TRAJECTORIES <- complete_blood_counts_lcmm %>%
  merge(lcmm_fit$pprob,by = "subject") %>%
  mutate(class = factor(ifelse(class == 1,'A','B'),levels=c("A","B"))) %>%
  ggplot(aes(x = Age,y = PLT)) + 
  geom_line(aes(group = SardID),alpha = 0.5,colour = 'grey') +
  facet_wrap( ~ class) +
  geom_smooth(method = 'loess',formula = y ~ x,span = 0.8,method.args = list(degree = 1),
              colour = 'black',fill = 'grey20') + 
  theme_gerstung(base_size = 15) + 
  theme(legend.position = 'bottom') + 
  xlab("Years since study entry")

LCMM_MODELS$PLT$BOXPLOT <- r_values %>%
  select(individual,coefficient,b_clone) %>%
  group_by(individual) %>%
  filter((b_clone) == max(b_clone)) %>%
  distinct() %>%
  merge(lcmm_fit$pprob,by.x = "individual",by.y = "subject",all.y = T) %>%
  mutate(class = factor(ifelse(class == 1,'A','B'),levels=c("A","B"))) %>%
  mutate(b_clone = ifelse(is.na(b_clone),0,b_clone)) %>% 
  ggplot(aes(x = as.factor(class),y = b_clone)) + 
  geom_boxplot(alpha = 0.8) + 
  geom_signif(comparisons = list(c("A","B")),
                map_signif_level = function(x) return(paste("p-value =",round(x,4))),
              test = wilcox.test) + 
  stat_summary(geom = 'text',
               fontface = 'italic',
               fun.data = function(x) return(list(y = 0.22,label = sprintf('n = %s',length(x))))) +
  xlab("Trajectory type") + 
  ylab("Growth effect") + 
  theme_gerstung(base_size = 15) 
```

```{r visualise_all_lcmm,fig.height = 6.5,fig.width = 14}
plot_grid(
  plot_grid(LCMM_MODELS$WBC$TRAJECTORIES + 
              theme(plot.title = element_text(hjust = 0.5)) + 
              xlab("") + 
              ylab(bquote('WBC'~(10^3/uL))),
            LCMM_MODELS$WBC$BOXPLOT + 
              xlab("") + 
              ylab("Mutation-independent effect") + 
              theme(panel.grid.major.x = element_blank()),
            ncol = 1,
            rel_heights = c(0.9,0.8)),
  plot_grid(LCMM_MODELS$HB$TRAJECTORIES + 
              theme(plot.title = element_text(hjust = 0.5)) + 
              ylab("Hb (g/dL)"),
            LCMM_MODELS$HB$BOXPLOT + 
              ylab("") + 
              theme(panel.grid.major.x = element_blank()),
            ncol = 1,
            rel_heights = c(0.9,0.8)),
  plot_grid(LCMM_MODELS$PLT$TRAJECTORIES + 
              theme(plot.title = element_text(hjust = 0.5)) + 
              xlab("") +
              ylab(bquote('Plt'~(10^3/uL))),
            LCMM_MODELS$PLT$BOXPLOT + 
              xlab("") + 
              ylab("") + 
              theme(panel.grid.major.x = element_blank()),
            ncol = 1,
            rel_heights = c(0.9,0.8)),
  plot_grid(LCMM_MODELS$ESR$TRAJECTORIES + 
              theme(plot.title = element_text(hjust = 0.5)) + 
              xlab("") + 
              ylab("ESR (mm/h)"),
            LCMM_MODELS$ESR$BOXPLOT + 
              xlab("") + 
              ylab("") + 
              theme(panel.grid.major.x = element_blank()),
            ncol = 1,
            rel_heights = c(0.9,0.8)),
  nrow = 1
) + 
  ggsave(useDingbats=FALSE,filename = sprintf("figures/%s/phenotype_associations/FigureBloodCounts.pdf",model_id),
         height = 6.5,width = 14) +
  ggsave(filename = sprintf("figures/%s/phenotype_associations/FigureBloodCounts.svg",model_id),
         height = 6.5,width = 14)
```

We can also analyse this data if we consider each blood count/chemistry parameter trajectory independently and calculate its slope and average value. These values can then be used in a simple model as predictors of unknown growth effects. We also use the LCMM model trajectory probabilities in a second linear model to assess their utility in these scenarios. 

```{r linear_model_blood_counts,fig.width=10,fig.height=4} 
model_probs <- list(select(LCMM_MODELS$ESR$MODEL$pprob,SardID=subject,ESR_traj_B=prob2), 
                    select(LCMM_MODELS$WBC$MODEL$pprob,SardID=subject,WBC_traj_B=prob2),
                    select(LCMM_MODELS$HB$MODEL$pprob,SardID=subject,HB_traj_B=prob2),
                    select(LCMM_MODELS$PLT$MODEL$pprob,SardID=subject,PLT_traj_B=prob2)
                    ) %>% 
  reduce(left_join, by = "SardID") 
           
slope <- function(x,y) {
  s_x <- sum(x)
  s_x2 <- sum(x^2)
  s_y <- sum(y)
  s_xy <- sum(x*y)
  return((s_xy - s_x * s_y) / (s_x2 - s_x^2))
}                             

data_linear_model <- merge(load_data(),blood_count_data,by = c("SardID","Phase"),all.y = T) %>% 
  group_by(SardID) %>%
  mutate(MeanAge = mean(Age)) %>% 
  group_by(SardID,Gender,MeanAge) %>%
  summarise(
    WBC_mean = mean(WBC),
    Hb_mean = mean(HB),
    Plt_mean = mean(PLT),
    ESR_mean = mean(ESR),
    WBC_slope = slope(Age,WBC),
    Hb_slope = slope(Age,HB),
    Plt_slope = slope(Age,PLT),
    ESR_slope = slope(Age,ESR)
  ) %>% 
  merge(distinct(select(r_values,"individual","b_clone","coefficient","site")),by.x = "SardID",by.y = "individual",all.x = T)  %>% 
  merge(model_probs,by = "SardID",all=T)

data_linear_model <- data_linear_model[
  apply(data_linear_model[,4:11],1,function(x) !any(is.na(x))),
] 
data_linear_model[,c(3:11,15:18)] <- apply(data_linear_model[,c(3:11,15:18)],2,scale)
data_linear_model <- data_linear_model %>% 
  mutate(b_clone = ifelse(is.na(b_clone),0,b_clone),
         coefficient = exp(ifelse(is.na(coefficient),0,coefficient))) %>%
  group_by(SardID) %>%
  filter(abs(b_clone) == max(abs(b_clone)))

linear_model_slopes <- lm(
  b_clone ~ WBC_mean + Hb_mean + Plt_mean + ESR_mean + WBC_slope + Hb_slope + Plt_slope + ESR_slope + MeanAge + Gender,
  data = data_linear_model) 

linear_model_traj_probs <- lm(
  b_clone ~ WBC_mean + Hb_mean + Plt_mean + ESR_mean + ESR_traj_B + WBC_traj_B + HB_traj_B + PLT_traj_B + MeanAge + Gender,
  data = data_linear_model
)

lm_res_plot <- data.frame(
  b_clone = data_linear_model$b_clone,
  Slopes = predict(linear_model_slopes,data_linear_model),
  Trajectories = predict(linear_model_traj_probs,data_linear_model)
) %>% 
  gather("key","value",Slopes,Trajectories) %>%
  ggplot(aes(x = b_clone,y = b_clone)) +
  geom_abline(slope = 1,size = 1.5) +
  geom_segment(aes(xend = b_clone,yend = value,colour = key),alpha = 0.2) + 
  geom_point(aes(y = value,colour = key),alpha=0.5) + 
  scale_color_aaas(name = "Model") + 
  theme_gerstung(base_size = 15) + 
  theme(legend.position = "bottom") + 
  xlab("Unknown effect") + 
  ylab("Unknown effect")

lmss <- summary(linear_model_slopes)
lmtps <- summary(linear_model_traj_probs)
lm_coef_plot <- data.frame(coefficient = c(lmss$coefficients[,1],lmtps$coefficients[,1]),
           std_error = c(lmss$coefficients[,2],lmtps$coefficients[,2]),
           p.val = c(lmss$coefficients[,4],lmtps$coefficients[,4]),
           names = c(rownames(lmss$coefficients),rownames(lmtps$coefficients)),
           model = c(rep(sprintf("Slopes\n(Adj. R sq. = %.4f)",lmss$adj.r.squared),nrow(lmss$coefficients)),
                     rep(sprintf("Trajectories\n(Adj. R sq. = %.4f)",lmtps$adj.r.squared),nrow(lmtps$coefficients)))
           ) %>% 
  mutate(signif = cut(p.val,breaks = c(0,0.001,0.01,0.05,0.1,1),labels = c("***","**","*",".","n.s."))) %>%
  mutate(namesGroup = ifelse(
    grepl("_mean",names),
    "Mean",
    ifelse(
      grepl("_slope",names),
      "Slope",
      ifelse(
        grepl("_traj_",names),
        "Trajectories",
        "Other"
      )
    )
  )) %>% 
  mutate(namesGroup = factor(namesGroup,levels = c("Mean","Slope","Trajectories","Other"))) %>% 
  mutate(names = str_match(names,"[A-Za-z()]+")) %>% 
  ggplot(aes(x = names,y = coefficient,colour = model)) + 
  geom_hline(yintercept = 0,alpha = 0.7) +
  geom_point() + 
  geom_errorbar(aes(ymin = coefficient - std_error,ymax = coefficient + std_error),
                width = 0.2) + 
  geom_text(aes(label = signif,y = coefficient + std_error + 0.02),size=3) +
  theme_gerstung(base_size = 15) + 
  facet_wrap(~ model + namesGroup,scales = "free_x",nrow = 1) + 
  rotate_x_text() + 
  xlab("Parameters") + 
  ylab("Coefficient") + 
  scale_color_aaas() + 
  theme(legend.position = "none",
        strip.text.x = element_text(margin = margin(0,0,0,0, "cm")))

plot_grid(lm_res_plot + theme(legend.position = "none"),lm_coef_plot,rel_widths = c(2,3.8)) %>%
  plot_grid(get_legend(lm_res_plot),ncol=1,rel_heights = c(1,0.1))
```

From this analysis we can observe that using the slopes in this context does allows us to make relatively little conclusions between clonal growth and other factors (average parameter value ,trajectory membership and slopes). Apart from a positive association with Plt slope and a negative association with Plt mean in the slopes model, little else can be inferred from this.

```{r compare_trajectories}
trajectory_class_data <- rbind(
  data.frame(LCMM_MODELS$ESR$MODEL$pprob[,c(1,2)],
             Parameter = "ESR"),
  data.frame(LCMM_MODELS$WBC$MODEL$pprob[,c(1,2)],
             Parameter = "WBC"),
  data.frame(LCMM_MODELS$HB$MODEL$pprob[,c(1,2)],
             Parameter = "Hb"),
  data.frame(LCMM_MODELS$PLT$MODEL$pprob[,c(1,2)],
             Parameter = "Plt")
) %>%
  group_by(subject) %>%
  mutate(N = sum(class == 2)) %>%
  ungroup() %>%
  arrange(N) %>%
  mutate(subject = factor(subject,
                          levels = rev(unique(subject)),
                          ordered = T)) 

N_trajectory_plot <- trajectory_class_data %>%
  ggplot(aes(y = "N",x = subject,fill = N)) + 
  geom_tile() +
  scale_fill_material(palette = "blue") + 
  theme_gerstung() +
  scale_x_discrete(expand = c(0,0)) + 
  scale_y_discrete(expand = c(0,0)) + 
  theme(axis.text.x = element_blank(),
        legend.position = 'none') + 
  ylab("") + 
  xlab("")

class_trajectory_plot <- trajectory_class_data %>%
  ggplot(aes(x = subject,y = Parameter,fill = as.factor(class))) + 
  geom_tile(color = 'black') + 
  theme_gerstung() + 
  scale_x_discrete(expand = c(0,0)) + 
  scale_y_discrete(expand = c(0,0)) +
  theme(axis.text.x = element_blank(),
        legend.position = 'bottom') +
  scale_fill_lancet(name="Trajectory") + 
  xlab("Individual")

plot_grid(
  N_trajectory_plot,
  class_trajectory_plot,
  ncol=1,
  rel_heights = c(0.15,1.0),align = T
)
```

```{r table_trajectory_coefficients}
trajectory_coefficient_data <- rbind(
  data.frame(coeff = c(coef(LCMM_MODELS$WBC$MODEL)[c(3,4)],
                       coef(LCMM_MODELS$WBC$MODEL_FULL)[1]),
             Parameter = "WBC",
             class = c(1,2,"Full")),
  data.frame(coeff = c(coef(LCMM_MODELS$HB$MODEL)[c(3,4)],
                       coef(LCMM_MODELS$HB$MODEL_FULL)[1]),
             Parameter = "Hb",
             class = c(1,2,"Full")),
  data.frame(coeff = c(coef(LCMM_MODELS$PLT$MODEL)[c(3,4)],
                       coef(LCMM_MODELS$PLT$MODEL_FULL)[1]),
             Parameter = "Plt",
             class = c(1,2,"Full")),
  data.frame(coeff = c(coef(LCMM_MODELS$ESR$MODEL)[c(3,4)],
                       coef(LCMM_MODELS$ESR$MODEL_FULL)[1]),
             Parameter = "ESR",
             class = c(1,2,"Full"))
)

trajectory_coefficient_data %>% 
  spread(key = "class",value = "coeff") %>%
  mutate_if(.predicate = is.numeric,function(x) round(x,4))
```

To better assess the relevance of these data we need to compare them with reference values. A few studies ([here](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3446744/) and [here](https://www.jstage.jst.go.jp/article/geriatrics1964/28/4/28_4_509/_pdf/-char/ja)) have assessed this, so we can use them to approximate a reference value. 

```{r reference_values}
reference_values_women_HD <- data.frame(
  Decade = c("50-59","60-69","70-79","80-89","+90"),
  Decade_numeric = c(5,6,7,8,9),
  Hb = c(12.1,10.5,10.2,9.3,8.8),
  Plt = c(442.40,385.91,443.72,388.04,360.02),
  WBC = c(4.6,3.2,4.5,4.2,4.0),
  N = c(122,100,121,111,49),
  Sex = "F",
  Place = "HD"
)
reference_values_men_HD <- data.frame(
  Decade = c("50-59","60-69","70-79","80-89","+90"),
  Decade_numeric = c(5,6,7,8,9),
  Hb = c(12.1,11.7,9.9,9.6,9.7),
  Plt = c(136.5,123.9,82.8,104.8,137.4),
  WBC = c(4.3,3.9,4.1,4.2,NA),
  N = c(99,102,105,113,20),
  Sex = "M",
  Place = "HD"
)

reference_values_women_JP_1 <- data.frame(
  Decade = c("50-59","60-69","70-79","80-89","+90"),
  Decade_numeric = c(5,6,7,8,9),
  Hb = c(13.6,13.3,NA,NA,NA),
  Plt = c(235,244,NA,NA,NA),
  WBC = c(5.4,5.5,NA,NA,NA),
  N = c(152,33,NA,NA,NA),
  Sex = "F",
  Place = "JP"
)
reference_values_men_JP_1 <- data.frame(
  Decade = c("50-59","60-69","70-79","80-89","+90"),
  Decade_numeric = c(5,6,7,8,9),
  Hb = c(15.3,15.0,14.6,NA,NA),
  Plt = c(231,232,237,NA,NA),
  WBC = c(6.3,6.1,5.9,NA,NA),
  N = c(300,110,6,NA,NA),
  Sex = "M",
  Place = "JP"
)

reference_values_women_JP_2 <- data.frame(
  Decade = c("50-59","60-69","70-79","80-89","+90"),
  Decade_numeric = c(5,6,7,8,9),
  Hb = c(NA,13.8,13.8,13.4,13.5),
  Plt = c(NA,279,268,269,205),
  WBC = c(NA,5.9,5.8,5.8,5.3),
  N = c(NA,81,208,109,4),
  Sex = "F",
  Place = "JP"
)
reference_values_men_JP_2 <- data.frame(
  Decade = c("50-59","60-69","70-79","80-89","+90"),
  Decade_numeric = c(5,6,7,8,9),
  Hb = c(NA,13.8,13.8,13.4,13.5),
  Plt = c(NA,269,261,239,269),
  WBC = c(NA,6.7,6.3,6.6,6.2),
  N = c(NA,72,106,61,4),
  Sex = "M",
  Place = "JP"
)

reference_values <- rbind(
  reference_values_women_HD,
  reference_values_women_JP_1,
  reference_values_women_JP_2,
  reference_values_men_HD,
  reference_values_men_JP_1,
  reference_values_men_JP_2
)

glm(Hb ~ Decade_numeric + Place,data = reference_values,weights = reference_values$N) %>%
  summary
glm(Plt ~ Decade_numeric + Place,data = reference_values,weights = reference_values$N) %>%
  summary 
glm(WBC ~ Decade_numeric + Place,data = reference_values,weights = reference_values$N) %>%
  summary
```

## Sex, smoking and other factors

```{r differences_sex, fig.width=15,fig.height=5}
fisher_exact_factor <- function(x,totals) {
  pres <- table(x)
  levels_pres <- names(pres)
  levels_totals <- names(totals)
  DF <- data.frame(
    level = c(levels_pres,levels_totals),
    counts = c(pres,totals),
    P = c(rep(T,length(levels_pres)),rep(F,length(totals)))
  ) 
  DF <- DF %>%
    spread(key = 'level',value = 'counts')
  pres <- gtools::na.replace(DF[2,2:ncol(DF)],0)
  not_pres <- gtools::na.replace(DF[1,2:ncol(DF)],0) - gtools::na.replace(DF[2,2:ncol(DF)],0)
  out <- fisher.test(t(rbind(pres,not_pres)))
  return(out)
}

totals <- table(distinct(select(full_data,SardID,Gender))$Gender)

p_val_comparisons <- full_data %>%
  select(individual = SardID,Gender,Gene) %>% 
  distinct %>%
  #mutate(Gender = ifelse(Gender == 'F','Female','Male')) %>%
  mutate(Gene = str_match(Gene,'[A-Z0-9]+')) %>%
  group_by(Gene,Gender) %>%
  mutate(size = length(Gene)) %>%
  group_by(Gene) %>%
  summarise(p.val = fisher_exact_factor(Gender,totals)$p.value,
            max_total = max(size),
            min_total = min(size)) %>%
  mutate(p.val.label = ifelse(p.val < 0.05 / length(unique(full_data$Gene)),'*',NA))

effect_distribution_gender <- full_data %>%
  select(individual = SardID,Gender) %>% 
  distinct %>%
  merge(r_values %>% select(individual,site,coefficient,b_clone,genes),by = c('individual')) %>%
  group_by(individual) %>%
  mutate(effect = coefficient) %>% 
  distinct() %>% 
  filter(effect == max(effect)) %>%
  mutate(Gender = ifelse(Gender == 'F','Female','Male')) %>%
  mutate(Truncating = !grepl("nt",genes)) %>%
  mutate(Truncating = ifelse(is.na(Truncating),F,T)) %>%
  mutate(genes = str_match(genes,'[A-Z0-9]+')) %>%
  ggplot(aes(x = as.factor(Gender),y = effect)) + 
  geom_boxplot(alpha = 0.8) + 
  theme_gerstung(base_size = 15) + 
  geom_signif(comparisons = list(c("Male","Female")),
              map_signif_level = function(x) return(paste("p-value =",format(x,scientifier = T,digits = 4))),
              test = wilcox.test) + 
  ylab("Genetic effect") + 
  xlab("Sex") 

gene_mutation_distribution_gender <- full_data %>%
  select(individual = SardID,Gender,Gene) %>% 
  distinct %>%
  mutate(Gender = ifelse(Gender == 'F','Female','Male')) %>%
  mutate(Gene = str_match(Gene,'[A-Z0-9]+')) %>%
  ggplot(aes(x = Gene,fill = Gender)) + 
  geom_bar(stat = "count",position = 'dodge') + 
  geom_text(data = p_val_comparisons,
            inherit.aes = F,
            aes(x = Gene,label = p.val.label,y = max_total+1),
            size = 5) +
  theme_gerstung(base_size = 15) +
  scale_fill_lancet(name = NULL) + 
  rotate_x_text() + 
  theme(legend.position = "bottom") + 
  ylab("Count") + 
  xlab("Gene")

age_distribution_gender <- full_data %>%
  select(individual = SardID,Gender) %>% 
  distinct %>%
  merge(r_values %>% select(age,individual,site,coefficient,b_clone,genes),by = c('individual')) %>%
  mutate(full_effect = coefficient + b_clone) %>% 
  group_by(individual) %>%
  filter(coefficient == max(coefficient)) %>%
  mutate(Gender = ifelse(Gender == 'F','Female','Male')) %>%
  mutate(Truncating = !grepl("nt",genes)) %>%
  mutate(Truncating = ifelse(is.na(Truncating),F,T)) %>%
  mutate(genes = str_match(genes,'[A-Z0-9]+')) %>%
  ggplot(aes(x = Gender,y = age,fill = Gender)) + 
  geom_boxplot() + 
  theme_gerstung(base_size = 15) +
  scale_fill_lancet(name = NULL) + 
  rotate_x_text() + 
  theme(legend.position = "bottom") + 
  ylab("Mutation count") + 
  xlab("Gene")

effect_distribution_gender %>% 
  ggsave(useDingbats=FALSE,filename = sprintf("figures/%s/phenotype_associations/FigureGender.pdf",model_id),
         width = 3.5,height = 4)
effect_distribution_gender %>% 
  ggsave(filename = sprintf("figures/%s/phenotype_associations/FigureGender.svg",model_id),
         width = 3.5,height = 4)

plot_grid(gene_mutation_distribution_gender,effect_distribution_gender,nrow = 1,rel_widths = c(1,0.3)) + 
  ggsave(useDingbats=FALSE,filename = sprintf("figures/%s/phenotype_associations/FigureGenderWithBarplot.pdf",model_id),
         width = 15,height = 4.5)
```

```{r untangling_mf_differences,fig.width=14,fig.height=5}
p_values_gender <- merge(
  r_values %>% 
    select(individual,site,coefficient,genes,b_clone) %>% 
    distinct %>% 
    merge(distinct(select(full_data,individual=SardID,Gender)),by = "individual") %>% 
    mutate(genes = str_match(genes,'[A-Z0-9]+')) %>%
    filter(!is.na(coefficient)) %>%
    group_by(genes) %>%
    summarise(
      pval.coefficient = ifelse(
        all(c("M","F") %in% Gender),
        wilcox.test(coefficient[Gender == 'M'],coefficient[Gender == 'F'],exact = F)$p.val,
        NA)),
  r_values %>% 
    select(individual,site,coefficient,genes,b_clone) %>% 
    distinct %>% 
    merge(distinct(select(full_data,individual=SardID,Gender)),by = "individual") %>% 
    mutate(genes = str_match(genes,'[A-Z0-9]+')) %>%
    group_by(genes) %>%
    summarise(
      pval.b_clone = ifelse(
        all(c("M","F") %in% Gender),
        wilcox.test(b_clone[Gender == 'M'],b_clone[Gender == 'F'],exact = F)$p.val,
        NA)),
  by = "genes"
)

clone_specific_gender_plot <- r_values %>% 
  select(individual,site,coefficient,genes,b_clone) %>% 
  distinct %>% 
  merge(distinct(select(full_data,individual=SardID,Gender)),by = "individual") %>% 
  mutate(genes = str_match(genes,'[A-Z0-9]+')) %>% 
  ggplot(aes(x = genes,y = b_clone,fill = Gender)) + 
  geom_boxplot(lwd=0.2,outlier.size = 0.5) + 
  theme_gerstung() + 
  rotate_x_text() + 
  scale_fill_lancet(name = NULL,labels = c("Female","Male")) +
  theme(legend.position = "bottom") + 
  xlab("Genes") + 
  ylab("Mutation-independent\ngrowth effect")

vaf_gender_plot <- r_values %>% 
  select(individual,site,coefficient,genes,b_clone,true,coverage,age) %>% 
  group_by(individual,site) %>%
  filter(age == max(age)) %>% 
  merge(distinct(select(full_data,individual=SardID,Gender)),by = "individual") %>% 
  mutate(genes = str_match(genes,'[A-Z0-9]+')) %>% 
  ggplot(aes(x = genes,y = true/coverage,fill = Gender)) + 
  geom_boxplot(lwd=0.2,outlier.size = 0.5) + 
  theme_gerstung() + 
  rotate_x_text() + 
  scale_fill_lancet(name = NULL,labels = c("Female","Male")) +
  theme(legend.position = "bottom") + 
  xlab("Genes") + 
  ylab("VAF") + scale_y_continuous(
    trans = 'log10',
    minor_breaks = c(sapply(10^c(-4:-1),function(x) seq(x*2,x*10-x,length.out = 8))))

clone_ages_gender_plot <- clone_ages %>% 
  select(site,individual,genes,CloneAgeHenryLower_adj,CloneAgeHenryUpper_adj) %>%
  distinct() %>%
  subset(abs(CloneAgeHenryLower_adj) < 120 & CloneAgeHenryUpper_adj > -10) %>%
  merge(distinct(select(full_data,individual=SardID,Gender)),by = "individual") %>%
  mutate(genes = str_match(genes,'[A-Z0-9]+')) %>% 
  ggplot(aes(x = genes,y = (CloneAgeHenryLower_adj+CloneAgeHenryUpper_adj)/2,fill = Gender)) + 
  geom_boxplot(lwd=0.2,outlier.size = 0.5) + 
  theme_gerstung() + 
  rotate_x_text() + 
  scale_fill_lancet(name = NULL,labels = c("Female","Male")) +
  theme(legend.position = "bottom") + 
  xlab("Genes") + 
  ylab("Clone age")

full_data %>%
  subset(grepl('DNMT3A',Gene)) %>%
  ggplot(aes(x = Age,y = VAF,group = paste(SardID,mutation_identifier),color = Gender)) + 
  geom_line() + 
  theme_gerstung() +
  scale_color_lancet() +
  scale_y_continuous(trans = 'log10') + 
  facet_wrap(~ Gender,nrow = 1)

plot_grid(clone_specific_gender_plot,vaf_gender_plot,clone_ages_gender_plot,nrow = 1)
```

```{r smoking_gene_composition}
load_smoking_data() %>% 
  mutate(status = ifelse(QuitSmoking == 'N' & CurrentSmoker == 'N',"Never smoked",
                         ifelse(QuitSmoking == 'N' & CurrentSmoker == 'Y',"Smoker",
                                "Previous smoker"))) %>% 
  subset(!is.na(CurrentSmoker)) %>% 
  group_by(SardID) %>%
  summarise(
    status = ifelse(
      sum(status == 'Smoker') > 0,'Smoked during study',
      ifelse(sum(status == 'Previous smoker') > 0, 'Smoked before study',
             'Never smoked')
    )
  ) %>% 
  mutate(status_binary = ifelse(status == 'Never smoked','Never smoked','Smoked')) %>% 
  merge(clone_ages,by.x = "SardID",by.y = "individual") %>% 
  select(genes,SardID,status,status_binary) %>%
  distinct() %>%
  mutate(genes = str_match(genes,'[A-Z0-9]+')) %>%
  distinct() %>%
  ggplot(aes(x = genes,fill = status_binary)) +
  geom_bar(position = 'dodge') + 
  #geom_text(aes(label = total)) +
  theme_gerstung() + 
  theme(legend.position = 'bottom') + 
  rotate_x_text() + 
  xlab("Gene") + 
  ylab("Enrichment") +
  facet_wrap(~ status_binary,scales = "free_y") +
  scale_fill_lancet(name = NULL)
```

```{r differences_smoking_coefficient, fig.height=7,fig.width=6}
load_smoking_data() %>% 
  mutate(status = ifelse(QuitSmoking == 'N' & CurrentSmoker == 'N',"Never smoked",
                         ifelse(QuitSmoking == 'N' & CurrentSmoker == 'Y',"Smoker",
                                "Previous smoker"))) %>% 
  subset(!is.na(CurrentSmoker)) %>% 
  group_by(SardID) %>%
  summarise(
    status = ifelse(
      sum(status == 'Smoker') > 0,'During',
      ifelse(sum(status == 'Previous smoker') > 0, 'Before',
             'Never')
    )
  ) %>% 
  mutate(status = factor(status,levels = c("Never","Before","During"))) %>%
  merge(r_values,by.x = "SardID",by.y = "individual") %>%
  mutate(individual = SardID) %>%
  merge(statistics_data,by = c("individual","site")) %>%
  subset(sum_stat < 0.7) %>% 
  mutate(genes = str_match(genes,'[A-Z0-9]+')) %>%
  group_by(SardID) %>%
  mutate(effect = coefficient) %>%
  filter(effect == max(effect)) %>% 
  select(SardID,effect,status) %>%
  distinct() %>% 
  ggplot(aes(x = status,y = effect)) + 
  #geom_point(alpha = 0.8) +
  geom_boxplot(alpha = 0.8) +
  theme_gerstung(base_size = 15) +
  geom_signif(comparisons = list(c("Never","Before")),
              y_position = 0.45,
              #map_signif_level = function(x) return(format(x,scientifier = T,digits = 4)),
              test = wilcox.test) + 
  geom_signif(comparisons = list(c("Never","During")),
              #map_signif_level = function(x) return(format(x,scientifier = T,digits = 4)),
              test = wilcox.test) +
  geom_signif(comparisons = list(c("Before","During")),
              y_position = 0.45,
              #map_signif_level = function(x) return(format(x,scientifier = T,digits = 4)),
              test = wilcox.test) + 
  theme(legend.position = 'bottom') + 
  scale_color_manual(values = gene_colours,name = NULL) + 
  xlab("Smoked") + 
  ylab("Genetic effect") + 
  ggsave(useDingbats=FALSE,filename = sprintf("figures/%s/phenotype_associations/FigureSmoking.pdf",model_id),
         width = 2.8,height = 4) +
  ggsave(filename = sprintf("figures/%s/phenotype_associations/FigureSmoking.svg",model_id),
         width = 2.8,height = 4)

load_smoking_data() %>% 
  mutate(status = ifelse(QuitSmoking == 'N' & CurrentSmoker == 'N',"Never smoked",
                         ifelse(QuitSmoking == 'N' & CurrentSmoker == 'Y',"Smoker",
                                "Previous smoker"))) %>% 
  subset(!is.na(CurrentSmoker)) %>% 
  group_by(SardID) %>%
  summarise(
    status = ifelse(
      sum(status == 'Smoker') > 0,'During',
      ifelse(sum(status == 'Previous smoker') > 0, 'Before',
             'Never')
    )
  ) %>% 
  mutate(status = factor(status,levels = c("Never","Before","During"))) %>%
  merge(r_values,by.x = "SardID",by.y = "individual") %>%
  mutate(individual = SardID) %>%
  merge(statistics_data,by = c("individual","site")) %>%
  subset(sum_stat < 0.7) %>% 
  mutate(genes = str_match(genes,'[A-Z0-9]+')) %>%
  group_by(SardID) %>%
  mutate(effect = coefficient) %>%
  filter(effect == max(effect)) %>% 
  select(SardID,effect,status) %>%
  mutate(status = ifelse(status == 'Never','No','Yes')) %>%
  distinct() %>% 
  ggplot(aes(x = status,y = effect)) + 
  #geom_point(alpha = 0.8) +
  geom_boxplot(alpha = 0.8) +
  theme_gerstung(base_size = 15) +
  geom_signif(comparisons = list(c("Yes","No")),
              map_signif_level = function(x) return(paste("p-value =",format(x,scientifier = T,digits = 4))),
              test = wilcox.test) + 
  theme(legend.position = 'bottom') + 
  scale_color_manual(values = gene_colours,name = NULL) + 
  xlab("Smoked") + 
  ylab("Genetic effect") + 
  ggsave(filename = sprintf("figures/%s/phenotype_associations/FigureSmoking_2levels.svg",model_id),
         width = 3.5,height = 4) +
  ggsave(useDingbats=FALSE,filename = sprintf("figures/%s/phenotype_associations/FigureSmoking_2levels.pdf",model_id),
         width = 3.5,height = 4)
```

```{r smoking_clone_age, fig.width=8, fig.height=3}
INTERVAL <- 5
smoking_clone_ages <- load_smoking_data() %>% 
  mutate(status = ifelse(QuitSmoking == 'N' & CurrentSmoker == 'N',"Never smoked",
                         ifelse(QuitSmoking == 'N' & CurrentSmoker == 'Y',"Smoker",
                                "Previous smoker"))) %>% 
  subset(!is.na(CurrentSmoker)) %>% 
  group_by(SardID) %>%
  filter(phase == max(phase)) %>%
  ungroup() %>%
  merge(clone_ages,by.x = "SardID",by.y = "individual") %>%
  merge(select(subset(load_data_keep(),Gene %in% load_included_genes() | is.na(Gene)),Gender,SardID),
        by = "SardID",all.y = T) %>%
  mutate(individual = SardID) %>%
  #merge(statistics_data,by = c("individual","site")) %>%
  mutate(genes = str_match(genes,'[A-Z0-9]+')) %>%
  group_by(SardID) %>%
  mutate(effect = (CloneAgeHenryLower_adj + CloneAgeHenryUpper_adj)/2) %>%
  filter(age == max(age)) %>%
  select(SardID,effect,site,CurrentSmoker, QuitYears,status,age,genes,Gender) %>% 
  subset(effect > 18 & effect < 100) %>% 
  mutate(effect = age - effect) %>% 
  mutate(QuitYears = ifelse(CurrentSmoker == 'Y',0,QuitYears)) %>% 
  mutate(CurrentlySmoking = effect > QuitYears) %>%
  mutate(CurrentlySmoking = ifelse(status == "Never smoked",FALSE,CurrentlySmoking)) %>%
  mutate(AgeGroups = floor((age - effect) / INTERVAL) * INTERVAL)

smoking_clone_ages %>% 
  ungroup() %>%
  subset(AgeGroups <= 60) %>% 
  mutate(AgeGroups = paste(AgeGroups,AgeGroups + INTERVAL,sep='-')) %>% 
  mutate(AgeGroups = ifelse(AgeGroups == '10-20','18-20',AgeGroups)) %>%
  group_by(SardID,AgeGroups) %>%
  distinct() %>%
  summarise(NClones = length(site),
            CurrentlySmoking = CurrentlySmoking[1]) %>%
  ggplot(aes(x = AgeGroups,
             y = NClones,
             fill = CurrentlySmoking,
             colour = CurrentlySmoking,
             group = paste(AgeGroups,CurrentlySmoking))) +
  stat_summary(geom = 'line',
               inherit.aes = F,
               alpha = 0.7,
               mapping = aes(x = AgeGroups,
                             y = NClones,
                             colour = CurrentlySmoking,
                             group = CurrentlySmoking),
               fun.y = mean,
               size = 1) +
  # geom_point(position = position_jitterdodge(jitter.width = 0.4,
  #                                            dodge.width = 0.8,
  #                                            jitter.height = 0.02),
  #            alpha = 0.6) +
  
  theme_gerstung(base_size = 15) +
  #theme(legend.position = 'bottom') +
  scale_colour_d3(labels = c("No","Yes"),name = "Smoking") +
  scale_fill_d3(labels = c("No","Yes"),name = "Smoking") + 
  #rotate_x_text() +
  xlab("Age groups") + 
  ylab("No. of clones\nappearing") + 
  #scale_y_continuous(trans = 'log10') +
  theme(axis.text.x = element_text(size = 11.2)) + 
  ggsave(useDingbats=FALSE,sprintf("figures/%s/phenotype_associations/FigureCloneSmoking.pdf",model_id),height = 4,width = 7) + 
  ggsave(sprintf("figures/%s/phenotype_associations/FigureCloneSmoking.svg",model_id),height = 4,width = 7)
```

```{r smoking_clone_age_boxplot}
load_smoking_data() %>% 
  mutate(status = ifelse(QuitSmoking == 'N' & CurrentSmoker == 'N',"Never smoked",
                         ifelse(QuitSmoking == 'N' & CurrentSmoker == 'Y',"Smoker",
                                "Previous smoker"))) %>% 
  subset(!is.na(CurrentSmoker)) %>% 
  group_by(SardID) %>%
  filter(phase == max(phase)) %>%
  ungroup() %>%
  merge(clone_ages,by.x = "SardID",by.y = "individual") %>%
  mutate(individual = SardID) %>%
  merge(statistics_data,by = c("individual","site")) %>%
  mutate(genes = str_match(genes,'[A-Z0-9]+')) %>% 
  group_by(SardID) %>%
  mutate(effect = (CloneAgeHenryLower_adj + CloneAgeHenryUpper_adj)/2) %>%
  filter(age == max(age)) %>%
  select(SardID,effect,site,CurrentSmoker, QuitYears,status,age) %>% 
  subset(effect > 18 & effect < 100) %>% 
  mutate(effect = age - effect) %>% 
  mutate(QuitYears = ifelse(CurrentSmoker == 'Y',0,QuitYears)) %>% 
  mutate(CurrentlySmoking = effect > QuitYears) %>%
  mutate(CurrentlySmoking = ifelse(status == "Never smoked",FALSE,CurrentlySmoking)) %>%
  ggplot(aes(x = 0,y = effect,colour = CurrentlySmoking)) + 
  geom_boxplot() + 
  theme_gerstung()
```

## Is the unknown cause effect in JAK2 provoked by specific genotypes?

We have access to gene dosage data for different mutations which relate with JAK2 oncogenesis. We can, as such, test if they are in any way or form correlated with the unknown cause effect in JAK2, a gene whose unknown cause effects appear to be quite varied.

```{r}
jak2_phenotype <- load_jak2_phenotype() %>%
  na.omit()
jak2_phenotype %>%
  na.omit() %>% 
  ggplot(aes(x = Genotype)) +
  geom_bar() + 
  facet_wrap(~ sprintf('%s (chr. %s)',Site,Chromosome),scales = 'free_x') + 
  theme_gerstung(base_size = 15) + 
  scale_y_continuous(breaks = seq(0,20,by=2)) + 
  ylab("Number of individuals")

jak2_phenotype_and_effect <- r_values %>%
  select(b_clone,SardID=individual,genes,age) %>%
  group_by(SardID) %>%
  filter(age == min(age)) %>% 
  distinct() %>% 
  subset(genes == 'JAK2') %>%
  merge(jak2_phenotype,by = "SardID") %>%
  select(-Dose) %>% 
  mutate(Chromosome = paste('Chr',Chromosome,sep='')) %>% 
  pivot_wider(names_from = c(Chromosome,Site),values_from = Genotype,names_sep = 'Site')

jak2_phenotype_and_effect %>% 
  gather("key","value",-SardID,-b_clone,-genes,-age) %>% 
  ggplot(aes(x = value,y = b_clone)) + 
  geom_point() +
  theme_gerstung(base_size = 15) + 
  facet_wrap(~ key,scales = 'free_x') + 
  xlab("Genotype") + 
  ylab("Unknown cause effect")

iou_genotype <- apply(t(combn(colnames(jak2_phenotype_and_effect)[5:16],2)),1,
      function(x) {
        tmp <- data.frame(A = jak2_phenotype_and_effect[,x[1]],
                          B = jak2_phenotype_and_effect[,x[2]]) %>% 
          table
        I <- sum(diag(tmp))
        U <- sum(tmp)
        return(c(A=x[1],B=x[2],IoU=I/U))
      }) %>% 
  t %>%
  as.data.frame(stringsAsFactors=F) %>%
  mutate(IoU = as.numeric(as.character(IoU)))

all_chr_site <- colnames(jak2_phenotype_and_effect)[c(5:16)]
all_chr_site <- all_chr_site[!(all_chr_site %in% iou_genotype$A[iou_genotype$IoU == 1])]

lm_formula <- formula(
  paste('b_clone',paste(c(all_chr_site),collapse = ' + '),sep=' ~ ')
)

lm(formula = lm_formula,data = jak2_phenotype_and_effect) %>%
  summary
```

From the data we have it is hard to conclude anything regarding the effect of different genotypes in differences in JAK2 dynamics. Looking at dot plots, there appears to be different distributions between the three different genotypes in site 1287194 in chromosome 5, but the very small sample size does not allow us to draw any conclusions from this.

## Is the U2AF1 prevalence in blood cells we observe in our cohort normal or explained by the old age?

To assess this we need to get the prevalence of U2AF1 in blood cells, which we obtain by marginalising over AML, CML, MDS and CH

```{r checking_U2AF1_prevalence}
u2af1_CH_list <- list(
  "Jaiswal 2014" = c(n_mutations = 0,total_individuals = 17182),
  "Genovese 2014" = c(n_mutations = 0,total_individuals = 12380),
  "McKerrell 2015" = c(n_mutations = NA,total_individuals = 4219),
  "Zink 2017" = c(n_mutations = 0, total_individuals = 11262),
  "Acuna-Hidalgo" = c(n_mutations = 1, total_individuals = 9350),
  "Coombs 2017" = c(n_mutations = 1,total_individuals = 6898), # a few other U2AF1 mutations, but only in individuals which developed a malignant haematological condition
  "Young 2016" = c(n_mutations = 0,total_individuals = 20),
  "Young 2019" = c(n_mutations = 1,75), # 3 more U2AF1 mutations in AML
  "Desai 2018" = c(n_mutations = 0,total_individuals = 181) # 6 U2AF1 mutations in MAL
)
p_u2af1_c_ch <- sum(sapply(u2af1_CH_list,function(x) x[1]),na.rm = T) / sum(sapply(u2af1_CH_list,function(x) x[2]))
p_u2af1_c_mds <- 8.73/100
p_u2af1_c_cml <- 6.67/100
p_u2af1_c_aml <- 4.63/100
p_ch <- c(0.9/100,20/100)
p_mds <- 2 / 1e5
p_cml <- 1.6 / 1e5
p_aml <- 4.3 / 1e5

# marginalise P(U2AF1)
p_u2af1 <- p_u2af1_c_ch * p_ch + p_u2af1_c_mds * p_mds + p_u2af1_c_aml * p_aml + p_u2af1_c_cml * p_cml

load_data() %>% 
  select(Gene,SardID) %>% 
  distinct %>% 
  summarise(`U2AF1 incidence in our cohort` = length(unique(SardID[Gene == 'U2AF1'])) / length(unique(SardID)),
            `Expected U2AF1 incidence for individuals over 70` = p_u2af1[2]) 
```

The incidence of U2AF1 mutations in our cohort is quite higher than what would be expected in a sample of individuals over 70. As such, we can, to some extent, validate our findings regarding the late appearance of most U2AF1 clones.

## Survival analysis 

```{r survival_analysis,fig.height=6.5,fig.width=12,dpi=300}
library(survival)
library(survminer)
# quick_plot_survival <- function(s,) {
#   s <- survfit(s)
#   summary_s <- summary(s)
#   data_plot <- data.frame(
#     y = summary_s$surv,
#     ymin = summary_s$lower,
#     ymax = summary_s$upper,
#     x = summary_s$time
#   )
# }

comorbidity_data <- load_comorbidity_data()
smoking_data <- load_smoking_data() %>% 
  mutate(status = ifelse(QuitSmoking == 'N' & CurrentSmoker == 'N',"Never smoked",
                         ifelse(QuitSmoking == 'N' & CurrentSmoker == 'Y',"Smoker",
                                "Previous smoker"))) %>% 
  subset(!is.na(CurrentSmoker)) %>% 
  group_by(SardID) %>%
  summarise(
    status = ifelse(
      sum(status == 'Smoker') > 0,'During',
      ifelse(sum(status == 'Previous smoker') > 0, 'Before',
             'Never')
    )
  ) %>% 
  mutate(status = factor(status,levels = c("Never","Before","During"))) %>%
  mutate(EverSmoked = ifelse(status == "Never","No","Yes"))

blood_parameter_ltp <- load_blood_count_data() %>%
  group_by(SardID) %>% 
  filter(Phase == max(Phase))

grouping_data_size <- load_data() %>% 
  group_by(SardID) %>% 
  filter(Phase == max(Phase)) %>% 
  group_by(SardID) %>% 
  summarise(
    LargestCloneVAF = max(VAF),
    NClones = length(unique(AAChange.refGene[!is.na(AAChange.refGene)])),
    LastTimePoint = min(Age),
    Gender = unique(Gender)) %>%
  mutate(HasLargeClone=LargestCloneVAF>0.025) %>%
  distinct()
grouping_data_dynamics <- r_values %>% 
  select(SardID=individual,site,b_clone,coefficient) %>%
  group_by(SardID) %>% 
  summarise(FastestClone=max(b_clone+coefficient),
            FastestGene=max(coefficient)) %>% 
  distinct() 
 
survival_data <- load_survival_data() %>%
  as_tibble() %>%
  subset(DEAD != "U") %>%
  subset(gtools::na.replace(as.character(AGE_of_death) != 'Unknown',T)) %>%
  mutate(Age_General = as.numeric(gtools::na.replace(AGE_of_death,0))+as.numeric(gtools::na.replace(AGE_atStudyEnd,0))) %>%
  mutate(DEAD = ifelse(DEAD=='Y',1,0))
full_survival_data <- merge(grouping_data_size,survival_data,by = "SardID",all=T) %>%
  merge(grouping_data_dynamics,by = "SardID",all=T) %>%
  merge(smoking_data,by = "SardID") %>%
  merge(comorbidity_data,by = "SardID") %>% 
  mutate(Diabetes2Status = ifelse(DIABETES_2,"Yes","No"),
         HadCancer = ifelse(CANCER,"Yes","No")) %>% 
  mutate(DEAD=as.numeric(DEAD),
         Age_General=as.numeric(Age_General),
         FastestClone=ifelse(is.na(FastestClone),0,FastestClone),
         FastestGene=ifelse(is.na(FastestGene),0,FastestGene),
         HasLargeClone=ifelse(is.na(HasLargeClone),F,HasLargeClone),
         NClones=ifelse(is.na(NClones),0,NClones),
         LargestCloneVAF=ifelse(is.na(LargestCloneVAF),0,LargestCloneVAF)) %>% 
  mutate(NClones_Factor = ifelse(
    NClones >= 3,
    "3+",
    ifelse(NClones > 0,"1-2",NClones))) %>% 
  mutate(NoClones=NClones_Factor=='0',
         OneTwoClones=NClones_Factor=='1-2',
         ThreeMoreClones=NClones_Factor=='3+') %>% 
  mutate(FastClone = ifelse(FastestClone >= 0.1,">=0.1","<0.1")) %>%
  filter(!is.na(Age_General)) %>%
  mutate(Age_General = Age_General - LastTimePoint) %>% 
  merge(blood_parameter_ltp,by = 'SardID') %>% 
  subset(!(SardID %in% c(load_excluded_individuals(),load_excluded_individuals_lymph())))

model.base <- coxph(
  Surv(Age_General,DEAD) ~ LastTimePoint + Gender + FastestClone + NClones + LargestCloneVAF + Diabetes2Status + EverSmoked,
  x=T,y=T,model=T,
  data=full_survival_data) 

forest_plot <- ggforest(model.base,
                        full_survival_data,
                        main = "",refLabel = "")

plot_size <- ggsurvplot(
  coxph(
    Surv(Age_General,DEAD) ~ LastTimePoint + Gender + FastestClone + NClones + LargestCloneVAF + strata(HasLargeClone),
    x=T,y=T,model=T,
    data=full_survival_data) %>% survfit,
  data = full_survival_data
)$plot + 
  scale_color_lancet(name="Clone size",labels=c("<2.5%",">=2.5%"))

plot_number <- ggsurvplot(
  coxph(
    Surv(Age_General,DEAD) ~ LastTimePoint + Gender + FastestClone + NClones + LargestCloneVAF + strata(NClones_Factor),
    x=T,y=T,model=T,
    data=full_survival_data) %>% survfit,
  data = full_survival_data
)$plot + 
  scale_color_uchicago(name="Number of clones")

plot_dynamics <- ggsurvplot(
  coxph(
    Surv(Age_General,DEAD) ~ LastTimePoint + Gender + FastestClone + NClones + LargestCloneVAF + strata(FastClone),
    x=T,y=T,model=T,
    data=full_survival_data) %>% survfit,
  data = full_survival_data
)$plot + 
  scale_color_d3(name = "Mutation-independent growth coefficient")

plot_grid(
  plot_grid(plot_size,plot_number,plot_dynamics,nrow = 1),
  plot_grid(ggplot()+theme_void(),forest_plot+theme_gerstung(),ggplot()+theme_void(),
            rel_widths = c(0.3,0.8,0.3),nrow = 1),
  ncol = 1,
  rel_heights = c(1,1.2)
) + 
  ggsave(useDingbats=FALSE,sprintf("figures/%s/survival_analysis/general_figure.pdf",model_id),
         height = 5,width = 12)

```

This survival analysis shows fairly clearly that there does not seem to be any strong association with dynamic parameters. This may be due to the small sample size, but we are still not able to replicate results by others regarding the number of clones and clone size. We suspect this may be due to the fact that our cohort is fairly older than those previously used (the median age during the first time point was of 70.7 years). 

# Session details

```{r session_details}
cat(system("git log -n1", intern=TRUE), sep="\n")
```

```{r session_objects}
l <- ls()
data.frame(variable=l, Reduce("rbind",lapply(l, function(x) data.frame(classes=paste(class(get(x)),collapse = ','), size=format(object.size(get(x)), units="auto")))))
```

```{r session_packages}
sessionInfo()
```